<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Week 5 Bioinformatics with Bioconductor | 40 Days and 40 Nights</title>
  <meta name="description" content="An ad hoc course to learn R for bioinformatics, using the
COVID-19 epidemic as an excuse." />
  <meta name="generator" content="bookdown 0.18 and GitBook 2.6.7" />

  <meta property="og:title" content="Week 5 Bioinformatics with Bioconductor | 40 Days and 40 Nights" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An ad hoc course to learn R for bioinformatics, using the
COVID-19 epidemic as an excuse." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Week 5 Bioinformatics with Bioconductor | 40 Days and 40 Nights" />
  
  <meta name="twitter:description" content="An ad hoc course to learn R for bioinformatics, using the
COVID-19 epidemic as an excuse." />
  

<meta name="author" content="Martin Morgan" />
<meta name="author" content="L. Shawn Matott" />


<meta name="date" content="2020-05-01" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="four.html"/>
<link rel="next" href="six.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">QuaRantine</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Motivation</a><ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#introduction"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#what-to-expect"><i class="fa fa-check"></i>What to expect</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="one.html"><a href="one.html"><i class="fa fa-check"></i><b>1</b> Basics</a><ul>
<li class="chapter" data-level="1.1" data-path="one.html"><a href="one.html#day-1-monday-zoom-orientation"><i class="fa fa-check"></i><b>1.1</b> Day 1 (Monday) Zoom orientation</a><ul>
<li class="chapter" data-level="1.1.1" data-path="one.html"><a href="one.html#logistics-10-minutes"><i class="fa fa-check"></i><b>1.1.1</b> Logistics (10 minutes)</a></li>
<li class="chapter" data-level="1.1.2" data-path="one.html"><a href="one.html#installing-r-and-rstudio-25-minutes-shawn"><i class="fa fa-check"></i><b>1.1.2</b> Installing <em>R</em> and <em>RStudio</em> (25 minutes, Shawn)</a></li>
<li class="chapter" data-level="1.1.3" data-path="one.html"><a href="one.html#basics-of-r-25-minutes"><i class="fa fa-check"></i><b>1.1.3</b> Basics of <em>R</em> (25 minutes)</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="one.html"><a href="one.html#day-2-vectors-and-variables"><i class="fa fa-check"></i><b>1.2</b> Day 2: Vectors and variables</a></li>
<li class="chapter" data-level="1.3" data-path="one.html"><a href="one.html#day-3-factor-date-and-na"><i class="fa fa-check"></i><b>1.3</b> Day 3: <code>factor()</code>, <code>Date()</code>, and <code>NA</code></a></li>
<li class="chapter" data-level="1.4" data-path="one.html"><a href="one.html#day-4-working-with-variables"><i class="fa fa-check"></i><b>1.4</b> Day 4: Working with variables</a></li>
<li class="chapter" data-level="1.5" data-path="one.html"><a href="one.html#day-5-friday-zoom-check-in"><i class="fa fa-check"></i><b>1.5</b> Day 5 (Friday) Zoom check-in</a><ul>
<li class="chapter" data-level="1.5.1" data-path="one.html"><a href="one.html#logistics"><i class="fa fa-check"></i><b>1.5.1</b> Logistics</a></li>
<li class="chapter" data-level="1.5.2" data-path="one.html"><a href="one.html#review-and-trouble-shoot-25-minutes-martin"><i class="fa fa-check"></i><b>1.5.2</b> Review and trouble shoot (25 minutes; Martin)</a></li>
<li class="chapter" data-level="1.5.3" data-path="one.html"><a href="one.html#weekend-activities-25-minutes-shawn"><i class="fa fa-check"></i><b>1.5.3</b> Weekend activities (25 minutes; Shawn)</a></li>
</ul></li>
<li class="chapter" data-level="1.6" data-path="one.html"><a href="one.html#day-6-r-scripts"><i class="fa fa-check"></i><b>1.6</b> Day 6: <em>R</em> scripts</a></li>
<li class="chapter" data-level="1.7" data-path="one.html"><a href="one.html#day-7-saving-data"><i class="fa fa-check"></i><b>1.7</b> Day 7: Saving data</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="two.html"><a href="two.html"><i class="fa fa-check"></i><b>2</b> The data frame</a><ul>
<li class="chapter" data-level="2.1" data-path="two.html"><a href="two.html#day-8-monday-zoom-check-in"><i class="fa fa-check"></i><b>2.1</b> Day 8 (Monday) Zoom check-in</a><ul>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#logistics-1"><i class="fa fa-check"></i>Logistics</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#review-and-troubleshoot-15-minutes"><i class="fa fa-check"></i>Review and troubleshoot (15 minutes)</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#the-data-frame-40-minutes"><i class="fa fa-check"></i>The data frame (40 minutes)</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#this-weeks-activities-5-minutes"><i class="fa fa-check"></i>This week’s activities (5 minutes)</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="two.html"><a href="two.html#day-9-creation-and-manipulation"><i class="fa fa-check"></i><b>2.2</b> Day 9: Creation and manipulation</a><ul>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#creation"><i class="fa fa-check"></i>Creation</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#column-selection"><i class="fa fa-check"></i>Column selection</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#adding-or-updating-columns"><i class="fa fa-check"></i>Adding or updating columns</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#reading-and-writing"><i class="fa fa-check"></i>Reading and writing</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#reading-from-a-remote-file"><i class="fa fa-check"></i>Reading from a remote file (!)</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="two.html"><a href="two.html#day-10-subset-with-and-within"><i class="fa fa-check"></i><b>2.3</b> Day 10: <code>subset()</code>, <code>with()</code>, and <code>within()</code></a><ul>
<li><a href="two.html#subset"><code>subset()</code></a></li>
<li><a href="two.html#with"><code>with()</code></a></li>
<li><a href="two.html#within"><code>within()</code></a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="two.html"><a href="two.html#day-11-aggregate-and-an-initial-work-flow"><i class="fa fa-check"></i><b>2.4</b> Day 11: <code>aggregate()</code> and an initial work flow</a><ul>
<li><a href="two.html#aggregate-for-summarizing-columns-by-group"><code>aggregate()</code> for summarizing columns by group</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#an-initial-work-flow"><i class="fa fa-check"></i>An initial work flow</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="two.html"><a href="two.html#day-12-friday-zoom-check-in"><i class="fa fa-check"></i><b>2.5</b> Day 12 (Friday) Zoom check-in</a><ul>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#review-and-troubleshoot-20-minutes"><i class="fa fa-check"></i>Review and troubleshoot (20 minutes)</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#user-defined-functions"><i class="fa fa-check"></i>User-defined functions</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#plotting-data"><i class="fa fa-check"></i>Plotting data</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="two.html"><a href="two.html#day-13-basic-visualization"><i class="fa fa-check"></i><b>2.6</b> Day 13: Basic visualization</a></li>
<li class="chapter" data-level="2.7" data-path="two.html"><a href="two.html#day-14-functions"><i class="fa fa-check"></i><b>2.7</b> Day 14: Functions</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="three.html"><a href="three.html"><i class="fa fa-check"></i><b>3</b> Packages and the ‘tidyverse’</a><ul>
<li class="chapter" data-level="3.1" data-path="three.html"><a href="three.html#day-15-monday-zoom-check-in"><i class="fa fa-check"></i><b>3.1</b> Day 15 (Monday) Zoom check-in</a><ul>
<li class="chapter" data-level="" data-path="three.html"><a href="three.html#review-and-troubleshoot-15-minutes-1"><i class="fa fa-check"></i>Review and troubleshoot (15 minutes)</a></li>
<li class="chapter" data-level="" data-path="three.html"><a href="three.html#packages-20-minutes"><i class="fa fa-check"></i>Packages (20 minutes)</a></li>
<li class="chapter" data-level="" data-path="three.html"><a href="three.html#the-tidyverse-of-packages-20-minutes"><i class="fa fa-check"></i>The ‘tidyverse’ of packages (20 minutes)</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="three.html"><a href="three.html#day-16-key-tidyverse-packages-readr-and-dplyr"><i class="fa fa-check"></i><b>3.2</b> Day 16 Key tidyverse packages: [readr][] and [dplyr][]</a></li>
<li class="chapter" data-level="3.3" data-path="three.html"><a href="three.html#day-17-visualization-with-ggplot2"><i class="fa fa-check"></i><b>3.3</b> Day 17 Visualization with [ggplot2][]</a><ul>
<li class="chapter" data-level="" data-path="three.html"><a href="three.html#setup"><i class="fa fa-check"></i>Setup</a></li>
<li class="chapter" data-level="" data-path="three.html"><a href="three.html#ggplot2-essentials"><i class="fa fa-check"></i>[ggplot2][] essentials</a></li>
<li class="chapter" data-level="" data-path="three.html"><a href="three.html#covid-19-in-erie-county"><i class="fa fa-check"></i>COVID-19 in Erie county</a></li>
<li class="chapter" data-level="" data-path="three.html"><a href="three.html#covid-19-in-new-york-state"><i class="fa fa-check"></i>COVID-19 in New York State</a></li>
<li class="chapter" data-level="" data-path="three.html"><a href="three.html#covid-19-nationally"><i class="fa fa-check"></i>COVID-19 nationally</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="three.html"><a href="three.html#day-18-worldwide-covid-data"><i class="fa fa-check"></i><b>3.4</b> Day 18 Worldwide COVID data</a></li>
<li class="chapter" data-level="3.5" data-path="three.html"><a href="three.html#day-19-friday-zoom-check-in"><i class="fa fa-check"></i><b>3.5</b> Day 19 (Friday) Zoom check-in</a><ul>
<li class="chapter" data-level="3.5.1" data-path="three.html"><a href="three.html#logistics-2"><i class="fa fa-check"></i><b>3.5.1</b> Logistics</a></li>
<li class="chapter" data-level="3.5.2" data-path="three.html"><a href="three.html#review-and-trouble-shoot-40-minutes"><i class="fa fa-check"></i><b>3.5.2</b> Review and trouble shoot (40 minutes)</a></li>
<li class="chapter" data-level="3.5.3" data-path="three.html"><a href="three.html#weekend-activities-15-minutes"><i class="fa fa-check"></i><b>3.5.3</b> Weekend activities (15 minutes)</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="three.html"><a href="three.html#day-20-exploring-the-course-of-pandemic-in-different-regions"><i class="fa fa-check"></i><b>3.6</b> Day 20 Exploring the course of pandemic in different regions</a></li>
<li class="chapter" data-level="3.7" data-path="three.html"><a href="three.html#day-21-critical-evaluation"><i class="fa fa-check"></i><b>3.7</b> Day 21 Critical evaluation</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="four.html"><a href="four.html"><i class="fa fa-check"></i><b>4</b> Machine learning</a><ul>
<li class="chapter" data-level="4.1" data-path="four.html"><a href="four.html#day-22-monday-zoom-check-in"><i class="fa fa-check"></i><b>4.1</b> Day 22 (Monday) Zoom check-in</a><ul>
<li class="chapter" data-level="4.1.1" data-path="four.html"><a href="four.html#a-machine-learning-primer"><i class="fa fa-check"></i><b>4.1.1</b> A Machine Learning Primer</a></li>
<li class="chapter" data-level="4.1.2" data-path="four.html"><a href="four.html#the-random-forest-algorithm"><i class="fa fa-check"></i><b>4.1.2</b> The Random Forest Algorithm</a></li>
<li class="chapter" data-level="4.1.3" data-path="four.html"><a href="four.html#a-covid-19-dataset-for-machine-learning"><i class="fa fa-check"></i><b>4.1.3</b> A COVID-19 Dataset for Machine Learning</a></li>
<li class="chapter" data-level="4.1.4" data-path="four.html"><a href="four.html#a-random-forest-example-using-covid-19-data"><i class="fa fa-check"></i><b>4.1.4</b> A Random Forest Example Using COVID-19 Data</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="four.html"><a href="four.html#day-23---support-vector-machines"><i class="fa fa-check"></i><b>4.2</b> Day 23 - Support Vector Machines</a><ul>
<li class="chapter" data-level="4.2.1" data-path="four.html"><a href="four.html#an-svm-example-using-covid-19"><i class="fa fa-check"></i><b>4.2.1</b> An SVM Example using COVID-19</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="four.html"><a href="four.html#day-24---the-k-nearest-neighbors-algorithm"><i class="fa fa-check"></i><b>4.3</b> Day 24 - the <span class="math inline">\(k\)</span>-Nearest Neighbors Algorithm</a></li>
<li class="chapter" data-level="4.4" data-path="four.html"><a href="four.html#day-25---artificial-neural-networks"><i class="fa fa-check"></i><b>4.4</b> Day 25 - Artificial Neural Networks</a><ul>
<li class="chapter" data-level="4.4.1" data-path="four.html"><a href="four.html#introduction-to-artifical-neural-networks"><i class="fa fa-check"></i><b>4.4.1</b> Introduction to Artifical Neural Networks</a></li>
<li class="chapter" data-level="4.4.2" data-path="four.html"><a href="four.html#keras-and-tensorflow"><i class="fa fa-check"></i><b>4.4.2</b> Keras and TensorFlow</a></li>
<li class="chapter" data-level="4.4.3" data-path="four.html"><a href="four.html#applying-keras-and-tensorflow-to-the-covid-19-dataset"><i class="fa fa-check"></i><b>4.4.3</b> Applying Keras and TensorFlow to the COVID-19 Dataset</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="four.html"><a href="four.html#day-26-friday-zoom-check-in"><i class="fa fa-check"></i><b>4.5</b> Day 26 (Friday) Zoom check-in</a><ul>
<li class="chapter" data-level="4.5.1" data-path="four.html"><a href="four.html#review-and-trouble-shoot-25-minutes"><i class="fa fa-check"></i><b>4.5.1</b> Review and trouble shoot (25 minutes)</a></li>
<li class="chapter" data-level="4.5.2" data-path="four.html"><a href="four.html#this-weekend-25-minutes"><i class="fa fa-check"></i><b>4.5.2</b> This weekend (25 minutes)</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="four.html"><a href="four.html#day-27"><i class="fa fa-check"></i><b>4.6</b> Day 27</a></li>
<li class="chapter" data-level="4.7" data-path="four.html"><a href="four.html#day-28"><i class="fa fa-check"></i><b>4.7</b> Day 28</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="five.html"><a href="five.html"><i class="fa fa-check"></i><b>5</b> Bioinformatics with <span>Bioconductor</span></a><ul>
<li class="chapter" data-level="5.1" data-path="five.html"><a href="five.html#day-29-monday-zoom-check-in"><i class="fa fa-check"></i><b>5.1</b> Day 29 (Monday) Zoom check-in</a><ul>
<li class="chapter" data-level="5.1.1" data-path="five.html"><a href="five.html#review-and-trouble-shoot"><i class="fa fa-check"></i><b>5.1.1</b> Review and trouble shoot</a></li>
<li class="chapter" data-level="5.1.2" data-path="five.html"><a href="five.html#bioconductor-packages-and-biological-data"><i class="fa fa-check"></i><b>5.1.2</b> <span>Bioconductor</span>, packages, and biological data</a></li>
<li class="chapter" data-level="5.1.3" data-path="five.html"><a href="five.html#this-week-sars-cov-2-sequence-and-human-tissue-specific-gene-expression-data"><i class="fa fa-check"></i><b>5.1.3</b> This week: SARS-CoV-2 sequence and human tissue-specific gene expression data</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="five.html"><a href="five.html#day-30-dna-sequences-and-annotations"><i class="fa fa-check"></i><b>5.2</b> Day 30 DNA sequences and annotations</a><ul>
<li class="chapter" data-level="5.2.1" data-path="five.html"><a href="five.html#representing-dna-sequences"><i class="fa fa-check"></i><b>5.2.1</b> Representing DNA sequences</a></li>
<li class="chapter" data-level="5.2.2" data-path="five.html"><a href="five.html#working-with-genbankrecord-objects"><i class="fa fa-check"></i><b>5.2.2</b> Working with <code>GenBankRecord</code> objects</a></li>
<li class="chapter" data-level="5.2.3" data-path="five.html"><a href="five.html#genomic-ranges-and-dna-sequences"><i class="fa fa-check"></i><b>5.2.3</b> Genomic ranges and DNA sequences</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="five.html"><a href="five.html#day-31-sequence-alignment-and-visualization"><i class="fa fa-check"></i><b>5.3</b> Day 31 Sequence alignment and visualization</a><ul>
<li class="chapter" data-level="5.3.1" data-path="five.html"><a href="five.html#sequence-alignment"><i class="fa fa-check"></i><b>5.3.1</b> Sequence alignment</a></li>
<li class="chapter" data-level="5.3.2" data-path="five.html"><a href="five.html#phylogenetic-tree-visualization"><i class="fa fa-check"></i><b>5.3.2</b> Phylogenetic tree visualization</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="five.html"><a href="five.html#day-32-single-cell-expression-data"><i class="fa fa-check"></i><b>5.4</b> Day 32 Single-cell expression data</a></li>
<li class="chapter" data-level="5.5" data-path="five.html"><a href="five.html#day-33-friday-zoom-check-in"><i class="fa fa-check"></i><b>5.5</b> Day 33 (Friday) Zoom check-in</a><ul>
<li class="chapter" data-level="5.5.1" data-path="five.html"><a href="five.html#review-and-trouble-shoot-25-minutes-1"><i class="fa fa-check"></i><b>5.5.1</b> Review and trouble shoot (25 minutes)</a></li>
<li class="chapter" data-level="5.5.2" data-path="five.html"><a href="five.html#next-week-25-minutes"><i class="fa fa-check"></i><b>5.5.2</b> Next week (25 minutes)</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="five.html"><a href="five.html#day-34"><i class="fa fa-check"></i><b>5.6</b> Day 34</a></li>
<li class="chapter" data-level="5.7" data-path="five.html"><a href="five.html#day-35"><i class="fa fa-check"></i><b>5.7</b> Day 35</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="six.html"><a href="six.html"><i class="fa fa-check"></i><b>6</b> Collaboration</a><ul>
<li class="chapter" data-level="6.1" data-path="six.html"><a href="six.html#days-monday-zoom-check-in"><i class="fa fa-check"></i><b>6.1</b> 5 Days (Monday) Zoom check-in</a></li>
<li class="chapter" data-level="6.2" data-path="six.html"><a href="six.html#days"><i class="fa fa-check"></i><b>6.2</b> 4 Days</a></li>
<li class="chapter" data-level="6.3" data-path="six.html"><a href="six.html#days-1"><i class="fa fa-check"></i><b>6.3</b> 3 Days</a></li>
<li class="chapter" data-level="6.4" data-path="six.html"><a href="six.html#days-2"><i class="fa fa-check"></i><b>6.4</b> 2 Days</a></li>
<li class="chapter" data-level="6.5" data-path="six.html"><a href="six.html#today-friday-zoom-check-in"><i class="fa fa-check"></i><b>6.5</b> Today! (Friday) Zoom check-in</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="frequently-asked-questions.html"><a href="frequently-asked-questions.html"><i class="fa fa-check"></i>Frequently asked questions</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">40 Days and 40 Nights</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="five" class="section level1">
<h1><span class="header-section-number">Week 5</span> Bioinformatics with <a href="https://bioconductor.org">Bioconductor</a></h1>
<div id="day-29-monday-zoom-check-in" class="section level2">
<h2><span class="header-section-number">5.1</span> Day 29 (Monday) Zoom check-in</h2>
<div id="review-and-trouble-shoot" class="section level3">
<h3><span class="header-section-number">5.1.1</span> Review and trouble shoot</h3>
</div>
<div id="bioconductor-packages-and-biological-data" class="section level3">
<h3><span class="header-section-number">5.1.2</span> <a href="https://bioconductor.org">Bioconductor</a>, packages, and biological data</h3>
<p><a href="https://bioconductor.org">Bioconductor</a></p>
<ul>
<li><p>Repository of more than 1900 packages for the ‘analysis and comprehension of high throughput genomic data’</p></li>
<li><p>Bulk and single cell RNASeq; ChIP and other gene regulation, called variants, flow cytometery, proteomics, …</p></li>
<li><p>Package discovery using <a href="https://bioconductor.org/packages">biocViews</a> and <a href="https://bioconductor.org/packages/release/BiocViews.html#___Workflow">workflows</a></p></li>
<li><p>Landing pages &amp; vignettes</p></li>
</ul>
<p>Packages</p>
<ul>
<li><p>One-time installation of <a href="https://cran.r-project.org/package=BiocManager">BiocManager</a></p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>))</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">repos =</span> <span class="st">&quot;https://cran.r-project.org&quot;</span>)</a></code></pre></div></li>
<li><p>Install <em>Bioconductor</em> or <em>CRAN</em> packages</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co">## Biostrings and genbankr are Bioconductor packages, rentrez is from CRAN</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2">BiocMangaer<span class="op">::</span><span class="kw">install</span>(<span class="kw">c</span>(<span class="st">&quot;Biostrings&quot;</span>, <span class="st">&quot;rentrez&quot;</span>))</a></code></pre></div></li>
<li><p>Validate installation</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">BiocManager<span class="op">::</span><span class="kw">version</span>() <span class="co"># current release is 3.11; needs R 4.0</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="co">## [1] &#39;3.10&#39;</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3">BiocManager<span class="op">::</span><span class="kw">valid</span>()</a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="co">## [1] TRUE</span></a></code></pre></div></li>
<li><p>Types of packages</p>
<ul>
<li>Software (e.g., <a href="https://bioconductor.org/packages/Biostrings">Biostrings</a>, <a href="https://bioconductor.org/packages/SingleCellExperiment">SingleCellExperiment</a>)</li>
<li>Annotation (e.g., <a href="https://bioconductor.org/packages/org.Hs.eg.db">org.Hs.eg.db</a>, <a href="https://bioconductor.org/packages/TxDb.Hsapiens.UCSC.hg38.knownGene">TxDb.Hsapiens.UCSC.hg38.knownGene</a>, <a href="https://bioconductor.org/packages/GO.db">GO.db</a>)</li>
<li>Experiment data (e.g., <a href="https://bioconductor.org/packages/airway">airway</a>, <a href="https://bioconductor.org/packages/scRNAseq">scRNAseq</a>)</li>
</ul></li>
</ul>
<p>Key packages</p>
<ul>
<li><a href="https://bioconductor.org/packages/Biostrings">Biostrings</a> for representing sequence data</li>
<li><a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a> for working with genomic coordinates</li>
<li><a href="https://bioconductor.org/packages/Summarizedexperiment">SummarizedExperiment</a> and <a href="https://bioconductor.org/packages/SingleCellExperiment">SingleCellExperiment</a> for representing experimental summaries, e.g., a gene x sample matrix of RNA-seq expression values.</li>
</ul>
<p>‘Class’ and ‘method’</p>
<ul>
<li><p>Classes represent data in ways that allow the specialized nature of the data to be exploited, e.g., it makes sense to calculate the <code>reverseComplement()</code> of a DNA sequence, but not of an arbitrary character vector.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw">library</span>(Biostrings)</a>
<a class="sourceLine" id="cb4-2" data-line-number="2">sequences &lt;-<span class="st"> </span><span class="kw">c</span>(</a>
<a class="sourceLine" id="cb4-3" data-line-number="3">    <span class="dt">chr1 =</span> <span class="st">&quot;CTGACT&quot;</span>,</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">    <span class="dt">chr2 =</span> <span class="st">&quot;CTGGGAACT&quot;</span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5">)</a>
<a class="sourceLine" id="cb4-6" data-line-number="6">dna &lt;-<span class="st"> </span><span class="kw">DNAStringSet</span>(sequences)</a>
<a class="sourceLine" id="cb4-7" data-line-number="7">dna</a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="co">##   A DNAStringSet instance of length 2</span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9"><span class="co">##     width seq                                               names               </span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10"><span class="co">## [1]     6 CTGACT                                            chr1</span></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="co">## [2]     9 CTGGGAACT                                         chr2</span></a>
<a class="sourceLine" id="cb4-12" data-line-number="12"></a>
<a class="sourceLine" id="cb4-13" data-line-number="13"><span class="co">## fails! `T` not in the RNA alphabet</span></a>
<a class="sourceLine" id="cb4-14" data-line-number="14">result &lt;-<span class="st"> </span><span class="kw">try</span>( <span class="kw">RNAStringSet</span>(sequences) )</a>
<a class="sourceLine" id="cb4-15" data-line-number="15"><span class="co">## Error in .Call2(&quot;new_XStringSet_from_CHARACTER&quot;, class(x0), elementType(x0),  : </span></a>
<a class="sourceLine" id="cb4-16" data-line-number="16"><span class="co">##   key 84 (char &#39;T&#39;) not in lookup table</span></a></code></pre></div></li>
<li><p>Methods represent the implementation of common (‘generic’) operations, specialized to the specific classes that the method operates on.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">length</span>(dna)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="co">## [1] 2</span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="kw">reverseComplement</span>(dna)</a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="co">##   A DNAStringSet instance of length 2</span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="co">##     width seq                                               names               </span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="co">## [1]     6 AGTCAG                                            chr1</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="co">## [2]     9 AGTTCCCAG                                         chr2</span></a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="kw">translate</span>(dna)</a>
<a class="sourceLine" id="cb5-9" data-line-number="9"><span class="co">##   A AAStringSet instance of length 2</span></a>
<a class="sourceLine" id="cb5-10" data-line-number="10"><span class="co">##     width seq                                               names               </span></a>
<a class="sourceLine" id="cb5-11" data-line-number="11"><span class="co">## [1]     2 MT                                                chr1</span></a>
<a class="sourceLine" id="cb5-12" data-line-number="12"><span class="co">## [2]     3 MGT                                               chr2</span></a></code></pre></div>
<ul>
<li>E.g., one could <code>translate()</code> both a DNA sequence, and an RNA sequence, so there is both <code>translate,DNAStringSet-method</code> and <code>translate,RNAStringSet-method</code></li>
</ul></li>
<li><p>Not all operations are implemented as methods on a class</p>
<ul>
<li><p>E.g., <code>length()</code> is a concept shared by collections of DNA and RNA sequences, so the method is defined on a common ‘parent’ class</p></li>
<li><p>E.g., <code>dim()</code> of a two-dimensional object is sufficient to implement <code>nrow()</code> (as <code>dim()[1]</code>) and <code>ncol()</code> (as <code>dim()[2]</code>) so one could use a ‘plain old’ function <code>nrow &lt;- function(x) dim(x)[1]</code> for <code>nrow()</code> / <code>ncol()</code>, relying on a <code>dim()</code> generic and methods to provide specialized behavior.</p></li>
</ul></li>
</ul>
<p>Getting help</p>
<ul>
<li>Vignettes, e.g., <a href="">DESeq2 vignette</a>! <code>browseVignettes(package = &quot;DESeq2&quot;)</code></li>
<li>Help pages for classes, e.g., <code>?DNAStringSet</code>, <code>?GRanges</code></li>
<li><p>Discovering methods</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">methods</span>(<span class="st">&quot;reverse&quot;</span>)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw">methods</span>(<span class="dt">class =</span> <span class="st">&quot;DNAStringSet&quot;</span>)</a></code></pre></div></li>
<li><p>Looking at symbols made available by packages (after the package is attached to the <code>search()</code> path): <code>ls(&quot;package:GenomicRanges&quot;)</code></p></li>
</ul>
</div>
<div id="this-week-sars-cov-2-sequence-and-human-tissue-specific-gene-expression-data" class="section level3">
<h3><span class="header-section-number">5.1.3</span> This week: SARS-CoV-2 sequence and human tissue-specific gene expression data</h3>
<p>Biological background</p>
<ul>
<li><p>RNA virus (<a href="https://en.wikipedia.org/wiki/File:Coronavirus_virion_structure.svg">source</a>, license: <a href="https://creativecommons.org/licenses/by-sa/4.0/deed.en">CC BY-SA 4.0</a>)</p>
<p><img src="images/512px-Coronavirus_virion_structure.svg.png" width="60%" /></p></li>
<li><p>Key domains (<a href="https://en.wikipedia.org/wiki/File:SARS-CoV_genome_organization.jpg">source</a>, license: <a href="https://creativecommons.org/licenses/by/2.0/deed.en">CC BY 2</a>)</p>
<p><img src="images/SARS-CoV_genome_organization.jpg" width="100%" /></p></li>
</ul>
<p>Setup</p>
<ul>
<li><p>Some essential <em>Bioconductor</em> data structures</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw">library</span>(Biostrings)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="kw">library</span>(GenomicRanges)</a></code></pre></div></li>
<li><p>Additional software</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw">library</span>(rentrez)</a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="kw">library</span>(genbankr)</a>
<a class="sourceLine" id="cb8-3" data-line-number="3"><span class="kw">library</span>(DECIPHER)</a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="kw">library</span>(ggtree)</a></code></pre></div></li>
<li><p>Need to install software?</p>
<ul>
<li><p>Install <a href="https://cran.r-project.org/package=BiocManager">BiocManager</a></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>))</a>
<a class="sourceLine" id="cb9-2" data-line-number="2">    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">repos =</span> <span class="st">&quot;https://cran.r-project.org&quot;</span>)</a></code></pre></div></li>
<li><p>Install other <em>CRAN</em> or <em>Bioconductor</em> packges</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">pkgs &lt;-<span class="st"> </span><span class="kw">c</span>(</a>
<a class="sourceLine" id="cb10-2" data-line-number="2">    <span class="co">## packages needing installation, e.g.,</span></a>
<a class="sourceLine" id="cb10-3" data-line-number="3">    <span class="st">&quot;Biostrings&quot;</span>, <span class="st">&quot;GenomicRanges&quot;</span>, <span class="st">&quot;genbankr&quot;</span>, <span class="st">&quot;rentrez&quot;</span>, <span class="st">&quot;DECIPHER&quot;</span>,</a>
<a class="sourceLine" id="cb10-4" data-line-number="4">    <span class="st">&quot;ggtree&quot;</span></a>
<a class="sourceLine" id="cb10-5" data-line-number="5">)</a>
<a class="sourceLine" id="cb10-6" data-line-number="6">BiocManager<span class="op">::</span><span class="kw">install</span>(pkgs)</a>
<a class="sourceLine" id="cb10-7" data-line-number="7">BiocManager<span class="op">::</span><span class="kw">valid</span>()</a></code></pre></div></li>
</ul></li>
</ul>
<p><a href="https://bioconductor.org/packages/Biostrings">Biostrings</a> and <a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></p>
<ul>
<li><p><a href="https://bioconductor.org/packages/Biostrings">Biostrings</a>: representing DNA sequences</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">sequences &lt;-<span class="st"> </span><span class="kw">c</span>(</a>
<a class="sourceLine" id="cb11-2" data-line-number="2">    <span class="dt">chr1 =</span> <span class="st">&quot;CTGACT&quot;</span>,</a>
<a class="sourceLine" id="cb11-3" data-line-number="3">    <span class="dt">chr2 =</span> <span class="st">&quot;CTGGGAACT&quot;</span></a>
<a class="sourceLine" id="cb11-4" data-line-number="4">)</a>
<a class="sourceLine" id="cb11-5" data-line-number="5">dna &lt;-<span class="st"> </span><span class="kw">DNAStringSet</span>(sequences)</a>
<a class="sourceLine" id="cb11-6" data-line-number="6">dna</a>
<a class="sourceLine" id="cb11-7" data-line-number="7"><span class="co">##   A DNAStringSet instance of length 2</span></a>
<a class="sourceLine" id="cb11-8" data-line-number="8"><span class="co">##     width seq                                               names               </span></a>
<a class="sourceLine" id="cb11-9" data-line-number="9"><span class="co">## [1]     6 CTGACT                                            chr1</span></a>
<a class="sourceLine" id="cb11-10" data-line-number="10"><span class="co">## [2]     9 CTGGGAACT                                         chr2</span></a></code></pre></div></li>
<li><p><a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a>: annotations in genome space, e.g., create a variable representing ‘regions of interest’</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">regions_of_interest &lt;-<span class="st"> </span><span class="kw">c</span>(</a>
<a class="sourceLine" id="cb12-2" data-line-number="2">    <span class="st">&quot;chr1:1-3&quot;</span>,</a>
<a class="sourceLine" id="cb12-3" data-line-number="3">    <span class="st">&quot;chr1:3-6&quot;</span>,</a>
<a class="sourceLine" id="cb12-4" data-line-number="4">    <span class="st">&quot;chr2:4-6&quot;</span></a>
<a class="sourceLine" id="cb12-5" data-line-number="5">)</a>
<a class="sourceLine" id="cb12-6" data-line-number="6">roi &lt;-<span class="st"> </span><span class="kw">GRanges</span>(regions_of_interest)</a>
<a class="sourceLine" id="cb12-7" data-line-number="7">roi</a>
<a class="sourceLine" id="cb12-8" data-line-number="8"><span class="co">## GRanges object with 3 ranges and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb12-9" data-line-number="9"><span class="co">##       seqnames    ranges strand</span></a>
<a class="sourceLine" id="cb12-10" data-line-number="10"><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></a>
<a class="sourceLine" id="cb12-11" data-line-number="11"><span class="co">##   [1]     chr1       1-3      *</span></a>
<a class="sourceLine" id="cb12-12" data-line-number="12"><span class="co">##   [2]     chr1       3-6      *</span></a>
<a class="sourceLine" id="cb12-13" data-line-number="13"><span class="co">##   [3]     chr2       4-6      *</span></a>
<a class="sourceLine" id="cb12-14" data-line-number="14"><span class="co">##   -------</span></a>
<a class="sourceLine" id="cb12-15" data-line-number="15"><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></a></code></pre></div></li>
<li><p>Many fun operations, e.g., extract sequences of regions of interest</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="kw">getSeq</span>(dna, roi)</a>
<a class="sourceLine" id="cb13-2" data-line-number="2"><span class="co">##   A DNAStringSet instance of length 3</span></a>
<a class="sourceLine" id="cb13-3" data-line-number="3"><span class="co">##     width seq</span></a>
<a class="sourceLine" id="cb13-4" data-line-number="4"><span class="co">## [1]     3 CTG</span></a>
<a class="sourceLine" id="cb13-5" data-line-number="5"><span class="co">## [2]     4 GACT</span></a>
<a class="sourceLine" id="cb13-6" data-line-number="6"><span class="co">## [3]     3 GGA</span></a></code></pre></div></li>
</ul>
<p>Sequences and annotations</p>
<ul>
<li><p>Input FASTA files from New York State samples (it makes little biological sense to use just New York samples, but we’ll do it anyway!) to <code>DNAStringSet</code></p></li>
<li><p>Work the Genbank accession of the reference sequence</p>
<ul>
<li>DNA sequence</li>
<li>Coding sequence annotations</li>
<li>Coding sequences</li>
</ul></li>
</ul>
<p>Alignment and visualization</p>
<ul>
<li>Multiple alignment using <a href="https://bioconductor.org/packages/DECIPHER">DECIPHER</a></li>
<li>Visualization using <a href="https://bioconductor.org/packages/ggtree">ggtree</a></li>
</ul>
<p>Gene expression</p>
<ul>
<li>Retrieving example data from ExperimentHub</li>
<li>Annotating cell types</li>
</ul>
</div>
</div>
<div id="day-30-dna-sequences-and-annotations" class="section level2">
<h2><span class="header-section-number">5.2</span> Day 30 DNA sequences and annotations</h2>
<p>Set up a directory for working</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">workdir &lt;-<span class="st"> &quot;workdir&quot;</span></a>
<a class="sourceLine" id="cb14-2" data-line-number="2"><span class="cf">if</span> (<span class="op">!</span><span class="kw">dir.exists</span>(workdir))</a>
<a class="sourceLine" id="cb14-3" data-line-number="3">    <span class="kw">dir.create</span>(workdir)</a></code></pre></div>
<p>Make sure the following packages are installed</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw">library</span>(Biostrings)</a>
<a class="sourceLine" id="cb15-2" data-line-number="2"><span class="kw">library</span>(GenomicRanges)</a>
<a class="sourceLine" id="cb15-3" data-line-number="3"><span class="kw">library</span>(rentrez)</a>
<a class="sourceLine" id="cb15-4" data-line-number="4"><span class="kw">library</span>(genbankr)</a></code></pre></div>
<ul>
<li><p>Install necessary packages, if needed</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>))</a>
<a class="sourceLine" id="cb16-2" data-line-number="2">    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">repos =</span> <span class="st">&quot;https://cran.r-project.org&quot;</span>)</a>
<a class="sourceLine" id="cb16-3" data-line-number="3"></a>
<a class="sourceLine" id="cb16-4" data-line-number="4"><span class="co">## for example...</span></a>
<a class="sourceLine" id="cb16-5" data-line-number="5">pkgs &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Biostrings&quot;</span>, <span class="st">&quot;GenomicRanges&quot;</span>, <span class="st">&quot;rentrez&quot;</span>, <span class="st">&quot;genbankr&quot;</span>)</a>
<a class="sourceLine" id="cb16-6" data-line-number="6">BiocManager<span class="op">::</span><span class="kw">install</span>(pkgs)</a></code></pre></div></li>
</ul>
<div id="representing-dna-sequences" class="section level3">
<h3><span class="header-section-number">5.2.1</span> Representing DNA sequences</h3>
<p>Sample information</p>
<ul>
<li><p>Visit the <a href="https://www.ncbi.nlm.nih.gov/projects/genome/sars-cov-2-seqs">NCBI SARS-CoV-2</a> site for orientation. We’ll retrieve and analyse some of the samples sequenced in New York state</p></li>
<li><p>Read a csv file summarizing records available from the US NCBI.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">url &lt;-<span class="st"> &quot;https://raw.githubusercontent.com/mtmorgan/QuaRantine/master/assets/05-genbank-accessions.csv&quot;</span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2">csv_file &lt;-<span class="st"> </span><span class="kw">file.path</span>(workdir, <span class="kw">basename</span>(url))</a>
<a class="sourceLine" id="cb17-3" data-line-number="3"><span class="cf">if</span> ( <span class="op">!</span><span class="kw">file.exists</span>(csv_file) )</a>
<a class="sourceLine" id="cb17-4" data-line-number="4">    <span class="kw">download.file</span>(url, csv_file)</a>
<a class="sourceLine" id="cb17-5" data-line-number="5"></a>
<a class="sourceLine" id="cb17-6" data-line-number="6">records &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw">read_csv</span>(csv_file)</a>
<a class="sourceLine" id="cb17-7" data-line-number="7"><span class="co">## Parsed with column specification:</span></a>
<a class="sourceLine" id="cb17-8" data-line-number="8"><span class="co">## cols(</span></a>
<a class="sourceLine" id="cb17-9" data-line-number="9"><span class="co">##   accession = col_character(),</span></a>
<a class="sourceLine" id="cb17-10" data-line-number="10"><span class="co">##   link = col_character(),</span></a>
<a class="sourceLine" id="cb17-11" data-line-number="11"><span class="co">##   date = col_date(format = &quot;&quot;),</span></a>
<a class="sourceLine" id="cb17-12" data-line-number="12"><span class="co">##   country = col_character(),</span></a>
<a class="sourceLine" id="cb17-13" data-line-number="13"><span class="co">##   region = col_character()</span></a>
<a class="sourceLine" id="cb17-14" data-line-number="14"><span class="co">## )</span></a>
<a class="sourceLine" id="cb17-15" data-line-number="15">records</a>
<a class="sourceLine" id="cb17-16" data-line-number="16"><span class="co">## # A tibble: 2,434 x 5</span></a>
<a class="sourceLine" id="cb17-17" data-line-number="17"><span class="co">##    accession link                                    date       country   region</span></a>
<a class="sourceLine" id="cb17-18" data-line-number="18"><span class="co">##    &lt;chr&gt;     &lt;chr&gt;                                   &lt;date&gt;     &lt;chr&gt;     &lt;chr&gt; </span></a>
<a class="sourceLine" id="cb17-19" data-line-number="19"><span class="co">##  1 NC_045512 https://www.ncbi.nlm.nih.gov/nuccore/N… NA         China     &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb17-20" data-line-number="20"><span class="co">##  2 MT447189  https://www.ncbi.nlm.nih.gov/nuccore/M… 2020-03-02 Uzbekist… &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb17-21" data-line-number="21"><span class="co">##  3 MT447188  https://www.ncbi.nlm.nih.gov/nuccore/M… 2020-03-02 Uzbekist… &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb17-22" data-line-number="22"><span class="co">##  4 MT447177  https://www.ncbi.nlm.nih.gov/nuccore/M… 2020-03-26 Iran      &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb17-23" data-line-number="23"><span class="co">##  5 MT447176  https://www.ncbi.nlm.nih.gov/nuccore/M… 2020-03-20 Thailand  &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb17-24" data-line-number="24"><span class="co">##  6 MT447175  https://www.ncbi.nlm.nih.gov/nuccore/M… 2020-03-19 Thailand  &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb17-25" data-line-number="25"><span class="co">##  7 MT447174  https://www.ncbi.nlm.nih.gov/nuccore/M… 2020-03-01 Thailand  &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb17-26" data-line-number="26"><span class="co">##  8 MT447173  https://www.ncbi.nlm.nih.gov/nuccore/M… 2020-03-18 Thailand  &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb17-27" data-line-number="27"><span class="co">##  9 MT447172  https://www.ncbi.nlm.nih.gov/nuccore/M… 2020-03-18 Thailand  &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb17-28" data-line-number="28"><span class="co">## 10 MT447171  https://www.ncbi.nlm.nih.gov/nuccore/M… 2020-03-16 Thailand  &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb17-29" data-line-number="29"><span class="co">## # … with 2,424 more rows</span></a></code></pre></div></li>
</ul>
<p>FASTA sequence files</p>
<ul>
<li><p>Download a FASTA file containing DNA sequences of all samples sequenced in New York State (this doesn’t make much sense biologically!)</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">url &lt;-<span class="st"> &quot;https://raw.githubusercontent.com/mtmorgan/QuaRantine/master/assets/05-SARS-CoV2-NY.fasta&quot;</span></a>
<a class="sourceLine" id="cb18-2" data-line-number="2">fasta_file &lt;-<span class="st"> </span><span class="kw">file.path</span>(workdir, <span class="kw">basename</span>(url))</a>
<a class="sourceLine" id="cb18-3" data-line-number="3"><span class="cf">if</span> ( <span class="op">!</span><span class="kw">file.exists</span>(fasta_file) )</a>
<a class="sourceLine" id="cb18-4" data-line-number="4">    <span class="kw">download.file</span>(url, fasta_file)</a></code></pre></div></li>
<li><p>This is a plain text file with a simple format. Each DNA sequence is on a line starting with an identifier, <code>&gt;MT...</code>, and then lines representing the DNA (or RNA or amino acid) sequence.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="kw">readLines</span>(fasta_file, <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="co">## [1] &quot;&gt;MT434817&quot;                                                                       </span></a>
<a class="sourceLine" id="cb19-3" data-line-number="3"><span class="co">## [2] &quot;ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAA&quot;</span></a>
<a class="sourceLine" id="cb19-4" data-line-number="4"><span class="co">## [3] &quot;AATCTGTGTGGCTGTCACTCGGCTGCATGCTTAGTGCACTCACGCAGTATAATTAATAACTAATTACTGTCGTTGACAGG&quot;</span></a>
<a class="sourceLine" id="cb19-5" data-line-number="5"><span class="co">## [4] &quot;ACACGAGTAACTCGTCTATCTTCTGCAGGCTGCTTACGGTTTCGTCCGTGTTGCAGCCGATCATCAGCACATCTAGGTTT&quot;</span></a>
<a class="sourceLine" id="cb19-6" data-line-number="6"><span class="co">## [5] &quot;TGTCCGGGTGTGACCGAAAGGTAAGATGGAGAGCCTTGTCCCTGGTTTCAACGAGAAAACACACGTCCAACTCAGTTTGC&quot;</span></a></code></pre></div></li>
<li><p>Load the <a href="https://bioconductor.org/packages/Biostrings">Biostrings</a> package</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="kw">library</span>(Biostrings)</a>
<a class="sourceLine" id="cb20-2" data-line-number="2"><span class="kw">options</span>(<span class="dt">Biostrings.coloring =</span> <span class="ot">FALSE</span>)</a></code></pre></div></li>
<li><p>Read the DNA sequences and explore basic properties</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">dna &lt;-<span class="st"> </span><span class="kw">readDNAStringSet</span>(fasta_file)</a>
<a class="sourceLine" id="cb21-2" data-line-number="2">dna</a>
<a class="sourceLine" id="cb21-3" data-line-number="3"><span class="co">##   A DNAStringSet instance of length 245</span></a>
<a class="sourceLine" id="cb21-4" data-line-number="4"><span class="co">##       width seq                                             names               </span></a>
<a class="sourceLine" id="cb21-5" data-line-number="5"><span class="co">##   [1] 29873 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434817</span></a>
<a class="sourceLine" id="cb21-6" data-line-number="6"><span class="co">##   [2] 29879 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434816</span></a>
<a class="sourceLine" id="cb21-7" data-line-number="7"><span class="co">##   [3] 29873 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434815</span></a>
<a class="sourceLine" id="cb21-8" data-line-number="8"><span class="co">##   [4] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434814</span></a>
<a class="sourceLine" id="cb21-9" data-line-number="9"><span class="co">##   [5] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434813</span></a>
<a class="sourceLine" id="cb21-10" data-line-number="10"><span class="co">##   ...   ... ...</span></a>
<a class="sourceLine" id="cb21-11" data-line-number="11"><span class="co">## [241] 29724 GATCTGTTCTCTAAACGAACTT...TGTACAGTGAACAATGCTAGGG MT370833</span></a>
<a class="sourceLine" id="cb21-12" data-line-number="12"><span class="co">## [242] 29717 TCTAAACGAACTTTAAAATCTG...TACAGTGAACAATGCTAGGGAG MT370832</span></a>
<a class="sourceLine" id="cb21-13" data-line-number="13"><span class="co">## [243] 29716 TCTAAACGAACTTTAAAATCTG...GTACAGTGAACAATGCTAGGGA MT370831</span></a>
<a class="sourceLine" id="cb21-14" data-line-number="14"><span class="co">## [244] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT325627</span></a>
<a class="sourceLine" id="cb21-15" data-line-number="15"><span class="co">## [245] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT304486</span></a></code></pre></div></li>
<li><p>Perform simple manipulations</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="co">## how many records?</span></a>
<a class="sourceLine" id="cb22-2" data-line-number="2"><span class="kw">length</span>(dna)</a>
<a class="sourceLine" id="cb22-3" data-line-number="3"><span class="co">## [1] 245</span></a>
<a class="sourceLine" id="cb22-4" data-line-number="4"></a>
<a class="sourceLine" id="cb22-5" data-line-number="5"><span class="co">## subset</span></a>
<a class="sourceLine" id="cb22-6" data-line-number="6">dna[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a>
<a class="sourceLine" id="cb22-7" data-line-number="7"><span class="co">##   A DNAStringSet instance of length 3</span></a>
<a class="sourceLine" id="cb22-8" data-line-number="8"><span class="co">##     width seq                                               names               </span></a>
<a class="sourceLine" id="cb22-9" data-line-number="9"><span class="co">## [1] 29873 ATTAAAGGTTTATACCTTCCCAG...AGGAGAATGACAAAAAAAAAAAA MT434817</span></a>
<a class="sourceLine" id="cb22-10" data-line-number="10"><span class="co">## [2] 29879 ATTAAAGGTTTATACCTTCCCAG...AGGAGAATGACAAAAAAAAAAAA MT434816</span></a>
<a class="sourceLine" id="cb22-11" data-line-number="11"><span class="co">## [3] 29873 ATTAAAGGTTTATACCTTCCCAG...AGGAGAATGACAAAAAAAAAAAA MT434815</span></a>
<a class="sourceLine" id="cb22-12" data-line-number="12">dna[<span class="kw">c</span>(<span class="st">&#39;MT434817&#39;</span>, <span class="st">&#39;MT434815&#39;</span>)]</a>
<a class="sourceLine" id="cb22-13" data-line-number="13"><span class="co">##   A DNAStringSet instance of length 2</span></a>
<a class="sourceLine" id="cb22-14" data-line-number="14"><span class="co">##     width seq                                               names               </span></a>
<a class="sourceLine" id="cb22-15" data-line-number="15"><span class="co">## [1] 29873 ATTAAAGGTTTATACCTTCCCAG...AGGAGAATGACAAAAAAAAAAAA MT434817</span></a>
<a class="sourceLine" id="cb22-16" data-line-number="16"><span class="co">## [2] 29873 ATTAAAGGTTTATACCTTCCCAG...AGGAGAATGACAAAAAAAAAAAA MT434815</span></a>
<a class="sourceLine" id="cb22-17" data-line-number="17"></a>
<a class="sourceLine" id="cb22-18" data-line-number="18"><span class="co">## extract</span></a>
<a class="sourceLine" id="cb22-19" data-line-number="19">dna[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb22-20" data-line-number="20"><span class="co">##   29873-letter &quot;DNAString&quot; instance</span></a>
<a class="sourceLine" id="cb22-21" data-line-number="21"><span class="co">## seq: ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAAC...TTAATAGCTTCTTAGGAGAATGACAAAAAAAAAAAA</span></a>
<a class="sourceLine" id="cb22-22" data-line-number="22"></a>
<a class="sourceLine" id="cb22-23" data-line-number="23"><span class="co">## number of characters in each record</span></a>
<a class="sourceLine" id="cb22-24" data-line-number="24"><span class="kw">head</span>(<span class="kw">nchar</span>(dna))</a>
<a class="sourceLine" id="cb22-25" data-line-number="25"><span class="co">## [1] 29873 29879 29873 29882 29882 29882</span></a>
<a class="sourceLine" id="cb22-26" data-line-number="26"><span class="kw">hist</span>(<span class="kw">nchar</span>(dna)) <span class="co"># histogram of sequence lengths</span></a></code></pre></div>
<p><img src="05-Bioinformatics_files/figure-html/unnamed-chunk-24-1.png" width="672" /></p></li>
</ul>
</div>
<div id="working-with-genbankrecord-objects" class="section level3">
<h3><span class="header-section-number">5.2.2</span> Working with <code>GenBankRecord</code> objects</h3>
<p>Data is often represented in complicated formats that do not fit into the tidy ‘table’ format of a CSV file. An example is the information displayed on the <a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_045512">SARS-CoV-2 reference genome</a> web page.</p>
<p>Often ‘someone’ has written software to manipulate this data in a more convient way.</p>
<ul>
<li><p>Attach the <a href="https://cran.r-project.org/package=rentrez">rentrez</a> package, which provides a way to query the NCBI’s ‘Entrez’ collection of data bases. Use <code>BiocManager::install('rentrez')</code> if you need to install the package</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="kw">library</span>(rentrez)</a>
<a class="sourceLine" id="cb23-2" data-line-number="2"></a>
<a class="sourceLine" id="cb23-3" data-line-number="3"><span class="co">## who&#39;s responsible?</span></a>
<a class="sourceLine" id="cb23-4" data-line-number="4"><span class="kw">maintainer</span>(<span class="st">&quot;rentrez&quot;</span>)</a>
<a class="sourceLine" id="cb23-5" data-line-number="5"><span class="co">## [1] &quot;David Winter &lt;david.winter@gmail.com&gt;&quot;</span></a>
<a class="sourceLine" id="cb23-6" data-line-number="6"></a>
<a class="sourceLine" id="cb23-7" data-line-number="7"><span class="co">## how to cite?</span></a>
<a class="sourceLine" id="cb23-8" data-line-number="8"><span class="kw">citation</span>(<span class="st">&quot;rentrez&quot;</span>)</a>
<a class="sourceLine" id="cb23-9" data-line-number="9"><span class="co">## </span></a>
<a class="sourceLine" id="cb23-10" data-line-number="10"><span class="co">## To cite rentrez in publications use:</span></a>
<a class="sourceLine" id="cb23-11" data-line-number="11"><span class="co">## </span></a>
<a class="sourceLine" id="cb23-12" data-line-number="12"><span class="co">##   Winter, D. J. (2017) rentrez: an R package for the NCBI eUtils API</span></a>
<a class="sourceLine" id="cb23-13" data-line-number="13"><span class="co">##   The R Journal 9(2):520-526</span></a>
<a class="sourceLine" id="cb23-14" data-line-number="14"><span class="co">## </span></a>
<a class="sourceLine" id="cb23-15" data-line-number="15"><span class="co">## A BibTeX entry for LaTeX users is</span></a>
<a class="sourceLine" id="cb23-16" data-line-number="16"><span class="co">## </span></a>
<a class="sourceLine" id="cb23-17" data-line-number="17"><span class="co">##   @Article{,</span></a>
<a class="sourceLine" id="cb23-18" data-line-number="18"><span class="co">##     title = {{rentrez}: an R package for the NCBI eUtils API},</span></a>
<a class="sourceLine" id="cb23-19" data-line-number="19"><span class="co">##     author = {David J. Winter},</span></a>
<a class="sourceLine" id="cb23-20" data-line-number="20"><span class="co">##     journal = {The R Journal},</span></a>
<a class="sourceLine" id="cb23-21" data-line-number="21"><span class="co">##     year = {2017},</span></a>
<a class="sourceLine" id="cb23-22" data-line-number="22"><span class="co">##     volume = {9},</span></a>
<a class="sourceLine" id="cb23-23" data-line-number="23"><span class="co">##     issue = {2},</span></a>
<a class="sourceLine" id="cb23-24" data-line-number="24"><span class="co">##     pages = {520--526},</span></a>
<a class="sourceLine" id="cb23-25" data-line-number="25"><span class="co">##   }</span></a></code></pre></div></li>
<li><p>The accession number for the SARS-CoV-2 reference sequence is <code>NC_045512</code>. Use <code>entrez_search()</code> to search the NCBI ‘nuccore’ database to discover information about this accession.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1">accession &lt;-<span class="st"> &quot;NC_045512&quot;</span></a>
<a class="sourceLine" id="cb24-2" data-line-number="2">nuccore_record &lt;-<span class="st"> </span><span class="kw">entrez_search</span>(<span class="st">&quot;nuccore&quot;</span>, accession)</a>
<a class="sourceLine" id="cb24-3" data-line-number="3">id &lt;-<span class="st"> </span>nuccore_record<span class="op">$</span>ids</a>
<a class="sourceLine" id="cb24-4" data-line-number="4">id</a>
<a class="sourceLine" id="cb24-5" data-line-number="5"><span class="co">## [1] &quot;1798174254&quot;</span></a></code></pre></div></li>
<li><p>The <code>id</code> is an identifier that can be used to actually retrieve the genbank record</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1">gb_record &lt;-</a>
<a class="sourceLine" id="cb25-2" data-line-number="2"><span class="st">    </span><span class="kw">entrez_fetch</span>(<span class="st">&quot;nuccore&quot;</span>, id, <span class="dt">rettype =</span> <span class="st">&quot;gbwithparts&quot;</span>, <span class="dt">retmode =</span> <span class="st">&quot;text&quot;</span>)</a></code></pre></div></li>
<li><p>The record contains all the text in the record; if you like visualize it with</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="kw">cat</span>(gb_record)</a></code></pre></div></li>
</ul>
<p>Represent the data as <code>GenBankR</code> object</p>
<ul>
<li><p>Rather than try to ‘munge’ (e.g., copy and paste) relevant information from the GenBank record into R, use the <a href="https://bioconductor.org/packages/genbankr">genbankr</a> package. Again, install with <code>BiocManager::install(&quot;genbankr&quot;)</code> if necessary</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="kw">library</span>(genbankr)</a>
<a class="sourceLine" id="cb27-2" data-line-number="2"></a>
<a class="sourceLine" id="cb27-3" data-line-number="3"><span class="kw">maintainer</span>(<span class="st">&quot;genbankr&quot;</span>)</a>
<a class="sourceLine" id="cb27-4" data-line-number="4"><span class="co">## [1] &quot;Gabriel Becker &lt;becker.gabriel@gene.com&gt;&quot;</span></a>
<a class="sourceLine" id="cb27-5" data-line-number="5"></a>
<a class="sourceLine" id="cb27-6" data-line-number="6"><span class="kw">citation</span>(<span class="st">&quot;genbankr&quot;</span>)</a>
<a class="sourceLine" id="cb27-7" data-line-number="7"><span class="co">## </span></a>
<a class="sourceLine" id="cb27-8" data-line-number="8"><span class="co">## To cite package &#39;genbankr&#39; in publications use:</span></a>
<a class="sourceLine" id="cb27-9" data-line-number="9"><span class="co">## </span></a>
<a class="sourceLine" id="cb27-10" data-line-number="10"><span class="co">##   Gabriel Becker and Michael Lawrence (2019). genbankr: Parsing GenBank</span></a>
<a class="sourceLine" id="cb27-11" data-line-number="11"><span class="co">##   files into semantically useful objects. R package version 1.14.0.</span></a>
<a class="sourceLine" id="cb27-12" data-line-number="12"><span class="co">## </span></a>
<a class="sourceLine" id="cb27-13" data-line-number="13"><span class="co">## A BibTeX entry for LaTeX users is</span></a>
<a class="sourceLine" id="cb27-14" data-line-number="14"><span class="co">## </span></a>
<a class="sourceLine" id="cb27-15" data-line-number="15"><span class="co">##   @Manual{,</span></a>
<a class="sourceLine" id="cb27-16" data-line-number="16"><span class="co">##     title = {genbankr: Parsing GenBank files into semantically useful objects},</span></a>
<a class="sourceLine" id="cb27-17" data-line-number="17"><span class="co">##     author = {Gabriel Becker and Michael Lawrence},</span></a>
<a class="sourceLine" id="cb27-18" data-line-number="18"><span class="co">##     year = {2019},</span></a>
<a class="sourceLine" id="cb27-19" data-line-number="19"><span class="co">##     note = {R package version 1.14.0},</span></a>
<a class="sourceLine" id="cb27-20" data-line-number="20"><span class="co">##   }</span></a></code></pre></div></li>
<li><p>Use the function <code>readGenBank()</code> to parse the complicated text of <code>gb_record</code></p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1">gb &lt;-<span class="st"> </span><span class="kw">readGenBank</span>(<span class="dt">text =</span> gb_record)</a>
<a class="sourceLine" id="cb28-2" data-line-number="2">gb</a>
<a class="sourceLine" id="cb28-3" data-line-number="3"><span class="co">## GenBank Annotations</span></a>
<a class="sourceLine" id="cb28-4" data-line-number="4"><span class="co">## Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome. </span></a>
<a class="sourceLine" id="cb28-5" data-line-number="5"><span class="co">## Accession: NC_045512</span></a>
<a class="sourceLine" id="cb28-6" data-line-number="6"><span class="co">## 1 Sequence(s) with total length length: 29903</span></a>
<a class="sourceLine" id="cb28-7" data-line-number="7"><span class="co">## 11 genes</span></a>
<a class="sourceLine" id="cb28-8" data-line-number="8"><span class="co">## 12 transcripts</span></a>
<a class="sourceLine" id="cb28-9" data-line-number="9"><span class="co">## 13 exons/cds elements</span></a>
<a class="sourceLine" id="cb28-10" data-line-number="10"><span class="co">## 0 variations</span></a>
<a class="sourceLine" id="cb28-11" data-line-number="11"><span class="co">## 34 other features</span></a></code></pre></div></li>
<li><p>Extract the reference sequence from the object</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1">ref &lt;-<span class="st"> </span><span class="kw">getSeq</span>(gb)</a>
<a class="sourceLine" id="cb29-2" data-line-number="2">ref</a>
<a class="sourceLine" id="cb29-3" data-line-number="3"><span class="co">##   A DNAStringSet instance of length 1</span></a>
<a class="sourceLine" id="cb29-4" data-line-number="4"><span class="co">##     width seq                                               names               </span></a>
<a class="sourceLine" id="cb29-5" data-line-number="5"><span class="co">## [1] 29903 ATTAAAGGTTTATACCTTCCCAG...AAAAAAAAAAAAAAAAAAAAAAA Severe acute resp...</span></a></code></pre></div></li>
<li><p>Extract the coding sequences from the reference sequence; the <code>cds()</code> command is explained a little later.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1">cds &lt;-<span class="st"> </span><span class="kw">getSeq</span>(ref, <span class="kw">cds</span>(gb))</a>
<a class="sourceLine" id="cb30-2" data-line-number="2">cds</a>
<a class="sourceLine" id="cb30-3" data-line-number="3"><span class="co">##   A DNAStringSet instance of length 13</span></a>
<a class="sourceLine" id="cb30-4" data-line-number="4"><span class="co">##      width seq</span></a>
<a class="sourceLine" id="cb30-5" data-line-number="5"><span class="co">##  [1] 13203 ATGGAGAGCCTTGTCCCTGGTTTCAACGAGAAA...CTTCAGTCAGCTGATGCACAATCGTTTTTAAAC</span></a>
<a class="sourceLine" id="cb30-6" data-line-number="6"><span class="co">##  [2]  8088 CGGGTTTGCGGTGTAAGTGCAGCCCGTCTTACA...GTTATTTCTAGTGATGTTCTTGTTAACAACTAA</span></a>
<a class="sourceLine" id="cb30-7" data-line-number="7"><span class="co">##  [3] 13218 ATGGAGAGCCTTGTCCCTGGTTTCAACGAGAAA...GCACAATCGTTTTTAAACGGGTTTGCGGTGTAA</span></a>
<a class="sourceLine" id="cb30-8" data-line-number="8"><span class="co">##  [4]  3822 ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTC...GTGCTCAAAGGAGTCAAATTACATTACACATAA</span></a>
<a class="sourceLine" id="cb30-9" data-line-number="9"><span class="co">##  [5]   828 ATGGATTTGTTTATGAGAATCTTCACAATTGGA...GAACCGACGACGACTACTAGCGTGCCTTTGTAA</span></a>
<a class="sourceLine" id="cb30-10" data-line-number="10"><span class="co">##  ...   ... ...</span></a>
<a class="sourceLine" id="cb30-11" data-line-number="11"><span class="co">##  [9]   366 ATGAAAATTATTCTTTTCTTGGCACTGATAACA...CTTTGCTTCACACTCAAAAGAAAGACAGAATGA</span></a>
<a class="sourceLine" id="cb30-12" data-line-number="12"><span class="co">## [10]   132 ATGATTGAACTTTCATTAATTGACTTCTATTTG...CTGCAAGATCATAATGAAACTTGTCACGCCTAA</span></a>
<a class="sourceLine" id="cb30-13" data-line-number="13"><span class="co">## [11]   366 ATGAAATTTCTTGTTTTCTTAGGAATCATCACA...CATGACGTTCGTGTTGTTTTAGATTTCATCTAA</span></a>
<a class="sourceLine" id="cb30-14" data-line-number="14"><span class="co">## [12]  1260 ATGTCTGATAATGGACCCCAAAATCAGCGAAAT...TCCATGAGCAGTGCTGACTCAACTCAGGCCTAA</span></a>
<a class="sourceLine" id="cb30-15" data-line-number="15"><span class="co">## [13]   117 ATGGGCTATATAAACGTTTTCGCTTTTCCGTTT...CAAGTAGATGTAGTTAACTTTAATCTCACATAG</span></a></code></pre></div></li>
<li><p>There are 13 coding sequences in the genome. Translate these to their amino acid sequences</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" data-line-number="1"><span class="kw">translate</span>(cds)</a>
<a class="sourceLine" id="cb31-2" data-line-number="2"><span class="co">##   A AAStringSet instance of length 13</span></a>
<a class="sourceLine" id="cb31-3" data-line-number="3"><span class="co">##      width seq</span></a>
<a class="sourceLine" id="cb31-4" data-line-number="4"><span class="co">##  [1]  4401 MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGD...VCTVCGMWKGYGCSCDQLREPMLQSADAQSFLN</span></a>
<a class="sourceLine" id="cb31-5" data-line-number="5"><span class="co">##  [2]  2696 RVCGVSAARLTPCGTGTSTDVVYRAFDIYNDKV...INDMILSLLSKGRLIIRENNRVVISSDVLVNN*</span></a>
<a class="sourceLine" id="cb31-6" data-line-number="6"><span class="co">##  [3]  4406 MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGD...GMWKGYGCSCDQLREPMLQSADAQSFLNGFAV*</span></a>
<a class="sourceLine" id="cb31-7" data-line-number="7"><span class="co">##  [4]  1274 MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFT...SCLKGCCSCGSCCKFDEDDSEPVLKGVKLHYT*</span></a>
<a class="sourceLine" id="cb31-8" data-line-number="8"><span class="co">##  [5]   276 MDLFMRIFTIGTVTLKQGEIKDATPSDFVRATA...VQIHTIDGSSGVVNPVMEPIYDEPTTTTSVPL*</span></a>
<a class="sourceLine" id="cb31-9" data-line-number="9"><span class="co">##  ...   ... ...</span></a>
<a class="sourceLine" id="cb31-10" data-line-number="10"><span class="co">##  [9]   122 MKIILFLALITLATCELYHYQECVRGTTVLLKE...QEEVQELYSPIFLIVAAIVFITLCFTLKRKTE*</span></a>
<a class="sourceLine" id="cb31-11" data-line-number="11"><span class="co">## [10]    44 MIELSLIDFYLCFLAFLLFLVLIMLIIFWFSLELQDHNETCHA*</span></a>
<a class="sourceLine" id="cb31-12" data-line-number="12"><span class="co">## [11]   122 MKFLVFLGIITTVAAFHQECSLQSCTQHQPYVV...CQEPKLGSLVVRCSFYEDFLEYHDVRVVLDFI*</span></a>
<a class="sourceLine" id="cb31-13" data-line-number="13"><span class="co">## [12]   420 MSDNGPQNQRNAPRITFGGPSDSTGSNQNGERS...KQQTVTLLPAADLDDFSKQLQQSMSSADSTQA*</span></a>
<a class="sourceLine" id="cb31-14" data-line-number="14"><span class="co">## [13]    39 MGYINVFAFPFTIYSLLLCRMNSRNYIAQVDVVNFNLT*</span></a></code></pre></div></li>
</ul>
</div>
<div id="genomic-ranges-and-dna-sequences" class="section level3">
<h3><span class="header-section-number">5.2.3</span> Genomic ranges and DNA sequences</h3>
<p>Genomic ranges</p>
<ul>
<li><p>Attach the <a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a> package. If necessary, install with <code>BiocManager::install(&quot;GenomicRanges&quot;)</code></p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1"><span class="kw">library</span>(GenomicRanges)</a></code></pre></div></li>
<li><p>Extract the coordinates of the coding sequences contained in the GenBank accession</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1">cds &lt;-<span class="st"> </span><span class="kw">cds</span>(gb)</a></code></pre></div></li>
<li><p>Look at the genomic ranges defined in <code>cds</code></p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1"><span class="kw">granges</span>(cds)</a>
<a class="sourceLine" id="cb34-2" data-line-number="2"><span class="co">## GRanges object with 13 ranges and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb34-3" data-line-number="3"><span class="co">##                                              seqnames      ranges strand</span></a>
<a class="sourceLine" id="cb34-4" data-line-number="4"><span class="co">##                                                 &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;</span></a>
<a class="sourceLine" id="cb34-5" data-line-number="5"><span class="co">##    [1] Severe acute respiratory syndrome coronavirus2   266-13468      +</span></a>
<a class="sourceLine" id="cb34-6" data-line-number="6"><span class="co">##    [2] Severe acute respiratory syndrome coronavirus2 13468-21555      +</span></a>
<a class="sourceLine" id="cb34-7" data-line-number="7"><span class="co">##    [3] Severe acute respiratory syndrome coronavirus2   266-13483      +</span></a>
<a class="sourceLine" id="cb34-8" data-line-number="8"><span class="co">##    [4] Severe acute respiratory syndrome coronavirus2 21563-25384      +</span></a>
<a class="sourceLine" id="cb34-9" data-line-number="9"><span class="co">##    [5] Severe acute respiratory syndrome coronavirus2 25393-26220      +</span></a>
<a class="sourceLine" id="cb34-10" data-line-number="10"><span class="co">##    ...                                            ...         ...    ...</span></a>
<a class="sourceLine" id="cb34-11" data-line-number="11"><span class="co">##    [9] Severe acute respiratory syndrome coronavirus2 27394-27759      +</span></a>
<a class="sourceLine" id="cb34-12" data-line-number="12"><span class="co">##   [10] Severe acute respiratory syndrome coronavirus2 27756-27887      +</span></a>
<a class="sourceLine" id="cb34-13" data-line-number="13"><span class="co">##   [11] Severe acute respiratory syndrome coronavirus2 27894-28259      +</span></a>
<a class="sourceLine" id="cb34-14" data-line-number="14"><span class="co">##   [12] Severe acute respiratory syndrome coronavirus2 28274-29533      +</span></a>
<a class="sourceLine" id="cb34-15" data-line-number="15"><span class="co">##   [13] Severe acute respiratory syndrome coronavirus2 29558-29674      +</span></a>
<a class="sourceLine" id="cb34-16" data-line-number="16"><span class="co">##   -------</span></a>
<a class="sourceLine" id="cb34-17" data-line-number="17"><span class="co">##   seqinfo: 1 sequence from NC_045512.2 genome</span></a></code></pre></div></li>
<li><p>There are 13 coding sequences in the reference genome. The sequence is named ‘Severe acute respiratory syndrome coronavirus2’. The first region starts at position 266 of the reference genome, and goes to position 13468. The coding sequence is on the ‘+’ strand.</p></li>
<li><p>The full <code>cds</code> object has additional detail about each coding sequence. Display the full object (the display ‘wraps’ around several lines</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" data-line-number="1">cds</a>
<a class="sourceLine" id="cb35-2" data-line-number="2"><span class="co">## GRanges object with 13 ranges and 14 metadata columns:</span></a>
<a class="sourceLine" id="cb35-3" data-line-number="3"><span class="co">##                                              seqnames      ranges strand |</span></a>
<a class="sourceLine" id="cb35-4" data-line-number="4"><span class="co">##                                                 &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; |</span></a>
<a class="sourceLine" id="cb35-5" data-line-number="5"><span class="co">##    [1] Severe acute respiratory syndrome coronavirus2   266-13468      + |</span></a>
<a class="sourceLine" id="cb35-6" data-line-number="6"><span class="co">##    [2] Severe acute respiratory syndrome coronavirus2 13468-21555      + |</span></a>
<a class="sourceLine" id="cb35-7" data-line-number="7"><span class="co">##    [3] Severe acute respiratory syndrome coronavirus2   266-13483      + |</span></a>
<a class="sourceLine" id="cb35-8" data-line-number="8"><span class="co">##    [4] Severe acute respiratory syndrome coronavirus2 21563-25384      + |</span></a>
<a class="sourceLine" id="cb35-9" data-line-number="9"><span class="co">##    [5] Severe acute respiratory syndrome coronavirus2 25393-26220      + |</span></a>
<a class="sourceLine" id="cb35-10" data-line-number="10"><span class="co">##    ...                                            ...         ...    ... .</span></a>
<a class="sourceLine" id="cb35-11" data-line-number="11"><span class="co">##    [9] Severe acute respiratory syndrome coronavirus2 27394-27759      + |</span></a>
<a class="sourceLine" id="cb35-12" data-line-number="12"><span class="co">##   [10] Severe acute respiratory syndrome coronavirus2 27756-27887      + |</span></a>
<a class="sourceLine" id="cb35-13" data-line-number="13"><span class="co">##   [11] Severe acute respiratory syndrome coronavirus2 27894-28259      + |</span></a>
<a class="sourceLine" id="cb35-14" data-line-number="14"><span class="co">##   [12] Severe acute respiratory syndrome coronavirus2 28274-29533      + |</span></a>
<a class="sourceLine" id="cb35-15" data-line-number="15"><span class="co">##   [13] Severe acute respiratory syndrome coronavirus2 29558-29674      + |</span></a>
<a class="sourceLine" id="cb35-16" data-line-number="16"><span class="co">##               type        gene   locus_tag ribosomal_slippage</span></a>
<a class="sourceLine" id="cb35-17" data-line-number="17"><span class="co">##        &lt;character&gt; &lt;character&gt; &lt;character&gt;          &lt;logical&gt;</span></a>
<a class="sourceLine" id="cb35-18" data-line-number="18"><span class="co">##    [1]         CDS      ORF1ab  GU280_gp01               TRUE</span></a>
<a class="sourceLine" id="cb35-19" data-line-number="19"><span class="co">##    [2]         CDS      ORF1ab  GU280_gp01               TRUE</span></a>
<a class="sourceLine" id="cb35-20" data-line-number="20"><span class="co">##    [3]         CDS      ORF1ab  GU280_gp01              FALSE</span></a>
<a class="sourceLine" id="cb35-21" data-line-number="21"><span class="co">##    [4]         CDS           S  GU280_gp02              FALSE</span></a>
<a class="sourceLine" id="cb35-22" data-line-number="22"><span class="co">##    [5]         CDS       ORF3a  GU280_gp03              FALSE</span></a>
<a class="sourceLine" id="cb35-23" data-line-number="23"><span class="co">##    ...         ...         ...         ...                ...</span></a>
<a class="sourceLine" id="cb35-24" data-line-number="24"><span class="co">##    [9]         CDS       ORF7a  GU280_gp07              FALSE</span></a>
<a class="sourceLine" id="cb35-25" data-line-number="25"><span class="co">##   [10]         CDS       ORF7b  GU280_gp08              FALSE</span></a>
<a class="sourceLine" id="cb35-26" data-line-number="26"><span class="co">##   [11]         CDS        ORF8  GU280_gp09              FALSE</span></a>
<a class="sourceLine" id="cb35-27" data-line-number="27"><span class="co">##   [12]         CDS           N  GU280_gp10              FALSE</span></a>
<a class="sourceLine" id="cb35-28" data-line-number="28"><span class="co">##   [13]         CDS       ORF10  GU280_gp11              FALSE</span></a>
<a class="sourceLine" id="cb35-29" data-line-number="29"><span class="co">##                                                note codon_start</span></a>
<a class="sourceLine" id="cb35-30" data-line-number="30"><span class="co">##                                         &lt;character&gt;   &lt;numeric&gt;</span></a>
<a class="sourceLine" id="cb35-31" data-line-number="31"><span class="co">##    [1] pp1ab; translated by -1 ribosomal frameshift           1</span></a>
<a class="sourceLine" id="cb35-32" data-line-number="32"><span class="co">##    [2] pp1ab; translated by -1 ribosomal frameshift           1</span></a>
<a class="sourceLine" id="cb35-33" data-line-number="33"><span class="co">##    [3]                                         pp1a           1</span></a>
<a class="sourceLine" id="cb35-34" data-line-number="34"><span class="co">##    [4]            structural protein; spike protein           1</span></a>
<a class="sourceLine" id="cb35-35" data-line-number="35"><span class="co">##    [5]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb35-36" data-line-number="36"><span class="co">##    ...                                          ...         ...</span></a>
<a class="sourceLine" id="cb35-37" data-line-number="37"><span class="co">##    [9]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb35-38" data-line-number="38"><span class="co">##   [10]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb35-39" data-line-number="39"><span class="co">##   [11]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb35-40" data-line-number="40"><span class="co">##   [12]                     ORF9; structural protein           1</span></a>
<a class="sourceLine" id="cb35-41" data-line-number="41"><span class="co">##   [13]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb35-42" data-line-number="42"><span class="co">##                            product     protein_id         db_xref   translation</span></a>
<a class="sourceLine" id="cb35-43" data-line-number="43"><span class="co">##                        &lt;character&gt;    &lt;character&gt; &lt;CharacterList&gt; &lt;AAStringSet&gt;</span></a>
<a class="sourceLine" id="cb35-44" data-line-number="44"><span class="co">##    [1]          ORF1ab polyprotein YP_009724389.1 GeneID:43740578     MES...VNN</span></a>
<a class="sourceLine" id="cb35-45" data-line-number="45"><span class="co">##    [2]          ORF1ab polyprotein YP_009724389.1 GeneID:43740578     MES...VNN</span></a>
<a class="sourceLine" id="cb35-46" data-line-number="46"><span class="co">##    [3]           ORF1a polyprotein YP_009725295.1 GeneID:43740578     MES...FAV</span></a>
<a class="sourceLine" id="cb35-47" data-line-number="47"><span class="co">##    [4]        surface glycoprotein YP_009724390.1 GeneID:43740568     MFV...HYT</span></a>
<a class="sourceLine" id="cb35-48" data-line-number="48"><span class="co">##    [5]               ORF3a protein YP_009724391.1 GeneID:43740569     MDL...VPL</span></a>
<a class="sourceLine" id="cb35-49" data-line-number="49"><span class="co">##    ...                         ...            ...             ...           ...</span></a>
<a class="sourceLine" id="cb35-50" data-line-number="50"><span class="co">##    [9]               ORF7a protein YP_009724395.1 GeneID:43740573     MKI...KTE</span></a>
<a class="sourceLine" id="cb35-51" data-line-number="51"><span class="co">##   [10]                       ORF7b YP_009725318.1 GeneID:43740574     MIE...CHA</span></a>
<a class="sourceLine" id="cb35-52" data-line-number="52"><span class="co">##   [11]                ORF8 protein YP_009724396.1 GeneID:43740577     MKF...DFI</span></a>
<a class="sourceLine" id="cb35-53" data-line-number="53"><span class="co">##   [12] nucleocapsid phosphoprotein YP_009724397.2 GeneID:43740575     MSD...TQA</span></a>
<a class="sourceLine" id="cb35-54" data-line-number="54"><span class="co">##   [13]               ORF10 protein YP_009725255.1 GeneID:43740576     MGY...NLT</span></a>
<a class="sourceLine" id="cb35-55" data-line-number="55"><span class="co">##            loctype       gene_synonym     gene_id transcript_id</span></a>
<a class="sourceLine" id="cb35-56" data-line-number="56"><span class="co">##        &lt;character&gt;    &lt;CharacterList&gt; &lt;character&gt;   &lt;character&gt;</span></a>
<a class="sourceLine" id="cb35-57" data-line-number="57"><span class="co">##    [1]      normal                 NA  GU280_gp01  GU280_gp01.1</span></a>
<a class="sourceLine" id="cb35-58" data-line-number="58"><span class="co">##    [2]      normal                 NA  GU280_gp01  GU280_gp01.1</span></a>
<a class="sourceLine" id="cb35-59" data-line-number="59"><span class="co">##    [3]      normal                 NA  GU280_gp01  GU280_gp01.2</span></a>
<a class="sourceLine" id="cb35-60" data-line-number="60"><span class="co">##    [4]      normal spike glycoprotein  GU280_gp02  GU280_gp02.1</span></a>
<a class="sourceLine" id="cb35-61" data-line-number="61"><span class="co">##    [5]      normal                 NA  GU280_gp03  GU280_gp03.1</span></a>
<a class="sourceLine" id="cb35-62" data-line-number="62"><span class="co">##    ...         ...                ...         ...           ...</span></a>
<a class="sourceLine" id="cb35-63" data-line-number="63"><span class="co">##    [9]      normal                 NA  GU280_gp07  GU280_gp07.1</span></a>
<a class="sourceLine" id="cb35-64" data-line-number="64"><span class="co">##   [10]      normal                 NA  GU280_gp08  GU280_gp08.1</span></a>
<a class="sourceLine" id="cb35-65" data-line-number="65"><span class="co">##   [11]      normal                 NA  GU280_gp09  GU280_gp09.1</span></a>
<a class="sourceLine" id="cb35-66" data-line-number="66"><span class="co">##   [12]      normal                 NA  GU280_gp10  GU280_gp10.1</span></a>
<a class="sourceLine" id="cb35-67" data-line-number="67"><span class="co">##   [13]      normal                 NA  GU280_gp11  GU280_gp11.1</span></a>
<a class="sourceLine" id="cb35-68" data-line-number="68"><span class="co">##   -------</span></a>
<a class="sourceLine" id="cb35-69" data-line-number="69"><span class="co">##   seqinfo: 1 sequence from NC_045512.2 genome</span></a></code></pre></div>
<p>or coerce to a tibble and explore</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1"><span class="kw">library</span>(tibble)</a>
<a class="sourceLine" id="cb36-2" data-line-number="2"><span class="kw">library</span>(dplyr)</a>
<a class="sourceLine" id="cb36-3" data-line-number="3"><span class="co">## </span></a>
<a class="sourceLine" id="cb36-4" data-line-number="4"><span class="co">## Attaching package: &#39;dplyr&#39;</span></a>
<a class="sourceLine" id="cb36-5" data-line-number="5"><span class="co">## The following objects are masked from &#39;package:GenomicRanges&#39;:</span></a>
<a class="sourceLine" id="cb36-6" data-line-number="6"><span class="co">## </span></a>
<a class="sourceLine" id="cb36-7" data-line-number="7"><span class="co">##     intersect, setdiff, union</span></a>
<a class="sourceLine" id="cb36-8" data-line-number="8"><span class="co">## The following object is masked from &#39;package:GenomeInfoDb&#39;:</span></a>
<a class="sourceLine" id="cb36-9" data-line-number="9"><span class="co">## </span></a>
<a class="sourceLine" id="cb36-10" data-line-number="10"><span class="co">##     intersect</span></a>
<a class="sourceLine" id="cb36-11" data-line-number="11"><span class="co">## The following objects are masked from &#39;package:Biostrings&#39;:</span></a>
<a class="sourceLine" id="cb36-12" data-line-number="12"><span class="co">## </span></a>
<a class="sourceLine" id="cb36-13" data-line-number="13"><span class="co">##     collapse, intersect, setdiff, setequal, union</span></a>
<a class="sourceLine" id="cb36-14" data-line-number="14"><span class="co">## The following object is masked from &#39;package:XVector&#39;:</span></a>
<a class="sourceLine" id="cb36-15" data-line-number="15"><span class="co">## </span></a>
<a class="sourceLine" id="cb36-16" data-line-number="16"><span class="co">##     slice</span></a>
<a class="sourceLine" id="cb36-17" data-line-number="17"><span class="co">## The following objects are masked from &#39;package:IRanges&#39;:</span></a>
<a class="sourceLine" id="cb36-18" data-line-number="18"><span class="co">## </span></a>
<a class="sourceLine" id="cb36-19" data-line-number="19"><span class="co">##     collapse, desc, intersect, setdiff, slice, union</span></a>
<a class="sourceLine" id="cb36-20" data-line-number="20"><span class="co">## The following objects are masked from &#39;package:S4Vectors&#39;:</span></a>
<a class="sourceLine" id="cb36-21" data-line-number="21"><span class="co">## </span></a>
<a class="sourceLine" id="cb36-22" data-line-number="22"><span class="co">##     first, intersect, rename, setdiff, setequal, union</span></a>
<a class="sourceLine" id="cb36-23" data-line-number="23"><span class="co">## The following objects are masked from &#39;package:BiocGenerics&#39;:</span></a>
<a class="sourceLine" id="cb36-24" data-line-number="24"><span class="co">## </span></a>
<a class="sourceLine" id="cb36-25" data-line-number="25"><span class="co">##     combine, intersect, setdiff, union</span></a>
<a class="sourceLine" id="cb36-26" data-line-number="26"><span class="co">## The following objects are masked from &#39;package:stats&#39;:</span></a>
<a class="sourceLine" id="cb36-27" data-line-number="27"><span class="co">## </span></a>
<a class="sourceLine" id="cb36-28" data-line-number="28"><span class="co">##     filter, lag</span></a>
<a class="sourceLine" id="cb36-29" data-line-number="29"><span class="co">## The following objects are masked from &#39;package:base&#39;:</span></a>
<a class="sourceLine" id="cb36-30" data-line-number="30"><span class="co">## </span></a>
<a class="sourceLine" id="cb36-31" data-line-number="31"><span class="co">##     intersect, setdiff, setequal, union</span></a>
<a class="sourceLine" id="cb36-32" data-line-number="32"><span class="kw">as_tibble</span>(cds)</a>
<a class="sourceLine" id="cb36-33" data-line-number="33"><span class="co">## # A tibble: 13 x 19</span></a>
<a class="sourceLine" id="cb36-34" data-line-number="34"><span class="co">##    seqnames start   end width strand type  gene  locus_tag ribosomal_slipp…</span></a>
<a class="sourceLine" id="cb36-35" data-line-number="35"><span class="co">##    &lt;fct&gt;    &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;fct&gt;  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;     &lt;lgl&gt;           </span></a>
<a class="sourceLine" id="cb36-36" data-line-number="36"><span class="co">##  1 Severe …   266 13468 13203 +      CDS   ORF1… GU280_gp… TRUE            </span></a>
<a class="sourceLine" id="cb36-37" data-line-number="37"><span class="co">##  2 Severe … 13468 21555  8088 +      CDS   ORF1… GU280_gp… TRUE            </span></a>
<a class="sourceLine" id="cb36-38" data-line-number="38"><span class="co">##  3 Severe …   266 13483 13218 +      CDS   ORF1… GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb36-39" data-line-number="39"><span class="co">##  4 Severe … 21563 25384  3822 +      CDS   S     GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb36-40" data-line-number="40"><span class="co">##  5 Severe … 25393 26220   828 +      CDS   ORF3a GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb36-41" data-line-number="41"><span class="co">##  6 Severe … 26245 26472   228 +      CDS   E     GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb36-42" data-line-number="42"><span class="co">##  7 Severe … 26523 27191   669 +      CDS   M     GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb36-43" data-line-number="43"><span class="co">##  8 Severe … 27202 27387   186 +      CDS   ORF6  GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb36-44" data-line-number="44"><span class="co">##  9 Severe … 27394 27759   366 +      CDS   ORF7a GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb36-45" data-line-number="45"><span class="co">## 10 Severe … 27756 27887   132 +      CDS   ORF7b GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb36-46" data-line-number="46"><span class="co">## 11 Severe … 27894 28259   366 +      CDS   ORF8  GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb36-47" data-line-number="47"><span class="co">## 12 Severe … 28274 29533  1260 +      CDS   N     GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb36-48" data-line-number="48"><span class="co">## 13 Severe … 29558 29674   117 +      CDS   ORF10 GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb36-49" data-line-number="49"><span class="co">## # … with 10 more variables: note &lt;chr&gt;, codon_start &lt;dbl&gt;, product &lt;chr&gt;,</span></a>
<a class="sourceLine" id="cb36-50" data-line-number="50"><span class="co">## #   protein_id &lt;chr&gt;, db_xref &lt;I&lt;list&gt;&gt;, translation &lt;chr&gt;, loctype &lt;chr&gt;,</span></a>
<a class="sourceLine" id="cb36-51" data-line-number="51"><span class="co">## #   gene_synonym &lt;I&lt;list&gt;&gt;, gene_id &lt;chr&gt;, transcript_id &lt;chr&gt;</span></a></code></pre></div>
<p>For instance, we can find the gene associated with each coding sequence, using the following equivalent commands</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb37-1" data-line-number="1">cds<span class="op">$</span>gene</a>
<a class="sourceLine" id="cb37-2" data-line-number="2"><span class="co">##  [1] &quot;ORF1ab&quot; &quot;ORF1ab&quot; &quot;ORF1ab&quot; &quot;S&quot;      &quot;ORF3a&quot;  &quot;E&quot;      &quot;M&quot;      &quot;ORF6&quot;  </span></a>
<a class="sourceLine" id="cb37-3" data-line-number="3"><span class="co">##  [9] &quot;ORF7a&quot;  &quot;ORF7b&quot;  &quot;ORF8&quot;   &quot;N&quot;      &quot;ORF10&quot;</span></a>
<a class="sourceLine" id="cb37-4" data-line-number="4"></a>
<a class="sourceLine" id="cb37-5" data-line-number="5"><span class="kw">subset</span>(cds, , gene)</a>
<a class="sourceLine" id="cb37-6" data-line-number="6"><span class="co">## GRanges object with 13 ranges and 1 metadata column:</span></a>
<a class="sourceLine" id="cb37-7" data-line-number="7"><span class="co">##                                              seqnames      ranges strand |</span></a>
<a class="sourceLine" id="cb37-8" data-line-number="8"><span class="co">##                                                 &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; |</span></a>
<a class="sourceLine" id="cb37-9" data-line-number="9"><span class="co">##    [1] Severe acute respiratory syndrome coronavirus2   266-13468      + |</span></a>
<a class="sourceLine" id="cb37-10" data-line-number="10"><span class="co">##    [2] Severe acute respiratory syndrome coronavirus2 13468-21555      + |</span></a>
<a class="sourceLine" id="cb37-11" data-line-number="11"><span class="co">##    [3] Severe acute respiratory syndrome coronavirus2   266-13483      + |</span></a>
<a class="sourceLine" id="cb37-12" data-line-number="12"><span class="co">##    [4] Severe acute respiratory syndrome coronavirus2 21563-25384      + |</span></a>
<a class="sourceLine" id="cb37-13" data-line-number="13"><span class="co">##    [5] Severe acute respiratory syndrome coronavirus2 25393-26220      + |</span></a>
<a class="sourceLine" id="cb37-14" data-line-number="14"><span class="co">##    ...                                            ...         ...    ... .</span></a>
<a class="sourceLine" id="cb37-15" data-line-number="15"><span class="co">##    [9] Severe acute respiratory syndrome coronavirus2 27394-27759      + |</span></a>
<a class="sourceLine" id="cb37-16" data-line-number="16"><span class="co">##   [10] Severe acute respiratory syndrome coronavirus2 27756-27887      + |</span></a>
<a class="sourceLine" id="cb37-17" data-line-number="17"><span class="co">##   [11] Severe acute respiratory syndrome coronavirus2 27894-28259      + |</span></a>
<a class="sourceLine" id="cb37-18" data-line-number="18"><span class="co">##   [12] Severe acute respiratory syndrome coronavirus2 28274-29533      + |</span></a>
<a class="sourceLine" id="cb37-19" data-line-number="19"><span class="co">##   [13] Severe acute respiratory syndrome coronavirus2 29558-29674      + |</span></a>
<a class="sourceLine" id="cb37-20" data-line-number="20"><span class="co">##               gene</span></a>
<a class="sourceLine" id="cb37-21" data-line-number="21"><span class="co">##        &lt;character&gt;</span></a>
<a class="sourceLine" id="cb37-22" data-line-number="22"><span class="co">##    [1]      ORF1ab</span></a>
<a class="sourceLine" id="cb37-23" data-line-number="23"><span class="co">##    [2]      ORF1ab</span></a>
<a class="sourceLine" id="cb37-24" data-line-number="24"><span class="co">##    [3]      ORF1ab</span></a>
<a class="sourceLine" id="cb37-25" data-line-number="25"><span class="co">##    [4]           S</span></a>
<a class="sourceLine" id="cb37-26" data-line-number="26"><span class="co">##    [5]       ORF3a</span></a>
<a class="sourceLine" id="cb37-27" data-line-number="27"><span class="co">##    ...         ...</span></a>
<a class="sourceLine" id="cb37-28" data-line-number="28"><span class="co">##    [9]       ORF7a</span></a>
<a class="sourceLine" id="cb37-29" data-line-number="29"><span class="co">##   [10]       ORF7b</span></a>
<a class="sourceLine" id="cb37-30" data-line-number="30"><span class="co">##   [11]        ORF8</span></a>
<a class="sourceLine" id="cb37-31" data-line-number="31"><span class="co">##   [12]           N</span></a>
<a class="sourceLine" id="cb37-32" data-line-number="32"><span class="co">##   [13]       ORF10</span></a>
<a class="sourceLine" id="cb37-33" data-line-number="33"><span class="co">##   -------</span></a>
<a class="sourceLine" id="cb37-34" data-line-number="34"><span class="co">##   seqinfo: 1 sequence from NC_045512.2 genome</span></a>
<a class="sourceLine" id="cb37-35" data-line-number="35"></a>
<a class="sourceLine" id="cb37-36" data-line-number="36"><span class="kw">as_tibble</span>(cds) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb37-37" data-line-number="37"><span class="st">    </span><span class="co">## info on gene and protein product</span></a>
<a class="sourceLine" id="cb37-38" data-line-number="38"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw">select</span>(gene, product)</a>
<a class="sourceLine" id="cb37-39" data-line-number="39"><span class="co">## # A tibble: 13 x 2</span></a>
<a class="sourceLine" id="cb37-40" data-line-number="40"><span class="co">##    gene   product                    </span></a>
<a class="sourceLine" id="cb37-41" data-line-number="41"><span class="co">##    &lt;chr&gt;  &lt;chr&gt;                      </span></a>
<a class="sourceLine" id="cb37-42" data-line-number="42"><span class="co">##  1 ORF1ab ORF1ab polyprotein         </span></a>
<a class="sourceLine" id="cb37-43" data-line-number="43"><span class="co">##  2 ORF1ab ORF1ab polyprotein         </span></a>
<a class="sourceLine" id="cb37-44" data-line-number="44"><span class="co">##  3 ORF1ab ORF1a polyprotein          </span></a>
<a class="sourceLine" id="cb37-45" data-line-number="45"><span class="co">##  4 S      surface glycoprotein       </span></a>
<a class="sourceLine" id="cb37-46" data-line-number="46"><span class="co">##  5 ORF3a  ORF3a protein              </span></a>
<a class="sourceLine" id="cb37-47" data-line-number="47"><span class="co">##  6 E      envelope protein           </span></a>
<a class="sourceLine" id="cb37-48" data-line-number="48"><span class="co">##  7 M      membrane glycoprotein      </span></a>
<a class="sourceLine" id="cb37-49" data-line-number="49"><span class="co">##  8 ORF6   ORF6 protein               </span></a>
<a class="sourceLine" id="cb37-50" data-line-number="50"><span class="co">##  9 ORF7a  ORF7a protein              </span></a>
<a class="sourceLine" id="cb37-51" data-line-number="51"><span class="co">## 10 ORF7b  ORF7b                      </span></a>
<a class="sourceLine" id="cb37-52" data-line-number="52"><span class="co">## 11 ORF8   ORF8 protein               </span></a>
<a class="sourceLine" id="cb37-53" data-line-number="53"><span class="co">## 12 N      nucleocapsid phosphoprotein</span></a>
<a class="sourceLine" id="cb37-54" data-line-number="54"><span class="co">## 13 ORF10  ORF10 protein</span></a></code></pre></div></li>
</ul>
<p>Using genomic ranges to extract DNA sequences</p>
<ul>
<li><p>Recall the sequence of the entire genome</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1">ref &lt;-<span class="st"> </span><span class="kw">getSeq</span>(gb)</a>
<a class="sourceLine" id="cb38-2" data-line-number="2">ref</a>
<a class="sourceLine" id="cb38-3" data-line-number="3"><span class="co">##   A DNAStringSet instance of length 1</span></a>
<a class="sourceLine" id="cb38-4" data-line-number="4"><span class="co">##     width seq                                               names               </span></a>
<a class="sourceLine" id="cb38-5" data-line-number="5"><span class="co">## [1] 29903 ATTAAAGGTTTATACCTTCCCAG...AAAAAAAAAAAAAAAAAAAAAAA Severe acute resp...</span></a></code></pre></div></li>
<li><p>Use <code>getSeq()</code> to extract the <code>DNAStringSet()</code> corresponding to the genomic ranges specified by <code>cds</code></p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" data-line-number="1">dna &lt;-<span class="st"> </span><span class="kw">getSeq</span>(ref, cds)</a>
<a class="sourceLine" id="cb39-2" data-line-number="2">dna</a>
<a class="sourceLine" id="cb39-3" data-line-number="3"><span class="co">##   A DNAStringSet instance of length 13</span></a>
<a class="sourceLine" id="cb39-4" data-line-number="4"><span class="co">##      width seq</span></a>
<a class="sourceLine" id="cb39-5" data-line-number="5"><span class="co">##  [1] 13203 ATGGAGAGCCTTGTCCCTGGTTTCAACGAGAAA...CTTCAGTCAGCTGATGCACAATCGTTTTTAAAC</span></a>
<a class="sourceLine" id="cb39-6" data-line-number="6"><span class="co">##  [2]  8088 CGGGTTTGCGGTGTAAGTGCAGCCCGTCTTACA...GTTATTTCTAGTGATGTTCTTGTTAACAACTAA</span></a>
<a class="sourceLine" id="cb39-7" data-line-number="7"><span class="co">##  [3] 13218 ATGGAGAGCCTTGTCCCTGGTTTCAACGAGAAA...GCACAATCGTTTTTAAACGGGTTTGCGGTGTAA</span></a>
<a class="sourceLine" id="cb39-8" data-line-number="8"><span class="co">##  [4]  3822 ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTC...GTGCTCAAAGGAGTCAAATTACATTACACATAA</span></a>
<a class="sourceLine" id="cb39-9" data-line-number="9"><span class="co">##  [5]   828 ATGGATTTGTTTATGAGAATCTTCACAATTGGA...GAACCGACGACGACTACTAGCGTGCCTTTGTAA</span></a>
<a class="sourceLine" id="cb39-10" data-line-number="10"><span class="co">##  ...   ... ...</span></a>
<a class="sourceLine" id="cb39-11" data-line-number="11"><span class="co">##  [9]   366 ATGAAAATTATTCTTTTCTTGGCACTGATAACA...CTTTGCTTCACACTCAAAAGAAAGACAGAATGA</span></a>
<a class="sourceLine" id="cb39-12" data-line-number="12"><span class="co">## [10]   132 ATGATTGAACTTTCATTAATTGACTTCTATTTG...CTGCAAGATCATAATGAAACTTGTCACGCCTAA</span></a>
<a class="sourceLine" id="cb39-13" data-line-number="13"><span class="co">## [11]   366 ATGAAATTTCTTGTTTTCTTAGGAATCATCACA...CATGACGTTCGTGTTGTTTTAGATTTCATCTAA</span></a>
<a class="sourceLine" id="cb39-14" data-line-number="14"><span class="co">## [12]  1260 ATGTCTGATAATGGACCCCAAAATCAGCGAAAT...TCCATGAGCAGTGCTGACTCAACTCAGGCCTAA</span></a>
<a class="sourceLine" id="cb39-15" data-line-number="15"><span class="co">## [13]   117 ATGGGCTATATAAACGTTTTCGCTTTTCCGTTT...CAAGTAGATGTAGTTAACTTTAATCTCACATAG</span></a></code></pre></div></li>
<li><p>It might be useful to coordinate information about each coding sequence with the sequence itself</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1">cds<span class="op">$</span>DNA &lt;-<span class="st"> </span>dna</a>
<a class="sourceLine" id="cb40-2" data-line-number="2">cds</a>
<a class="sourceLine" id="cb40-3" data-line-number="3"><span class="co">## GRanges object with 13 ranges and 15 metadata columns:</span></a>
<a class="sourceLine" id="cb40-4" data-line-number="4"><span class="co">##                                              seqnames      ranges strand |</span></a>
<a class="sourceLine" id="cb40-5" data-line-number="5"><span class="co">##                                                 &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; |</span></a>
<a class="sourceLine" id="cb40-6" data-line-number="6"><span class="co">##    [1] Severe acute respiratory syndrome coronavirus2   266-13468      + |</span></a>
<a class="sourceLine" id="cb40-7" data-line-number="7"><span class="co">##    [2] Severe acute respiratory syndrome coronavirus2 13468-21555      + |</span></a>
<a class="sourceLine" id="cb40-8" data-line-number="8"><span class="co">##    [3] Severe acute respiratory syndrome coronavirus2   266-13483      + |</span></a>
<a class="sourceLine" id="cb40-9" data-line-number="9"><span class="co">##    [4] Severe acute respiratory syndrome coronavirus2 21563-25384      + |</span></a>
<a class="sourceLine" id="cb40-10" data-line-number="10"><span class="co">##    [5] Severe acute respiratory syndrome coronavirus2 25393-26220      + |</span></a>
<a class="sourceLine" id="cb40-11" data-line-number="11"><span class="co">##    ...                                            ...         ...    ... .</span></a>
<a class="sourceLine" id="cb40-12" data-line-number="12"><span class="co">##    [9] Severe acute respiratory syndrome coronavirus2 27394-27759      + |</span></a>
<a class="sourceLine" id="cb40-13" data-line-number="13"><span class="co">##   [10] Severe acute respiratory syndrome coronavirus2 27756-27887      + |</span></a>
<a class="sourceLine" id="cb40-14" data-line-number="14"><span class="co">##   [11] Severe acute respiratory syndrome coronavirus2 27894-28259      + |</span></a>
<a class="sourceLine" id="cb40-15" data-line-number="15"><span class="co">##   [12] Severe acute respiratory syndrome coronavirus2 28274-29533      + |</span></a>
<a class="sourceLine" id="cb40-16" data-line-number="16"><span class="co">##   [13] Severe acute respiratory syndrome coronavirus2 29558-29674      + |</span></a>
<a class="sourceLine" id="cb40-17" data-line-number="17"><span class="co">##               type        gene   locus_tag ribosomal_slippage</span></a>
<a class="sourceLine" id="cb40-18" data-line-number="18"><span class="co">##        &lt;character&gt; &lt;character&gt; &lt;character&gt;          &lt;logical&gt;</span></a>
<a class="sourceLine" id="cb40-19" data-line-number="19"><span class="co">##    [1]         CDS      ORF1ab  GU280_gp01               TRUE</span></a>
<a class="sourceLine" id="cb40-20" data-line-number="20"><span class="co">##    [2]         CDS      ORF1ab  GU280_gp01               TRUE</span></a>
<a class="sourceLine" id="cb40-21" data-line-number="21"><span class="co">##    [3]         CDS      ORF1ab  GU280_gp01              FALSE</span></a>
<a class="sourceLine" id="cb40-22" data-line-number="22"><span class="co">##    [4]         CDS           S  GU280_gp02              FALSE</span></a>
<a class="sourceLine" id="cb40-23" data-line-number="23"><span class="co">##    [5]         CDS       ORF3a  GU280_gp03              FALSE</span></a>
<a class="sourceLine" id="cb40-24" data-line-number="24"><span class="co">##    ...         ...         ...         ...                ...</span></a>
<a class="sourceLine" id="cb40-25" data-line-number="25"><span class="co">##    [9]         CDS       ORF7a  GU280_gp07              FALSE</span></a>
<a class="sourceLine" id="cb40-26" data-line-number="26"><span class="co">##   [10]         CDS       ORF7b  GU280_gp08              FALSE</span></a>
<a class="sourceLine" id="cb40-27" data-line-number="27"><span class="co">##   [11]         CDS        ORF8  GU280_gp09              FALSE</span></a>
<a class="sourceLine" id="cb40-28" data-line-number="28"><span class="co">##   [12]         CDS           N  GU280_gp10              FALSE</span></a>
<a class="sourceLine" id="cb40-29" data-line-number="29"><span class="co">##   [13]         CDS       ORF10  GU280_gp11              FALSE</span></a>
<a class="sourceLine" id="cb40-30" data-line-number="30"><span class="co">##                                                note codon_start</span></a>
<a class="sourceLine" id="cb40-31" data-line-number="31"><span class="co">##                                         &lt;character&gt;   &lt;numeric&gt;</span></a>
<a class="sourceLine" id="cb40-32" data-line-number="32"><span class="co">##    [1] pp1ab; translated by -1 ribosomal frameshift           1</span></a>
<a class="sourceLine" id="cb40-33" data-line-number="33"><span class="co">##    [2] pp1ab; translated by -1 ribosomal frameshift           1</span></a>
<a class="sourceLine" id="cb40-34" data-line-number="34"><span class="co">##    [3]                                         pp1a           1</span></a>
<a class="sourceLine" id="cb40-35" data-line-number="35"><span class="co">##    [4]            structural protein; spike protein           1</span></a>
<a class="sourceLine" id="cb40-36" data-line-number="36"><span class="co">##    [5]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb40-37" data-line-number="37"><span class="co">##    ...                                          ...         ...</span></a>
<a class="sourceLine" id="cb40-38" data-line-number="38"><span class="co">##    [9]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb40-39" data-line-number="39"><span class="co">##   [10]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb40-40" data-line-number="40"><span class="co">##   [11]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb40-41" data-line-number="41"><span class="co">##   [12]                     ORF9; structural protein           1</span></a>
<a class="sourceLine" id="cb40-42" data-line-number="42"><span class="co">##   [13]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb40-43" data-line-number="43"><span class="co">##                            product     protein_id         db_xref   translation</span></a>
<a class="sourceLine" id="cb40-44" data-line-number="44"><span class="co">##                        &lt;character&gt;    &lt;character&gt; &lt;CharacterList&gt; &lt;AAStringSet&gt;</span></a>
<a class="sourceLine" id="cb40-45" data-line-number="45"><span class="co">##    [1]          ORF1ab polyprotein YP_009724389.1 GeneID:43740578     MES...VNN</span></a>
<a class="sourceLine" id="cb40-46" data-line-number="46"><span class="co">##    [2]          ORF1ab polyprotein YP_009724389.1 GeneID:43740578     MES...VNN</span></a>
<a class="sourceLine" id="cb40-47" data-line-number="47"><span class="co">##    [3]           ORF1a polyprotein YP_009725295.1 GeneID:43740578     MES...FAV</span></a>
<a class="sourceLine" id="cb40-48" data-line-number="48"><span class="co">##    [4]        surface glycoprotein YP_009724390.1 GeneID:43740568     MFV...HYT</span></a>
<a class="sourceLine" id="cb40-49" data-line-number="49"><span class="co">##    [5]               ORF3a protein YP_009724391.1 GeneID:43740569     MDL...VPL</span></a>
<a class="sourceLine" id="cb40-50" data-line-number="50"><span class="co">##    ...                         ...            ...             ...           ...</span></a>
<a class="sourceLine" id="cb40-51" data-line-number="51"><span class="co">##    [9]               ORF7a protein YP_009724395.1 GeneID:43740573     MKI...KTE</span></a>
<a class="sourceLine" id="cb40-52" data-line-number="52"><span class="co">##   [10]                       ORF7b YP_009725318.1 GeneID:43740574     MIE...CHA</span></a>
<a class="sourceLine" id="cb40-53" data-line-number="53"><span class="co">##   [11]                ORF8 protein YP_009724396.1 GeneID:43740577     MKF...DFI</span></a>
<a class="sourceLine" id="cb40-54" data-line-number="54"><span class="co">##   [12] nucleocapsid phosphoprotein YP_009724397.2 GeneID:43740575     MSD...TQA</span></a>
<a class="sourceLine" id="cb40-55" data-line-number="55"><span class="co">##   [13]               ORF10 protein YP_009725255.1 GeneID:43740576     MGY...NLT</span></a>
<a class="sourceLine" id="cb40-56" data-line-number="56"><span class="co">##            loctype       gene_synonym     gene_id transcript_id            DNA</span></a>
<a class="sourceLine" id="cb40-57" data-line-number="57"><span class="co">##        &lt;character&gt;    &lt;CharacterList&gt; &lt;character&gt;   &lt;character&gt; &lt;DNAStringSet&gt;</span></a>
<a class="sourceLine" id="cb40-58" data-line-number="58"><span class="co">##    [1]      normal                 NA  GU280_gp01  GU280_gp01.1      ATG...AAC</span></a>
<a class="sourceLine" id="cb40-59" data-line-number="59"><span class="co">##    [2]      normal                 NA  GU280_gp01  GU280_gp01.1      CGG...TAA</span></a>
<a class="sourceLine" id="cb40-60" data-line-number="60"><span class="co">##    [3]      normal                 NA  GU280_gp01  GU280_gp01.2      ATG...TAA</span></a>
<a class="sourceLine" id="cb40-61" data-line-number="61"><span class="co">##    [4]      normal spike glycoprotein  GU280_gp02  GU280_gp02.1      ATG...TAA</span></a>
<a class="sourceLine" id="cb40-62" data-line-number="62"><span class="co">##    [5]      normal                 NA  GU280_gp03  GU280_gp03.1      ATG...TAA</span></a>
<a class="sourceLine" id="cb40-63" data-line-number="63"><span class="co">##    ...         ...                ...         ...           ...            ...</span></a>
<a class="sourceLine" id="cb40-64" data-line-number="64"><span class="co">##    [9]      normal                 NA  GU280_gp07  GU280_gp07.1      ATG...TGA</span></a>
<a class="sourceLine" id="cb40-65" data-line-number="65"><span class="co">##   [10]      normal                 NA  GU280_gp08  GU280_gp08.1      ATG...TAA</span></a>
<a class="sourceLine" id="cb40-66" data-line-number="66"><span class="co">##   [11]      normal                 NA  GU280_gp09  GU280_gp09.1      ATG...TAA</span></a>
<a class="sourceLine" id="cb40-67" data-line-number="67"><span class="co">##   [12]      normal                 NA  GU280_gp10  GU280_gp10.1      ATG...TAA</span></a>
<a class="sourceLine" id="cb40-68" data-line-number="68"><span class="co">##   [13]      normal                 NA  GU280_gp11  GU280_gp11.1      ATG...TAG</span></a>
<a class="sourceLine" id="cb40-69" data-line-number="69"><span class="co">##   -------</span></a>
<a class="sourceLine" id="cb40-70" data-line-number="70"><span class="co">##   seqinfo: 1 sequence from NC_045512.2 genome</span></a></code></pre></div></li>
</ul>
</div>
</div>
<div id="day-31-sequence-alignment-and-visualization" class="section level2">
<h2><span class="header-section-number">5.3</span> Day 31 Sequence alignment and visualization</h2>
<div id="sequence-alignment" class="section level3">
<h3><span class="header-section-number">5.3.1</span> Sequence alignment</h3>
<p>Setup</p>
<ul>
<li><p>Create a working directory, if necessary. E.g.,</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" data-line-number="1">workdir &lt;-<span class="st"> &quot;workdir&quot;</span></a>
<a class="sourceLine" id="cb41-2" data-line-number="2"><span class="cf">if</span> (<span class="op">!</span><span class="kw">dir.exists</span>(workdir))</a>
<a class="sourceLine" id="cb41-3" data-line-number="3">    <span class="kw">dir.create</span>(workdir)</a></code></pre></div></li>
<li><p>Install (if necessary, using <code>BiocManager::install(&quot;DECIPHER&quot;)</code>) and attach the <a href="https://bioconductor.org/packages/DECIPHER">DECIPHER</a> package</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" data-line-number="1"><span class="kw">library</span>(DECIPHER)</a>
<a class="sourceLine" id="cb42-2" data-line-number="2"></a>
<a class="sourceLine" id="cb42-3" data-line-number="3"><span class="kw">maintainer</span>(<span class="st">&quot;DECIPHER&quot;</span>)</a>
<a class="sourceLine" id="cb42-4" data-line-number="4"><span class="co">## [1] &quot;Erik Wright &lt;eswright@pitt.edu&gt;&quot;</span></a>
<a class="sourceLine" id="cb42-5" data-line-number="5"></a>
<a class="sourceLine" id="cb42-6" data-line-number="6"><span class="kw">citation</span>(<span class="st">&quot;DECIPHER&quot;</span>)</a>
<a class="sourceLine" id="cb42-7" data-line-number="7"><span class="co">## </span></a>
<a class="sourceLine" id="cb42-8" data-line-number="8"><span class="co">## Wright ES (2016). &quot;Using DECIPHER v2.0 to Analyze Big Biological</span></a>
<a class="sourceLine" id="cb42-9" data-line-number="9"><span class="co">## Sequence Data in R.&quot; _The R Journal_, *8*(1), 352-359.</span></a>
<a class="sourceLine" id="cb42-10" data-line-number="10"><span class="co">## </span></a>
<a class="sourceLine" id="cb42-11" data-line-number="11"><span class="co">## A BibTeX entry for LaTeX users is</span></a>
<a class="sourceLine" id="cb42-12" data-line-number="12"><span class="co">## </span></a>
<a class="sourceLine" id="cb42-13" data-line-number="13"><span class="co">##   @Article{,</span></a>
<a class="sourceLine" id="cb42-14" data-line-number="14"><span class="co">##     title = {Using DECIPHER v2.0 to Analyze Big Biological Sequence Data in R},</span></a>
<a class="sourceLine" id="cb42-15" data-line-number="15"><span class="co">##     author = {Erik S. Wright},</span></a>
<a class="sourceLine" id="cb42-16" data-line-number="16"><span class="co">##     journal = {The R Journal},</span></a>
<a class="sourceLine" id="cb42-17" data-line-number="17"><span class="co">##     year = {2016},</span></a>
<a class="sourceLine" id="cb42-18" data-line-number="18"><span class="co">##     volume = {8},</span></a>
<a class="sourceLine" id="cb42-19" data-line-number="19"><span class="co">##     number = {1},</span></a>
<a class="sourceLine" id="cb42-20" data-line-number="20"><span class="co">##     pages = {352-359},</span></a>
<a class="sourceLine" id="cb42-21" data-line-number="21"><span class="co">##   }</span></a></code></pre></div></li>
<li><p>This package contains tools for multiple sequence alignment. Explore the introductory and other vignettes in this package.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" data-line-number="1"><span class="kw">browseVignettes</span>(<span class="dt">package =</span> <span class="st">&quot;DECIPHER&quot;</span>)</a></code></pre></div></li>
<li><p>Check out one of the help pages, e.g.,</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb44-1" data-line-number="1">?AlignSeqs</a></code></pre></div>
<p>Use <code>example(AlignSeqs)</code> to run the examples on the help page.</p></li>
</ul>
<p>Sequence download, input, and alignment</p>
<ul>
<li><p>Download, if necessary</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" data-line-number="1">url &lt;-<span class="st"> &quot;https://raw.githubusercontent.com/mtmorgan/QuaRantine/master/assets/05-SARS-CoV2-NY.fasta&quot;</span></a>
<a class="sourceLine" id="cb45-2" data-line-number="2">fasta_file &lt;-<span class="st"> </span><span class="kw">file.path</span>(workdir, <span class="kw">basename</span>(url))</a>
<a class="sourceLine" id="cb45-3" data-line-number="3"><span class="cf">if</span> ( <span class="op">!</span><span class="kw">file.exists</span>(fasta_file) )</a>
<a class="sourceLine" id="cb45-4" data-line-number="4">    <span class="kw">download.file</span>(url, fasta_file)</a></code></pre></div></li>
<li><p>Load as <code>DNAStringSet</code></p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb46-1" data-line-number="1">dna &lt;-<span class="st"> </span><span class="kw">readDNAStringSet</span>(fasta_file)</a>
<a class="sourceLine" id="cb46-2" data-line-number="2">dna</a>
<a class="sourceLine" id="cb46-3" data-line-number="3"><span class="co">##   A DNAStringSet instance of length 245</span></a>
<a class="sourceLine" id="cb46-4" data-line-number="4"><span class="co">##       width seq                                             names               </span></a>
<a class="sourceLine" id="cb46-5" data-line-number="5"><span class="co">##   [1] 29873 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434817</span></a>
<a class="sourceLine" id="cb46-6" data-line-number="6"><span class="co">##   [2] 29879 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434816</span></a>
<a class="sourceLine" id="cb46-7" data-line-number="7"><span class="co">##   [3] 29873 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434815</span></a>
<a class="sourceLine" id="cb46-8" data-line-number="8"><span class="co">##   [4] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434814</span></a>
<a class="sourceLine" id="cb46-9" data-line-number="9"><span class="co">##   [5] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434813</span></a>
<a class="sourceLine" id="cb46-10" data-line-number="10"><span class="co">##   ...   ... ...</span></a>
<a class="sourceLine" id="cb46-11" data-line-number="11"><span class="co">## [241] 29724 GATCTGTTCTCTAAACGAACTT...TGTACAGTGAACAATGCTAGGG MT370833</span></a>
<a class="sourceLine" id="cb46-12" data-line-number="12"><span class="co">## [242] 29717 TCTAAACGAACTTTAAAATCTG...TACAGTGAACAATGCTAGGGAG MT370832</span></a>
<a class="sourceLine" id="cb46-13" data-line-number="13"><span class="co">## [243] 29716 TCTAAACGAACTTTAAAATCTG...GTACAGTGAACAATGCTAGGGA MT370831</span></a>
<a class="sourceLine" id="cb46-14" data-line-number="14"><span class="co">## [244] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT325627</span></a>
<a class="sourceLine" id="cb46-15" data-line-number="15"><span class="co">## [245] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT304486</span></a></code></pre></div></li>
<li><p>Perform multiple alignment using DECIPHER</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb47-1" data-line-number="1">aln &lt;-<span class="st"> </span><span class="kw">AlignSeqs</span>(dna, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb47-2" data-line-number="2">aln</a>
<a class="sourceLine" id="cb47-3" data-line-number="3"><span class="co">##   A DNAStringSet instance of length 245</span></a>
<a class="sourceLine" id="cb47-4" data-line-number="4"><span class="co">##       width seq                                             names               </span></a>
<a class="sourceLine" id="cb47-5" data-line-number="5"><span class="co">##   [1] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434817</span></a>
<a class="sourceLine" id="cb47-6" data-line-number="6"><span class="co">##   [2] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434816</span></a>
<a class="sourceLine" id="cb47-7" data-line-number="7"><span class="co">##   [3] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434815</span></a>
<a class="sourceLine" id="cb47-8" data-line-number="8"><span class="co">##   [4] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434814</span></a>
<a class="sourceLine" id="cb47-9" data-line-number="9"><span class="co">##   [5] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434813</span></a>
<a class="sourceLine" id="cb47-10" data-line-number="10"><span class="co">##   ...   ... ...</span></a>
<a class="sourceLine" id="cb47-11" data-line-number="11"><span class="co">## [241] 29882 ----------------------...---------------------- MT370833</span></a>
<a class="sourceLine" id="cb47-12" data-line-number="12"><span class="co">## [242] 29882 ----------------------...---------------------- MT370832</span></a>
<a class="sourceLine" id="cb47-13" data-line-number="13"><span class="co">## [243] 29882 ----------------------...---------------------- MT370831</span></a>
<a class="sourceLine" id="cb47-14" data-line-number="14"><span class="co">## [244] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT325627</span></a>
<a class="sourceLine" id="cb47-15" data-line-number="15"><span class="co">## [245] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT304486</span></a></code></pre></div></li>
<li><p>The ‘distance matrix’ measures the proximity (Hamming distance) of each sequence to other sequences in sequence space. The result is a large matrix of pairwise distances</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb48-1" data-line-number="1">dist &lt;-<span class="st"> </span><span class="kw">DistanceMatrix</span>(aln)</a>
<a class="sourceLine" id="cb48-2" data-line-number="2"><span class="co">## ================================================================================</span></a>
<a class="sourceLine" id="cb48-3" data-line-number="3"><span class="co">## </span></a>
<a class="sourceLine" id="cb48-4" data-line-number="4"><span class="co">## Time difference of 3.28 secs</span></a>
<a class="sourceLine" id="cb48-5" data-line-number="5"><span class="kw">class</span>(dist)</a>
<a class="sourceLine" id="cb48-6" data-line-number="6"><span class="co">## [1] &quot;matrix&quot;</span></a>
<a class="sourceLine" id="cb48-7" data-line-number="7"><span class="kw">dim</span>(dist)</a>
<a class="sourceLine" id="cb48-8" data-line-number="8"><span class="co">## [1] 245 245</span></a>
<a class="sourceLine" id="cb48-9" data-line-number="9">dist[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb48-10" data-line-number="10"><span class="co">##              MT434817     MT434816     MT434815     MT434814     MT434813</span></a>
<a class="sourceLine" id="cb48-11" data-line-number="11"><span class="co">## MT434817 0.0000000000 0.0005019744 0.0007027642 0.0007362292 0.0006358343</span></a>
<a class="sourceLine" id="cb48-12" data-line-number="12"><span class="co">## MT434816 0.0005019744 0.0000000000 0.0004685095 0.0005019744 0.0004015795</span></a>
<a class="sourceLine" id="cb48-13" data-line-number="13"><span class="co">## MT434815 0.0007027642 0.0004685095 0.0000000000 0.0007027642 0.0006023693</span></a>
<a class="sourceLine" id="cb48-14" data-line-number="14"><span class="co">## MT434814 0.0007362292 0.0005019744 0.0007027642 0.0000000000 0.0003011847</span></a>
<a class="sourceLine" id="cb48-15" data-line-number="15"><span class="co">## MT434813 0.0006358343 0.0004015795 0.0006023693 0.0003011847 0.0000000000</span></a></code></pre></div></li>
<li><p>Use the distance matrix to cluster sequences using the ‘Neighbor Joining’ algorithm; visualize the result</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb49-1" data-line-number="1">dendrogram &lt;-<span class="st"> </span><span class="kw">IdClusters</span>(dist, <span class="dt">method =</span> <span class="st">&quot;NJ&quot;</span>, <span class="dt">type =</span> <span class="st">&quot;dendrogram&quot;</span>)</a>
<a class="sourceLine" id="cb49-2" data-line-number="2"><span class="co">## ================================================================================</span></a>
<a class="sourceLine" id="cb49-3" data-line-number="3"><span class="co">## </span></a>
<a class="sourceLine" id="cb49-4" data-line-number="4"><span class="co">## Time difference of 0.41 secs</span></a>
<a class="sourceLine" id="cb49-5" data-line-number="5"><span class="kw">plot</span>(dendrogram)</a></code></pre></div>
<p><img src="05-Bioinformatics_files/figure-html/unnamed-chunk-51-1.png" width="672" /></p></li>
<li><p>Save the tree to the working directory using the ‘Newick’ format</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb50-1" data-line-number="1">dendrogram_file &lt;-<span class="st"> </span><span class="kw">file.path</span>(workdir, <span class="st">&quot;SARS-CoV2-NY.newick&quot;</span>)</a>
<a class="sourceLine" id="cb50-2" data-line-number="2"><span class="kw">WriteDendrogram</span>(dendrogram, dendrogram_file, <span class="dt">quoteLabels =</span> <span class="ot">FALSE</span>)</a></code></pre></div></li>
</ul>
<p><strong>Advanced activity</strong></p>
<ul>
<li><p>It seems likely that a more biologically motivated analysis would focus on codiong sequence, in particular the S, E, M and / or N loci. And the analysis would use all sequences, not just those from a particular geo-political region.</p></li>
<li><p>The information can be extracted from a genbank record, e.g., from the <code>gb_record</code> retrieved above</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb51-1" data-line-number="1">gb &lt;-<span class="st"> </span><span class="kw">readGenBank</span>(<span class="dt">text =</span> gb_record)</a>
<a class="sourceLine" id="cb51-2" data-line-number="2"><span class="co">## No exons read from genbank file. Assuming sections of CDS are full exons</span></a>
<a class="sourceLine" id="cb51-3" data-line-number="3"><span class="co">## No transcript features (mRNA) found, using spans of CDSs</span></a>
<a class="sourceLine" id="cb51-4" data-line-number="4">semn &lt;-</a>
<a class="sourceLine" id="cb51-5" data-line-number="5"><span class="st">    </span><span class="kw">cds</span>(gb) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">subset</span>(gene <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;S&quot;</span>, <span class="st">&quot;E&quot;</span>, <span class="st">&quot;M&quot;</span>, <span class="st">&quot;N&quot;</span>))</a>
<a class="sourceLine" id="cb51-6" data-line-number="6">ref &lt;-<span class="st"> </span><span class="kw">getSeq</span>(gb)</a>
<a class="sourceLine" id="cb51-7" data-line-number="7"><span class="kw">getSeq</span>(ref, semn)</a>
<a class="sourceLine" id="cb51-8" data-line-number="8"><span class="co">##   A DNAStringSet instance of length 4</span></a>
<a class="sourceLine" id="cb51-9" data-line-number="9"><span class="co">##     width seq</span></a>
<a class="sourceLine" id="cb51-10" data-line-number="10"><span class="co">## [1]  3822 ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCT...GTGCTCAAAGGAGTCAAATTACATTACACATAA</span></a>
<a class="sourceLine" id="cb51-11" data-line-number="11"><span class="co">## [2]   228 ATGTACTCATTCGTTTCGGAAGAGACAGGTACGT...AATTCTTCTAGAGTTCCTGATCTTCTGGTCTAA</span></a>
<a class="sourceLine" id="cb51-12" data-line-number="12"><span class="co">## [3]   669 ATGGCAGATTCCAACGGTACTATTACCGTTGAAG...AGCAGTGACAATATTGCTTTGCTTGTACAGTAA</span></a>
<a class="sourceLine" id="cb51-13" data-line-number="13"><span class="co">## [4]  1260 ATGTCTGATAATGGACCCCAAAATCAGCGAAATG...TCCATGAGCAGTGCTGACTCAACTCAGGCCTAA</span></a></code></pre></div></li>
<li><p>The script used to retrieve New York State sequences is [available][get-SARS-CoV2-NY-fasta]. The crucial modification is in the <code>get_accession_sequence()</code> function. Modify this function to retrieve, e.g., the S (surface glycoprotein) coding sequence, along the lines of</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb52-1" data-line-number="1">...</a>
<a class="sourceLine" id="cb52-2" data-line-number="2">result &lt;-<span class="st"> </span><span class="kw">readGenBank</span>(<span class="dt">text =</span> text)</a>
<a class="sourceLine" id="cb52-3" data-line-number="3">ref &lt;-<span class="st"> </span><span class="kw">getSeq</span>(result)</a>
<a class="sourceLine" id="cb52-4" data-line-number="4"><span class="kw">getSeq</span>(ref, <span class="kw">subset</span>(<span class="kw">cds</span>(result), gene <span class="op">%in%</span><span class="st"> &quot;S&quot;</span>))</a>
<a class="sourceLine" id="cb52-5" data-line-number="5">...</a></code></pre></div></li>
<li><p>Debug an updated version of the script downloading just a few records. Then download <em>all</em> records (this could take 10’s of minutes).</p></li>
<li><p>Perform the multiple alignment on S gene sequences instead of whole-genome sequence</p></li>
</ul>
</div>
<div id="phylogenetic-tree-visualization" class="section level3">
<h3><span class="header-section-number">5.3.2</span> Phylogenetic tree visualization</h3>
<p>Setup</p>
<ul>
<li><p>Install (if necessary, using <code>BiocManager::install(&quot;ggtree&quot;)</code>) and attach the <a href="https://bioconductor.org/packages/ggtree">ggtree</a> package</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb53-1" data-line-number="1"><span class="kw">library</span>(ggtree)</a>
<a class="sourceLine" id="cb53-2" data-line-number="2"></a>
<a class="sourceLine" id="cb53-3" data-line-number="3"><span class="kw">maintainer</span>(<span class="st">&quot;ggtree&quot;</span>)</a>
<a class="sourceLine" id="cb53-4" data-line-number="4"><span class="co">## [1] &quot;Guangchuang Yu &lt;guangchuangyu@gmail.com&gt;&quot;</span></a>
<a class="sourceLine" id="cb53-5" data-line-number="5"></a>
<a class="sourceLine" id="cb53-6" data-line-number="6"><span class="kw">citation</span>(<span class="st">&quot;ggtree&quot;</span>)</a>
<a class="sourceLine" id="cb53-7" data-line-number="7"><span class="co">## </span></a>
<a class="sourceLine" id="cb53-8" data-line-number="8"><span class="co">## To cite ggtree in publications use:</span></a>
<a class="sourceLine" id="cb53-9" data-line-number="9"><span class="co">## </span></a>
<a class="sourceLine" id="cb53-10" data-line-number="10"><span class="co">##   Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk</span></a>
<a class="sourceLine" id="cb53-11" data-line-number="11"><span class="co">##   Lam. ggtree: an R package for visualization and annotation of</span></a>
<a class="sourceLine" id="cb53-12" data-line-number="12"><span class="co">##   phylogenetic trees with their covariates and other associated data.</span></a>
<a class="sourceLine" id="cb53-13" data-line-number="13"><span class="co">##   Methods in Ecology and Evolution 2017, 8(1):28-36</span></a>
<a class="sourceLine" id="cb53-14" data-line-number="14"><span class="co">## </span></a>
<a class="sourceLine" id="cb53-15" data-line-number="15"><span class="co">##   Guangchuang Yu, Tommy Tsan-Yuk Lam, Huachen Zhu, Yi Guan. Two methods</span></a>
<a class="sourceLine" id="cb53-16" data-line-number="16"><span class="co">##   for mapping and visualizing associated data on phylogeny using</span></a>
<a class="sourceLine" id="cb53-17" data-line-number="17"><span class="co">##   ggtree. Molecular Biology and Evolution 2018, 35(2):3041-3043. doi:</span></a>
<a class="sourceLine" id="cb53-18" data-line-number="18"><span class="co">##   10.1093/molbev/msy194</span></a>
<a class="sourceLine" id="cb53-19" data-line-number="19"><span class="co">## </span></a>
<a class="sourceLine" id="cb53-20" data-line-number="20"><span class="co">## To see these entries in BibTeX format, use &#39;print(&lt;citation&gt;,</span></a>
<a class="sourceLine" id="cb53-21" data-line-number="21"><span class="co">## bibtex=TRUE)&#39;, &#39;toBibtex(.)&#39;, or set</span></a>
<a class="sourceLine" id="cb53-22" data-line-number="22"><span class="co">## &#39;options(citation.bibtex.max=999)&#39;.</span></a></code></pre></div></li>
<li><p>The visualizations this package provides are not well-suited to <em>R</em>’s help pages; check out the online <a href="https://yulab-smu.github.io/treedata-book">ggtree book</a> to orient yourself.</p></li>
</ul>
<p>Data preparation and initial visualization</p>
<ul>
<li><p>Read the data from the Newick-format file created by DECIPHER</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb54-1" data-line-number="1">dendrogram_file &lt;-<span class="st"> </span><span class="kw">file.path</span>(workdir, <span class="st">&quot;SARS-CoV2-NY.newick&quot;</span>)</a>
<a class="sourceLine" id="cb54-2" data-line-number="2">tree &lt;-<span class="st"> </span>treeio<span class="op">::</span><span class="kw">read.newick</span>(dendrogram_file)</a>
<a class="sourceLine" id="cb54-3" data-line-number="3">tree</a>
<a class="sourceLine" id="cb54-4" data-line-number="4"><span class="co">## </span></a>
<a class="sourceLine" id="cb54-5" data-line-number="5"><span class="co">## Phylogenetic tree with 245 tips and 244 internal nodes.</span></a>
<a class="sourceLine" id="cb54-6" data-line-number="6"><span class="co">## </span></a>
<a class="sourceLine" id="cb54-7" data-line-number="7"><span class="co">## Tip labels:</span></a>
<a class="sourceLine" id="cb54-8" data-line-number="8"><span class="co">##  MT370841, MT370857, MT370848, MT370954, MT370985, MT370842, ...</span></a>
<a class="sourceLine" id="cb54-9" data-line-number="9"><span class="co">## </span></a>
<a class="sourceLine" id="cb54-10" data-line-number="10"><span class="co">## Rooted; includes branch lengths.</span></a>
<a class="sourceLine" id="cb54-11" data-line-number="11">tibble<span class="op">::</span><span class="kw">as_tibble</span>(tree)</a>
<a class="sourceLine" id="cb54-12" data-line-number="12"><span class="co">## # A tibble: 489 x 4</span></a>
<a class="sourceLine" id="cb54-13" data-line-number="13"><span class="co">##    parent  node branch.length label   </span></a>
<a class="sourceLine" id="cb54-14" data-line-number="14"><span class="co">##     &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt;   </span></a>
<a class="sourceLine" id="cb54-15" data-line-number="15"><span class="co">##  1    248     1   0.000336    MT370841</span></a>
<a class="sourceLine" id="cb54-16" data-line-number="16"><span class="co">##  2    249     2   0.000168    MT370857</span></a>
<a class="sourceLine" id="cb54-17" data-line-number="17"><span class="co">##  3    250     3   0.000639    MT370848</span></a>
<a class="sourceLine" id="cb54-18" data-line-number="18"><span class="co">##  4    252     4   0.000000316 MT370954</span></a>
<a class="sourceLine" id="cb54-19" data-line-number="19"><span class="co">##  5    253     5   0           MT370985</span></a>
<a class="sourceLine" id="cb54-20" data-line-number="20"><span class="co">##  6    253     6   0.000000117 MT370842</span></a>
<a class="sourceLine" id="cb54-21" data-line-number="21"><span class="co">##  7    254     7   0           MT370929</span></a>
<a class="sourceLine" id="cb54-22" data-line-number="22"><span class="co">##  8    255     8   0.0000227   MT370849</span></a>
<a class="sourceLine" id="cb54-23" data-line-number="23"><span class="co">##  9    256     9   0.000000558 MT434800</span></a>
<a class="sourceLine" id="cb54-24" data-line-number="24"><span class="co">## 10    256    10   0.0000329   MT434786</span></a>
<a class="sourceLine" id="cb54-25" data-line-number="25"><span class="co">## # … with 479 more rows</span></a></code></pre></div></li>
<li><p>plot the tree, adding tip labels to the longest branches</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb55-1" data-line-number="1"><span class="kw">ggtree</span>(tree) <span class="op">+</span></a>
<a class="sourceLine" id="cb55-2" data-line-number="2"><span class="st">    </span><span class="kw">geom_tiplab</span>(</a>
<a class="sourceLine" id="cb55-3" data-line-number="3">        <span class="kw">aes</span>(</a>
<a class="sourceLine" id="cb55-4" data-line-number="4">            <span class="dt">subset =</span> branch.length <span class="op">&gt;</span><span class="st"> </span><span class="fl">.0001</span>,</a>
<a class="sourceLine" id="cb55-5" data-line-number="5">            <span class="dt">label =</span> label</a>
<a class="sourceLine" id="cb55-6" data-line-number="6">        ),</a>
<a class="sourceLine" id="cb55-7" data-line-number="7">        <span class="dt">size=</span><span class="dv">2</span></a>
<a class="sourceLine" id="cb55-8" data-line-number="8">    )</a>
<a class="sourceLine" id="cb55-9" data-line-number="9"><span class="co">## Warning: `data_frame()` is deprecated as of tibble 1.1.0.</span></a>
<a class="sourceLine" id="cb55-10" data-line-number="10"><span class="co">## Please use `tibble()` instead.</span></a>
<a class="sourceLine" id="cb55-11" data-line-number="11"><span class="co">## This warning is displayed once every 8 hours.</span></a>
<a class="sourceLine" id="cb55-12" data-line-number="12"><span class="co">## Call `lifecycle::last_warnings()` to see where this warning was generated.</span></a></code></pre></div>
<p><img src="05-Bioinformatics_files/figure-html/unnamed-chunk-57-1.png" width="672" /></p></li>
<li><p>Collapse clades with short branches</p></li>
</ul>
</div>
</div>
<div id="day-32-single-cell-expression-data" class="section level2">
<h2><span class="header-section-number">5.4</span> Day 32 Single-cell expression data</h2>
</div>
<div id="day-33-friday-zoom-check-in" class="section level2">
<h2><span class="header-section-number">5.5</span> Day 33 (Friday) Zoom check-in</h2>
<div id="review-and-trouble-shoot-25-minutes-1" class="section level3">
<h3><span class="header-section-number">5.5.1</span> Review and trouble shoot (25 minutes)</h3>
<p><a href="https://bioconductor.org">Bioconductor</a></p>
<ul>
<li><p>Packages</p></li>
<li><p>Classes and methods</p></li>
</ul>
<p>SARS-CoV-2 sequence data</p>
<ul>
<li><p>Querying Entrez and parsing GenBank records</p></li>
<li><p>Working with DNA sequences and genomic ranges</p></li>
<li><p>Multiple alignment and phylogenetic tree visualization</p></li>
</ul>
<p>Single cell expression</p>
</div>
<div id="next-week-25-minutes" class="section level3">
<h3><span class="header-section-number">5.5.2</span> Next week (25 minutes)</h3>
</div>
</div>
<div id="day-34" class="section level2">
<h2><span class="header-section-number">5.6</span> Day 34</h2>
</div>
<div id="day-35" class="section level2">
<h2><span class="header-section-number">5.7</span> Day 35</h2>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="four.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="six.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["QuaRantine.pdf", "QuaRantine.epub"],
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
