<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Week 5 Bioinformatics with Bioconductor | 40 Days and 40 Nights</title>
  <meta name="description" content="An ad hoc course to learn R for bioinformatics, using the
COVID-19 epidemic as an excuse." />
  <meta name="generator" content="bookdown 0.18 and GitBook 2.6.7" />

  <meta property="og:title" content="Week 5 Bioinformatics with Bioconductor | 40 Days and 40 Nights" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An ad hoc course to learn R for bioinformatics, using the
COVID-19 epidemic as an excuse." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Week 5 Bioinformatics with Bioconductor | 40 Days and 40 Nights" />
  
  <meta name="twitter:description" content="An ad hoc course to learn R for bioinformatics, using the
COVID-19 epidemic as an excuse." />
  

<meta name="author" content="Martin Morgan" />
<meta name="author" content="L. Shawn Matott" />


<meta name="date" content="2020-05-22" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="four.html"/>
<link rel="next" href="six.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">QuaRantine</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Motivation</a><ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#introduction"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#what-to-expect"><i class="fa fa-check"></i>What to expect</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="one.html"><a href="one.html"><i class="fa fa-check"></i><b>1</b> Basics</a><ul>
<li class="chapter" data-level="1.1" data-path="one.html"><a href="one.html#day-1-monday-zoom-orientation"><i class="fa fa-check"></i><b>1.1</b> Day 1 (Monday) Zoom orientation</a><ul>
<li class="chapter" data-level="1.1.1" data-path="one.html"><a href="one.html#logistics-10-minutes"><i class="fa fa-check"></i><b>1.1.1</b> Logistics (10 minutes)</a></li>
<li class="chapter" data-level="1.1.2" data-path="one.html"><a href="one.html#installing-r-and-rstudio-25-minutes-shawn"><i class="fa fa-check"></i><b>1.1.2</b> Installing <em>R</em> and <em>RStudio</em> (25 minutes, Shawn)</a></li>
<li class="chapter" data-level="1.1.3" data-path="one.html"><a href="one.html#basics-of-r-25-minutes"><i class="fa fa-check"></i><b>1.1.3</b> Basics of <em>R</em> (25 minutes)</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="one.html"><a href="one.html#day-2-vectors-and-variables"><i class="fa fa-check"></i><b>1.2</b> Day 2: Vectors and variables</a></li>
<li class="chapter" data-level="1.3" data-path="one.html"><a href="one.html#day-3-factor-date-and-na"><i class="fa fa-check"></i><b>1.3</b> Day 3: <code>factor()</code>, <code>Date()</code>, and <code>NA</code></a></li>
<li class="chapter" data-level="1.4" data-path="one.html"><a href="one.html#day-4-working-with-variables"><i class="fa fa-check"></i><b>1.4</b> Day 4: Working with variables</a></li>
<li class="chapter" data-level="1.5" data-path="one.html"><a href="one.html#day-5-friday-zoom-check-in"><i class="fa fa-check"></i><b>1.5</b> Day 5 (Friday) Zoom check-in</a><ul>
<li class="chapter" data-level="1.5.1" data-path="one.html"><a href="one.html#logistics"><i class="fa fa-check"></i><b>1.5.1</b> Logistics</a></li>
<li class="chapter" data-level="1.5.2" data-path="one.html"><a href="one.html#review-and-trouble-shoot-25-minutes-martin"><i class="fa fa-check"></i><b>1.5.2</b> Review and trouble shoot (25 minutes; Martin)</a></li>
<li class="chapter" data-level="1.5.3" data-path="one.html"><a href="one.html#weekend-activities-25-minutes-shawn"><i class="fa fa-check"></i><b>1.5.3</b> Weekend activities (25 minutes; Shawn)</a></li>
</ul></li>
<li class="chapter" data-level="1.6" data-path="one.html"><a href="one.html#day-6-r-scripts"><i class="fa fa-check"></i><b>1.6</b> Day 6: <em>R</em> scripts</a></li>
<li class="chapter" data-level="1.7" data-path="one.html"><a href="one.html#day-7-saving-data"><i class="fa fa-check"></i><b>1.7</b> Day 7: Saving data</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="two.html"><a href="two.html"><i class="fa fa-check"></i><b>2</b> The data frame</a><ul>
<li class="chapter" data-level="2.1" data-path="two.html"><a href="two.html#day-8-monday-zoom-check-in"><i class="fa fa-check"></i><b>2.1</b> Day 8 (Monday) Zoom check-in</a><ul>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#logistics-1"><i class="fa fa-check"></i>Logistics</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#review-and-troubleshoot-15-minutes"><i class="fa fa-check"></i>Review and troubleshoot (15 minutes)</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#the-data-frame-40-minutes"><i class="fa fa-check"></i>The data frame (40 minutes)</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#this-weeks-activities-5-minutes"><i class="fa fa-check"></i>This week’s activities (5 minutes)</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="two.html"><a href="two.html#day-9-creation-and-manipulation"><i class="fa fa-check"></i><b>2.2</b> Day 9: Creation and manipulation</a><ul>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#creation"><i class="fa fa-check"></i>Creation</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#column-selection"><i class="fa fa-check"></i>Column selection</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#adding-or-updating-columns"><i class="fa fa-check"></i>Adding or updating columns</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#reading-and-writing"><i class="fa fa-check"></i>Reading and writing</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#reading-from-a-remote-file"><i class="fa fa-check"></i>Reading from a remote file (!)</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="two.html"><a href="two.html#day-10-subset-with-and-within"><i class="fa fa-check"></i><b>2.3</b> Day 10: <code>subset()</code>, <code>with()</code>, and <code>within()</code></a><ul>
<li><a href="two.html#subset"><code>subset()</code></a></li>
<li><a href="two.html#with"><code>with()</code></a></li>
<li><a href="two.html#within"><code>within()</code></a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="two.html"><a href="two.html#day-11-aggregate-and-an-initial-work-flow"><i class="fa fa-check"></i><b>2.4</b> Day 11: <code>aggregate()</code> and an initial work flow</a><ul>
<li><a href="two.html#aggregate-for-summarizing-columns-by-group"><code>aggregate()</code> for summarizing columns by group</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#an-initial-work-flow"><i class="fa fa-check"></i>An initial work flow</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="two.html"><a href="two.html#day-12-friday-zoom-check-in"><i class="fa fa-check"></i><b>2.5</b> Day 12 (Friday) Zoom check-in</a><ul>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#review-and-troubleshoot-20-minutes"><i class="fa fa-check"></i>Review and troubleshoot (20 minutes)</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#user-defined-functions"><i class="fa fa-check"></i>User-defined functions</a></li>
<li class="chapter" data-level="" data-path="two.html"><a href="two.html#plotting-data"><i class="fa fa-check"></i>Plotting data</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="two.html"><a href="two.html#day-13-basic-visualization"><i class="fa fa-check"></i><b>2.6</b> Day 13: Basic visualization</a></li>
<li class="chapter" data-level="2.7" data-path="two.html"><a href="two.html#day-14-functions"><i class="fa fa-check"></i><b>2.7</b> Day 14: Functions</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="three.html"><a href="three.html"><i class="fa fa-check"></i><b>3</b> Packages and the ‘tidyverse’</a><ul>
<li class="chapter" data-level="3.1" data-path="three.html"><a href="three.html#day-15-monday-zoom-check-in"><i class="fa fa-check"></i><b>3.1</b> Day 15 (Monday) Zoom check-in</a><ul>
<li class="chapter" data-level="" data-path="three.html"><a href="three.html#review-and-troubleshoot-15-minutes-1"><i class="fa fa-check"></i>Review and troubleshoot (15 minutes)</a></li>
<li class="chapter" data-level="" data-path="three.html"><a href="three.html#packages-20-minutes"><i class="fa fa-check"></i>Packages (20 minutes)</a></li>
<li class="chapter" data-level="" data-path="three.html"><a href="three.html#the-tidyverse-of-packages-20-minutes"><i class="fa fa-check"></i>The ‘tidyverse’ of packages (20 minutes)</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="three.html"><a href="three.html#day-16-key-tidyverse-packages-readr-and-dplyr"><i class="fa fa-check"></i><b>3.2</b> Day 16 Key tidyverse packages: <span>readr</span> and <span>dplyr</span></a></li>
<li class="chapter" data-level="3.3" data-path="three.html"><a href="three.html#day-17-visualization-with-ggplot2"><i class="fa fa-check"></i><b>3.3</b> Day 17 Visualization with <span>ggplot2</span></a><ul>
<li class="chapter" data-level="" data-path="three.html"><a href="three.html#setup"><i class="fa fa-check"></i>Setup</a></li>
<li><a href="three.html#ggplot2-essentials"><span>ggplot2</span> essentials</a></li>
<li class="chapter" data-level="" data-path="three.html"><a href="three.html#covid-19-in-erie-county"><i class="fa fa-check"></i>COVID-19 in Erie county</a></li>
<li class="chapter" data-level="" data-path="three.html"><a href="three.html#covid-19-in-new-york-state"><i class="fa fa-check"></i>COVID-19 in New York State</a></li>
<li class="chapter" data-level="" data-path="three.html"><a href="three.html#covid-19-nationally"><i class="fa fa-check"></i>COVID-19 nationally</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="three.html"><a href="three.html#day-18-worldwide-covid-data"><i class="fa fa-check"></i><b>3.4</b> Day 18 Worldwide COVID data</a></li>
<li class="chapter" data-level="3.5" data-path="three.html"><a href="three.html#day-19-friday-zoom-check-in"><i class="fa fa-check"></i><b>3.5</b> Day 19 (Friday) Zoom check-in</a><ul>
<li class="chapter" data-level="3.5.1" data-path="three.html"><a href="three.html#logistics-2"><i class="fa fa-check"></i><b>3.5.1</b> Logistics</a></li>
<li class="chapter" data-level="3.5.2" data-path="three.html"><a href="three.html#review-and-trouble-shoot-40-minutes"><i class="fa fa-check"></i><b>3.5.2</b> Review and trouble shoot (40 minutes)</a></li>
<li class="chapter" data-level="3.5.3" data-path="three.html"><a href="three.html#weekend-activities-15-minutes"><i class="fa fa-check"></i><b>3.5.3</b> Weekend activities (15 minutes)</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="three.html"><a href="three.html#day-20-exploring-the-course-of-pandemic-in-different-regions"><i class="fa fa-check"></i><b>3.6</b> Day 20 Exploring the course of pandemic in different regions</a></li>
<li class="chapter" data-level="3.7" data-path="three.html"><a href="three.html#day-21-critical-evaluation"><i class="fa fa-check"></i><b>3.7</b> Day 21 Critical evaluation</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="four.html"><a href="four.html"><i class="fa fa-check"></i><b>4</b> Machine learning</a><ul>
<li class="chapter" data-level="4.1" data-path="four.html"><a href="four.html#day-22-monday-zoom-check-in"><i class="fa fa-check"></i><b>4.1</b> Day 22 (Monday) Zoom check-in</a><ul>
<li class="chapter" data-level="4.1.1" data-path="four.html"><a href="four.html#a-covid-19-dataset-for-machine-learning"><i class="fa fa-check"></i><b>4.1.1</b> A COVID-19 Dataset for Machine Learning</a></li>
<li class="chapter" data-level="4.1.2" data-path="four.html"><a href="four.html#a-machine-learning-primer"><i class="fa fa-check"></i><b>4.1.2</b> A Machine Learning Primer</a></li>
<li class="chapter" data-level="4.1.3" data-path="four.html"><a href="four.html#the-random-forest-algorithm"><i class="fa fa-check"></i><b>4.1.3</b> The Random Forest Algorithm</a></li>
<li class="chapter" data-level="4.1.4" data-path="four.html"><a href="four.html#a-random-forest-example-using-covid-19-data"><i class="fa fa-check"></i><b>4.1.4</b> A Random Forest Example Using COVID-19 Data</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="four.html"><a href="four.html#day-23---support-vector-machines"><i class="fa fa-check"></i><b>4.2</b> Day 23 - Support Vector Machines</a><ul>
<li class="chapter" data-level="4.2.1" data-path="four.html"><a href="four.html#an-svm-example-using-covid-19"><i class="fa fa-check"></i><b>4.2.1</b> An SVM Example using COVID-19</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="four.html"><a href="four.html#day-24---the-k-nearest-neighbors-algorithm"><i class="fa fa-check"></i><b>4.3</b> Day 24 - the <span class="math inline">\(k\)</span>-Nearest Neighbors Algorithm</a></li>
<li class="chapter" data-level="4.4" data-path="four.html"><a href="four.html#day-25---exploring-the-knn-algorithm"><i class="fa fa-check"></i><b>4.4</b> Day 25 - Exploring the KNN Algorithm</a></li>
<li class="chapter" data-level="4.5" data-path="four.html"><a href="four.html#day-26-friday-zoom-check-in"><i class="fa fa-check"></i><b>4.5</b> Day 26 (Friday) Zoom check-in</a><ul>
<li class="chapter" data-level="4.5.1" data-path="four.html"><a href="four.html#review-and-trouble-shoot-25-minutes"><i class="fa fa-check"></i><b>4.5.1</b> Review and trouble shoot (25 minutes)</a></li>
<li class="chapter" data-level="4.5.2" data-path="four.html"><a href="four.html#this-weekend-25-minutes"><i class="fa fa-check"></i><b>4.5.2</b> This weekend (25 minutes)</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="four.html"><a href="four.html#day-27"><i class="fa fa-check"></i><b>4.6</b> Day 27</a></li>
<li class="chapter" data-level="4.7" data-path="four.html"><a href="four.html#day-28"><i class="fa fa-check"></i><b>4.7</b> Day 28</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="five.html"><a href="five.html"><i class="fa fa-check"></i><b>5</b> Bioinformatics with <span>Bioconductor</span></a><ul>
<li class="chapter" data-level="5.1" data-path="five.html"><a href="five.html#day-29-monday-zoom-check-in"><i class="fa fa-check"></i><b>5.1</b> Day 29 (Monday) Zoom check-in</a><ul>
<li class="chapter" data-level="5.1.1" data-path="five.html"><a href="five.html#review-and-trouble-shoot"><i class="fa fa-check"></i><b>5.1.1</b> Review and trouble shoot</a></li>
<li class="chapter" data-level="5.1.2" data-path="five.html"><a href="five.html#bioconductor-packages-and-biological-data"><i class="fa fa-check"></i><b>5.1.2</b> <span>Bioconductor</span>, packages, and biological data</a></li>
<li class="chapter" data-level="5.1.3" data-path="five.html"><a href="five.html#this-week-sars-cov-2-sequence-and-human-tissue-specific-gene-expression-data"><i class="fa fa-check"></i><b>5.1.3</b> This week: SARS-CoV-2 sequence and human tissue-specific gene expression data</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="five.html"><a href="five.html#day-30-dna-sequences-and-annotations"><i class="fa fa-check"></i><b>5.2</b> Day 30 DNA sequences and annotations</a><ul>
<li class="chapter" data-level="5.2.1" data-path="five.html"><a href="five.html#setup-1"><i class="fa fa-check"></i><b>5.2.1</b> Setup</a></li>
<li class="chapter" data-level="5.2.2" data-path="five.html"><a href="five.html#representing-dna-sequences"><i class="fa fa-check"></i><b>5.2.2</b> Representing DNA sequences</a></li>
<li class="chapter" data-level="5.2.3" data-path="five.html"><a href="five.html#working-with-genbankrecord-objects"><i class="fa fa-check"></i><b>5.2.3</b> Working with <code>GenBankRecord</code> objects</a></li>
<li class="chapter" data-level="5.2.4" data-path="five.html"><a href="five.html#genomic-ranges-and-dna-sequences"><i class="fa fa-check"></i><b>5.2.4</b> Genomic ranges and DNA sequences</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="five.html"><a href="five.html#day-31-sequence-alignment-and-visualization"><i class="fa fa-check"></i><b>5.3</b> Day 31 Sequence alignment and visualization</a><ul>
<li class="chapter" data-level="5.3.1" data-path="five.html"><a href="five.html#s-gene-exploration"><i class="fa fa-check"></i><b>5.3.1</b> ‘S’ gene exploration</a></li>
<li class="chapter" data-level="5.3.2" data-path="five.html"><a href="five.html#sequence-alignment"><i class="fa fa-check"></i><b>5.3.2</b> Sequence alignment</a></li>
<li class="chapter" data-level="5.3.3" data-path="five.html"><a href="five.html#phylogenetic-tree-visualization"><i class="fa fa-check"></i><b>5.3.3</b> Phylogenetic tree visualization</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="five.html"><a href="five.html#day-32-single-cell-expression-data"><i class="fa fa-check"></i><b>5.4</b> Day 32 Single-cell expression data</a><ul>
<li class="chapter" data-level="5.4.1" data-path="five.html"><a href="five.html#setup-2"><i class="fa fa-check"></i><b>5.4.1</b> Setup</a></li>
<li class="chapter" data-level="5.4.2" data-path="five.html"><a href="five.html#data-input-and-exploration"><i class="fa fa-check"></i><b>5.4.2</b> Data input and exploration</a></li>
<li class="chapter" data-level="5.4.3" data-path="five.html"><a href="five.html#coordinating-cell-and-expression-data"><i class="fa fa-check"></i><b>5.4.3</b> Coordinating cell and expression data</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="five.html"><a href="five.html#day-33-friday-zoom-check-in"><i class="fa fa-check"></i><b>5.5</b> Day 33 (Friday) Zoom check-in</a><ul>
<li class="chapter" data-level="5.5.1" data-path="five.html"><a href="five.html#monday-troubles-10-minutes"><i class="fa fa-check"></i><b>5.5.1</b> Monday troubles (10 minutes)</a></li>
<li class="chapter" data-level="5.5.2" data-path="five.html"><a href="five.html#review-and-trouble-shoot-40-minutes-1"><i class="fa fa-check"></i><b>5.5.2</b> Review and trouble shoot (40 minutes)</a></li>
<li class="chapter" data-level="5.5.3" data-path="five.html"><a href="five.html#this-weekend-10-minutes"><i class="fa fa-check"></i><b>5.5.3</b> This weekend (10 minutes)</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="five.html"><a href="five.html#day-34-revising-tree-visualization"><i class="fa fa-check"></i><b>5.6</b> Day 34 Revising tree visualization</a></li>
<li class="chapter" data-level="5.7" data-path="five.html"><a href="five.html#day-35-exploring-bioconductor"><i class="fa fa-check"></i><b>5.7</b> Day 35 Exploring <span>Bioconductor</span></a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="six.html"><a href="six.html"><i class="fa fa-check"></i><b>6</b> Collaboration</a><ul>
<li class="chapter" data-level="6.1" data-path="six.html"><a href="six.html#days-monday-zoom-check-in"><i class="fa fa-check"></i><b>6.1</b> 5 Days (Monday) Zoom check-in</a><ul>
<li class="chapter" data-level="6.1.1" data-path="six.html"><a href="six.html#weekend-review-5-minutes"><i class="fa fa-check"></i><b>6.1.1</b> Weekend review (5 minutes)</a></li>
<li class="chapter" data-level="6.1.2" data-path="six.html"><a href="six.html#vignettes-25-minutes-shawn"><i class="fa fa-check"></i><b>6.1.2</b> Vignettes (25 minutes; Shawn)</a></li>
<li class="chapter" data-level="6.1.3" data-path="six.html"><a href="six.html#writing-and-documenting-r-functions-25-minutes"><i class="fa fa-check"></i><b>6.1.3</b> Writing and documenting R functions (25 minutes)</a></li>
<li class="chapter" data-level="6.1.4" data-path="six.html"><a href="six.html#create-an-r-pacakge"><i class="fa fa-check"></i><b>6.1.4</b> Create an <em>R</em> pacakge!</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="six.html"><a href="six.html#days-write-a-vignette"><i class="fa fa-check"></i><b>6.2</b> 4 Days Write a vignette!</a></li>
<li class="chapter" data-level="6.3" data-path="six.html"><a href="six.html#days-create-documented-reusable-functions"><i class="fa fa-check"></i><b>6.3</b> 3 Days Create documented, reusable functions!</a></li>
<li class="chapter" data-level="6.4" data-path="six.html"><a href="six.html#days-share-your-work-as-a-package"><i class="fa fa-check"></i><b>6.4</b> 2 Days Share your work as a package!</a><ul>
<li class="chapter" data-level="6.4.1" data-path="six.html"><a href="six.html#description"><i class="fa fa-check"></i><b>6.4.1</b> DESCRIPTION</a></li>
<li class="chapter" data-level="6.4.2" data-path="six.html"><a href="six.html#documentation"><i class="fa fa-check"></i><b>6.4.2</b> Documentation</a></li>
<li class="chapter" data-level="6.4.3" data-path="six.html"><a href="six.html#vignettes"><i class="fa fa-check"></i><b>6.4.3</b> Vignettes</a></li>
<li class="chapter" data-level="6.4.4" data-path="six.html"><a href="six.html#build-and-share"><i class="fa fa-check"></i><b>6.4.4</b> Build and share</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="six.html"><a href="six.html#today-friday-zoom-check-in"><i class="fa fa-check"></i><b>6.5</b> Today! (Friday) Zoom check-in</a><ul>
<li class="chapter" data-level="6.5.1" data-path="six.html"><a href="six.html#review-and-troubleshoot"><i class="fa fa-check"></i><b>6.5.1</b> Review and troubleshoot</a></li>
<li class="chapter" data-level="6.5.2" data-path="six.html"><a href="six.html#course-review"><i class="fa fa-check"></i><b>6.5.2</b> Course review</a></li>
<li class="chapter" data-level="6.5.3" data-path="six.html"><a href="six.html#feedback-next-steps"><i class="fa fa-check"></i><b>6.5.3</b> Feedback &amp; next steps</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="acknowledgments.html"><a href="acknowledgments.html"><i class="fa fa-check"></i>Acknowledgments</a></li>
<li class="chapter" data-level="" data-path="frequently-asked-questions.html"><a href="frequently-asked-questions.html"><i class="fa fa-check"></i>Frequently asked questions</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">40 Days and 40 Nights</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="five" class="section level1">
<h1><span class="header-section-number">Week 5</span> Bioinformatics with <a href="https://bioconductor.org">Bioconductor</a></h1>
<div id="day-29-monday-zoom-check-in" class="section level2">
<h2><span class="header-section-number">5.1</span> Day 29 (Monday) Zoom check-in</h2>
<div id="review-and-trouble-shoot" class="section level3">
<h3><span class="header-section-number">5.1.1</span> Review and trouble shoot</h3>
</div>
<div id="bioconductor-packages-and-biological-data" class="section level3">
<h3><span class="header-section-number">5.1.2</span> <a href="https://bioconductor.org">Bioconductor</a>, packages, and biological data</h3>
<p><a href="https://bioconductor.org">Bioconductor</a></p>
<ul>
<li><p>Repository of more than 1900 packages for the ‘analysis and comprehension of high throughput genomic data’</p></li>
<li><p>Bulk and single cell RNASeq; ChIP and other gene regulation, called variants, flow cytometery, proteomics, …</p></li>
<li><p>Package discovery using <a href="https://bioconductor.org/packages">biocViews</a> and <a href="https://bioconductor.org/packages/release/BiocViews.html#___Workflow">workflows</a></p></li>
<li><p>Landing pages &amp; vignettes</p></li>
</ul>
<p>Packages</p>
<ul>
<li><p>One-time installation of <a href="https://cran.r-project.org/package=BiocManager">BiocManager</a></p>
<div class="sourceCode" id="cb316"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb316-1" data-line-number="1"><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>))</a>
<a class="sourceLine" id="cb316-2" data-line-number="2">    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">repos =</span> <span class="st">&quot;https://cran.r-project.org&quot;</span>)</a></code></pre></div></li>
<li><p>Install <em>Bioconductor</em> or <em>CRAN</em> packages</p>
<div class="sourceCode" id="cb317"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb317-1" data-line-number="1"><span class="co">## Biostrings and genbankr are Bioconductor packages, rentrez is from CRAN</span></a>
<a class="sourceLine" id="cb317-2" data-line-number="2">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="kw">c</span>(<span class="st">&quot;Biostrings&quot;</span>, <span class="st">&quot;rentrez&quot;</span>))</a></code></pre></div></li>
<li><p>Validate installation</p>
<div class="sourceCode" id="cb318"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb318-1" data-line-number="1">BiocManager<span class="op">::</span><span class="kw">version</span>() <span class="co"># current release is 3.11; needs R 4.0</span></a>
<a class="sourceLine" id="cb318-2" data-line-number="2"><span class="co">## [1] &#39;3.10&#39;</span></a>
<a class="sourceLine" id="cb318-3" data-line-number="3">BiocManager<span class="op">::</span><span class="kw">valid</span>()</a>
<a class="sourceLine" id="cb318-4" data-line-number="4"><span class="co">## Warning: 10 packages out-of-date; 0 packages too new</span></a>
<a class="sourceLine" id="cb318-5" data-line-number="5"><span class="co">## </span></a>
<a class="sourceLine" id="cb318-6" data-line-number="6"><span class="co">## * sessionInfo()</span></a>
<a class="sourceLine" id="cb318-7" data-line-number="7"><span class="co">## </span></a>
<a class="sourceLine" id="cb318-8" data-line-number="8"><span class="co">## R version 3.6.3 Patched (2020-02-29 r77906)</span></a>
<a class="sourceLine" id="cb318-9" data-line-number="9"><span class="co">## Platform: x86_64-apple-darwin17.7.0 (64-bit)</span></a>
<a class="sourceLine" id="cb318-10" data-line-number="10"><span class="co">## Running under: macOS High Sierra 10.13.6</span></a>
<a class="sourceLine" id="cb318-11" data-line-number="11"><span class="co">## </span></a>
<a class="sourceLine" id="cb318-12" data-line-number="12"><span class="co">## Matrix products: default</span></a>
<a class="sourceLine" id="cb318-13" data-line-number="13"><span class="co">## BLAS:   /Users/ma38727/bin/R-3-6-branch/lib/libRblas.dylib</span></a>
<a class="sourceLine" id="cb318-14" data-line-number="14"><span class="co">## LAPACK: /Users/ma38727/bin/R-3-6-branch/lib/libRlapack.dylib</span></a>
<a class="sourceLine" id="cb318-15" data-line-number="15"><span class="co">## </span></a>
<a class="sourceLine" id="cb318-16" data-line-number="16"><span class="co">## locale:</span></a>
<a class="sourceLine" id="cb318-17" data-line-number="17"><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></a>
<a class="sourceLine" id="cb318-18" data-line-number="18"><span class="co">## </span></a>
<a class="sourceLine" id="cb318-19" data-line-number="19"><span class="co">## attached base packages:</span></a>
<a class="sourceLine" id="cb318-20" data-line-number="20"><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></a>
<a class="sourceLine" id="cb318-21" data-line-number="21"><span class="co">## </span></a>
<a class="sourceLine" id="cb318-22" data-line-number="22"><span class="co">## loaded via a namespace (and not attached):</span></a>
<a class="sourceLine" id="cb318-23" data-line-number="23"><span class="co">##  [1] BiocManager_1.30.10 compiler_3.6.3      magrittr_1.5       </span></a>
<a class="sourceLine" id="cb318-24" data-line-number="24"><span class="co">##  [4] bookdown_0.18       htmltools_0.4.0     tools_3.6.3        </span></a>
<a class="sourceLine" id="cb318-25" data-line-number="25"><span class="co">##  [7] yaml_2.2.1          Rcpp_1.0.4.6        stringi_1.4.6      </span></a>
<a class="sourceLine" id="cb318-26" data-line-number="26"><span class="co">## [10] rmarkdown_2.1       knitr_1.28          stringr_1.4.0      </span></a>
<a class="sourceLine" id="cb318-27" data-line-number="27"><span class="co">## [13] digest_0.6.25       xfun_0.13           rlang_0.4.6        </span></a>
<a class="sourceLine" id="cb318-28" data-line-number="28"><span class="co">## [16] evaluate_0.14      </span></a>
<a class="sourceLine" id="cb318-29" data-line-number="29"><span class="co">## </span></a>
<a class="sourceLine" id="cb318-30" data-line-number="30"><span class="co">## Bioconductor version &#39;3.10&#39;</span></a>
<a class="sourceLine" id="cb318-31" data-line-number="31"><span class="co">## </span></a>
<a class="sourceLine" id="cb318-32" data-line-number="32"><span class="co">##   * 10 packages out-of-date</span></a>
<a class="sourceLine" id="cb318-33" data-line-number="33"><span class="co">##   * 0 packages too new</span></a>
<a class="sourceLine" id="cb318-34" data-line-number="34"><span class="co">## </span></a>
<a class="sourceLine" id="cb318-35" data-line-number="35"><span class="co">## create a valid installation with</span></a>
<a class="sourceLine" id="cb318-36" data-line-number="36"><span class="co">## </span></a>
<a class="sourceLine" id="cb318-37" data-line-number="37"><span class="co">##   BiocManager::install(c(</span></a>
<a class="sourceLine" id="cb318-38" data-line-number="38"><span class="co">##     &quot;bookdown&quot;, &quot;ellipsis&quot;, &quot;modelr&quot;, &quot;natserv&quot;, &quot;RcppArmadillo&quot;, &quot;sp&quot;,</span></a>
<a class="sourceLine" id="cb318-39" data-line-number="39"><span class="co">##     &quot;taxize&quot;, &quot;tidyr&quot;, &quot;tinytex&quot;, &quot;xfun&quot;</span></a>
<a class="sourceLine" id="cb318-40" data-line-number="40"><span class="co">##   ), update = TRUE, ask = FALSE)</span></a>
<a class="sourceLine" id="cb318-41" data-line-number="41"><span class="co">## </span></a>
<a class="sourceLine" id="cb318-42" data-line-number="42"><span class="co">## more details: BiocManager::valid()$too_new, BiocManager::valid()$out_of_date</span></a></code></pre></div></li>
<li><p>Types of packages</p>
<ul>
<li>Software (e.g., <a href="https://bioconductor.org/packages/Biostrings">Biostrings</a>, <a href="https://bioconductor.org/packages/SingleCellExperiment">SingleCellExperiment</a>)</li>
<li>Annotation (e.g., <a href="https://bioconductor.org/packages/org.Hs.eg.db">org.Hs.eg.db</a>, <a href="https://bioconductor.org/packages/TxDb.Hsapiens.UCSC.hg38.knownGene">TxDb.Hsapiens.UCSC.hg38.knownGene</a>, <a href="https://bioconductor.org/packages/GO.db">GO.db</a>)</li>
<li>Experiment data (e.g., <a href="https://bioconductor.org/packages/airway">airway</a>, <a href="https://bioconductor.org/packages/scRNAseq">scRNAseq</a>)</li>
</ul></li>
</ul>
<p>Key packages</p>
<ul>
<li><a href="https://bioconductor.org/packages/Biostrings">Biostrings</a> for representing sequence data</li>
<li><a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a> for working with genomic coordinates</li>
<li><a href="https://bioconductor.org/packages/Summarizedexperiment">SummarizedExperiment</a> and <a href="https://bioconductor.org/packages/SingleCellExperiment">SingleCellExperiment</a> for representing experimental summaries, e.g., a gene x sample matrix of RNA-seq expression values.</li>
</ul>
<p>‘Class’ and ‘method’</p>
<ul>
<li><p>Classes represent data in ways that allow the specialized nature of the data to be exploited, e.g., it makes sense to calculate the <code>reverseComplement()</code> of a DNA sequence, but not of an arbitrary character vector.</p>
<div class="sourceCode" id="cb319"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb319-1" data-line-number="1"><span class="kw">library</span>(Biostrings)</a>
<a class="sourceLine" id="cb319-2" data-line-number="2">sequences &lt;-<span class="st"> </span><span class="kw">c</span>(</a>
<a class="sourceLine" id="cb319-3" data-line-number="3">    <span class="dt">chr1 =</span> <span class="st">&quot;CTGACT&quot;</span>,</a>
<a class="sourceLine" id="cb319-4" data-line-number="4">    <span class="dt">chr2 =</span> <span class="st">&quot;CTGGGAACT&quot;</span></a>
<a class="sourceLine" id="cb319-5" data-line-number="5">)</a>
<a class="sourceLine" id="cb319-6" data-line-number="6">dna &lt;-<span class="st"> </span><span class="kw">DNAStringSet</span>(sequences)</a>
<a class="sourceLine" id="cb319-7" data-line-number="7">dna</a>
<a class="sourceLine" id="cb319-8" data-line-number="8"><span class="co">##   A DNAStringSet instance of length 2</span></a>
<a class="sourceLine" id="cb319-9" data-line-number="9"><span class="co">##     width seq                                               names               </span></a>
<a class="sourceLine" id="cb319-10" data-line-number="10"><span class="co">## [1]     6 CTGACT                                            chr1</span></a>
<a class="sourceLine" id="cb319-11" data-line-number="11"><span class="co">## [2]     9 CTGGGAACT                                         chr2</span></a>
<a class="sourceLine" id="cb319-12" data-line-number="12"></a>
<a class="sourceLine" id="cb319-13" data-line-number="13"><span class="co">## fails! `T` not in the RNA alphabet</span></a>
<a class="sourceLine" id="cb319-14" data-line-number="14">result &lt;-<span class="st"> </span><span class="kw">try</span>( <span class="kw">RNAStringSet</span>(sequences) )</a>
<a class="sourceLine" id="cb319-15" data-line-number="15"><span class="co">## Error in .Call2(&quot;new_XStringSet_from_CHARACTER&quot;, class(x0), elementType(x0),  : </span></a>
<a class="sourceLine" id="cb319-16" data-line-number="16"><span class="co">##   key 84 (char &#39;T&#39;) not in lookup table</span></a></code></pre></div></li>
<li><p>Methods represent the implementation of common (‘generic’) operations, specialized to the specific classes that the method operates on.</p>
<div class="sourceCode" id="cb320"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb320-1" data-line-number="1"><span class="kw">length</span>(dna)</a>
<a class="sourceLine" id="cb320-2" data-line-number="2"><span class="co">## [1] 2</span></a>
<a class="sourceLine" id="cb320-3" data-line-number="3"><span class="kw">reverseComplement</span>(dna)</a>
<a class="sourceLine" id="cb320-4" data-line-number="4"><span class="co">##   A DNAStringSet instance of length 2</span></a>
<a class="sourceLine" id="cb320-5" data-line-number="5"><span class="co">##     width seq                                               names               </span></a>
<a class="sourceLine" id="cb320-6" data-line-number="6"><span class="co">## [1]     6 AGTCAG                                            chr1</span></a>
<a class="sourceLine" id="cb320-7" data-line-number="7"><span class="co">## [2]     9 AGTTCCCAG                                         chr2</span></a>
<a class="sourceLine" id="cb320-8" data-line-number="8"><span class="kw">translate</span>(dna)</a>
<a class="sourceLine" id="cb320-9" data-line-number="9"><span class="co">##   A AAStringSet instance of length 2</span></a>
<a class="sourceLine" id="cb320-10" data-line-number="10"><span class="co">##     width seq                                               names               </span></a>
<a class="sourceLine" id="cb320-11" data-line-number="11"><span class="co">## [1]     2 MT                                                chr1</span></a>
<a class="sourceLine" id="cb320-12" data-line-number="12"><span class="co">## [2]     3 MGT                                               chr2</span></a></code></pre></div>
<ul>
<li>E.g., one could <code>translate()</code> both a DNA sequence, and an RNA sequence, so there is both <code>translate,DNAStringSet-method</code> and <code>translate,RNAStringSet-method</code></li>
</ul></li>
<li><p>Not all operations are implemented as methods on a class</p>
<ul>
<li><p>E.g., <code>length()</code> is a concept shared by collections of DNA and RNA sequences, so the method is defined on a common ‘parent’ class</p></li>
<li><p>E.g., <code>dim()</code> of a two-dimensional object is sufficient to implement <code>nrow()</code> (as <code>dim()[1]</code>) and <code>ncol()</code> (as <code>dim()[2]</code>) so one could use a ‘plain old’ function <code>nrow &lt;- function(x) dim(x)[1]</code> for <code>nrow()</code> / <code>ncol()</code>, relying on a <code>dim()</code> generic and methods to provide specialized behavior.</p></li>
</ul></li>
</ul>
<p>Getting help</p>
<ul>
<li>Vignettes, e.g., <a href="">DESeq2 vignette</a>! <code>browseVignettes(package = &quot;DESeq2&quot;)</code></li>
<li>Help pages for classes, e.g., <code>?DNAStringSet</code>, <code>?GRanges</code></li>
<li><p>Discovering methods</p>
<div class="sourceCode" id="cb321"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb321-1" data-line-number="1"><span class="kw">methods</span>(<span class="st">&quot;reverse&quot;</span>)</a>
<a class="sourceLine" id="cb321-2" data-line-number="2"><span class="kw">methods</span>(<span class="dt">class =</span> <span class="st">&quot;DNAStringSet&quot;</span>)</a></code></pre></div></li>
<li><p>Looking at symbols made available by packages (after the package is attached to the <code>search()</code> path): <code>ls(&quot;package:GenomicRanges&quot;)</code></p></li>
</ul>
</div>
<div id="this-week-sars-cov-2-sequence-and-human-tissue-specific-gene-expression-data" class="section level3">
<h3><span class="header-section-number">5.1.3</span> This week: SARS-CoV-2 sequence and human tissue-specific gene expression data</h3>
<p>Biological background</p>
<ul>
<li><p>RNA virus (<a href="https://en.wikipedia.org/wiki/File:Coronavirus_virion_structure.svg">source</a>, license: <a href="https://creativecommons.org/licenses/by-sa/4.0/deed.en">CC BY-SA 4.0</a>)</p>
<p><img src="images/512px-Coronavirus_virion_structure.svg.png" width="60%" /></p></li>
<li><p>Key domains (<a href="https://en.wikipedia.org/wiki/File:SARS-CoV_genome_organization.jpg">source</a>, license: <a href="https://creativecommons.org/licenses/by/2.0/deed.en">CC BY 2</a>)</p>
<p><img src="images/SARS-CoV_genome_organization.jpg" width="100%" /></p></li>
</ul>
<p>Setup</p>
<ul>
<li><p>Some essential <em>Bioconductor</em> data structures</p>
<div class="sourceCode" id="cb322"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb322-1" data-line-number="1"><span class="kw">library</span>(Biostrings)</a>
<a class="sourceLine" id="cb322-2" data-line-number="2"><span class="kw">library</span>(GenomicRanges)</a></code></pre></div></li>
<li><p>Additional software</p>
<div class="sourceCode" id="cb323"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb323-1" data-line-number="1"><span class="kw">library</span>(rentrez)</a>
<a class="sourceLine" id="cb323-2" data-line-number="2"><span class="kw">library</span>(genbankr)</a>
<a class="sourceLine" id="cb323-3" data-line-number="3"><span class="kw">library</span>(DECIPHER)</a>
<a class="sourceLine" id="cb323-4" data-line-number="4"><span class="kw">library</span>(ggtree)</a></code></pre></div></li>
<li><p>Need to install software?</p>
<ul>
<li><p>Install <a href="https://cran.r-project.org/package=BiocManager">BiocManager</a></p>
<div class="sourceCode" id="cb324"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb324-1" data-line-number="1"><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>))</a>
<a class="sourceLine" id="cb324-2" data-line-number="2">    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">repos =</span> <span class="st">&quot;https://cran.r-project.org&quot;</span>)</a></code></pre></div></li>
<li><p>Install other <em>CRAN</em> or <em>Bioconductor</em> packges</p>
<div class="sourceCode" id="cb325"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb325-1" data-line-number="1">pkgs &lt;-<span class="st"> </span><span class="kw">c</span>(</a>
<a class="sourceLine" id="cb325-2" data-line-number="2">    <span class="co">## packages needing installation, e.g.,</span></a>
<a class="sourceLine" id="cb325-3" data-line-number="3">    <span class="st">&quot;Biostrings&quot;</span>, <span class="st">&quot;GenomicRanges&quot;</span>, <span class="st">&quot;genbankr&quot;</span>, <span class="st">&quot;rentrez&quot;</span>, <span class="st">&quot;DECIPHER&quot;</span>,</a>
<a class="sourceLine" id="cb325-4" data-line-number="4">    <span class="st">&quot;ggtree&quot;</span></a>
<a class="sourceLine" id="cb325-5" data-line-number="5">)</a>
<a class="sourceLine" id="cb325-6" data-line-number="6">BiocManager<span class="op">::</span><span class="kw">install</span>(pkgs)</a>
<a class="sourceLine" id="cb325-7" data-line-number="7">BiocManager<span class="op">::</span><span class="kw">valid</span>()</a></code></pre></div></li>
</ul></li>
</ul>
<p><a href="https://bioconductor.org/packages/Biostrings">Biostrings</a> and <a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></p>
<ul>
<li><p><a href="https://bioconductor.org/packages/Biostrings">Biostrings</a>: representing DNA sequences</p>
<div class="sourceCode" id="cb326"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb326-1" data-line-number="1">sequences &lt;-<span class="st"> </span><span class="kw">c</span>(</a>
<a class="sourceLine" id="cb326-2" data-line-number="2">    <span class="dt">chr1 =</span> <span class="st">&quot;CTGACT&quot;</span>,</a>
<a class="sourceLine" id="cb326-3" data-line-number="3">    <span class="dt">chr2 =</span> <span class="st">&quot;CTGGGAACT&quot;</span></a>
<a class="sourceLine" id="cb326-4" data-line-number="4">)</a>
<a class="sourceLine" id="cb326-5" data-line-number="5">dna &lt;-<span class="st"> </span><span class="kw">DNAStringSet</span>(sequences)</a>
<a class="sourceLine" id="cb326-6" data-line-number="6">dna</a>
<a class="sourceLine" id="cb326-7" data-line-number="7"><span class="co">##   A DNAStringSet instance of length 2</span></a>
<a class="sourceLine" id="cb326-8" data-line-number="8"><span class="co">##     width seq                                               names               </span></a>
<a class="sourceLine" id="cb326-9" data-line-number="9"><span class="co">## [1]     6 CTGACT                                            chr1</span></a>
<a class="sourceLine" id="cb326-10" data-line-number="10"><span class="co">## [2]     9 CTGGGAACT                                         chr2</span></a></code></pre></div></li>
<li><p><a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a>: annotations in genome space, e.g., create a variable representing ‘regions of interest’</p>
<div class="sourceCode" id="cb327"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb327-1" data-line-number="1">regions_of_interest &lt;-<span class="st"> </span><span class="kw">c</span>(</a>
<a class="sourceLine" id="cb327-2" data-line-number="2">    <span class="st">&quot;chr1:1-3&quot;</span>,</a>
<a class="sourceLine" id="cb327-3" data-line-number="3">    <span class="st">&quot;chr1:3-6&quot;</span>,</a>
<a class="sourceLine" id="cb327-4" data-line-number="4">    <span class="st">&quot;chr2:4-6&quot;</span></a>
<a class="sourceLine" id="cb327-5" data-line-number="5">)</a>
<a class="sourceLine" id="cb327-6" data-line-number="6">roi &lt;-<span class="st"> </span><span class="kw">GRanges</span>(regions_of_interest)</a>
<a class="sourceLine" id="cb327-7" data-line-number="7">roi</a>
<a class="sourceLine" id="cb327-8" data-line-number="8"><span class="co">## GRanges object with 3 ranges and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb327-9" data-line-number="9"><span class="co">##       seqnames    ranges strand</span></a>
<a class="sourceLine" id="cb327-10" data-line-number="10"><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></a>
<a class="sourceLine" id="cb327-11" data-line-number="11"><span class="co">##   [1]     chr1       1-3      *</span></a>
<a class="sourceLine" id="cb327-12" data-line-number="12"><span class="co">##   [2]     chr1       3-6      *</span></a>
<a class="sourceLine" id="cb327-13" data-line-number="13"><span class="co">##   [3]     chr2       4-6      *</span></a>
<a class="sourceLine" id="cb327-14" data-line-number="14"><span class="co">##   -------</span></a>
<a class="sourceLine" id="cb327-15" data-line-number="15"><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></a></code></pre></div></li>
<li><p>Many fun operations, e.g., extract sequences of regions of interest</p>
<div class="sourceCode" id="cb328"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb328-1" data-line-number="1"><span class="kw">getSeq</span>(dna, roi)</a>
<a class="sourceLine" id="cb328-2" data-line-number="2"><span class="co">##   A DNAStringSet instance of length 3</span></a>
<a class="sourceLine" id="cb328-3" data-line-number="3"><span class="co">##     width seq</span></a>
<a class="sourceLine" id="cb328-4" data-line-number="4"><span class="co">## [1]     3 CTG</span></a>
<a class="sourceLine" id="cb328-5" data-line-number="5"><span class="co">## [2]     4 GACT</span></a>
<a class="sourceLine" id="cb328-6" data-line-number="6"><span class="co">## [3]     3 GGA</span></a></code></pre></div></li>
</ul>
<p>Sequences and annotations</p>
<ul>
<li><p>Input FASTA files from New York State samples (it makes little biological sense to use just New York samples, but we’ll do it anyway!) to <code>DNAStringSet</code></p></li>
<li><p>Work the Genbank accession of the reference sequence</p>
<ul>
<li>DNA sequence</li>
<li>Coding sequence annotations</li>
<li>Coding sequences</li>
</ul></li>
</ul>
<p>Alignment and visualization</p>
<ul>
<li>Multiple alignment using <a href="https://bioconductor.org/packages/DECIPHER">DECIPHER</a></li>
<li>Visualization using <a href="https://bioconductor.org/packages/ggtree">ggtree</a></li>
</ul>
<p>Gene expression</p>
<ul>
<li>Retrieving example data from ExperimentHub</li>
<li>Annotating cell types</li>
</ul>
</div>
</div>
<div id="day-30-dna-sequences-and-annotations" class="section level2">
<h2><span class="header-section-number">5.2</span> Day 30 DNA sequences and annotations</h2>
<div id="setup-1" class="section level3">
<h3><span class="header-section-number">5.2.1</span> Setup</h3>
<p>Set up a directory for working</p>
<div class="sourceCode" id="cb329"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb329-1" data-line-number="1">workdir &lt;-<span class="st"> &quot;workdir&quot;</span></a>
<a class="sourceLine" id="cb329-2" data-line-number="2"><span class="cf">if</span> (<span class="op">!</span><span class="kw">dir.exists</span>(workdir))</a>
<a class="sourceLine" id="cb329-3" data-line-number="3">    <span class="kw">dir.create</span>(workdir)</a></code></pre></div>
<p>Make sure the following packages are installed</p>
<div class="sourceCode" id="cb330"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb330-1" data-line-number="1"><span class="kw">library</span>(Biostrings)</a>
<a class="sourceLine" id="cb330-2" data-line-number="2"><span class="kw">library</span>(GenomicRanges)</a>
<a class="sourceLine" id="cb330-3" data-line-number="3"><span class="kw">library</span>(rentrez)</a>
<a class="sourceLine" id="cb330-4" data-line-number="4"><span class="kw">library</span>(genbankr)</a></code></pre></div>
<ul>
<li><p>Install necessary packages, if needed</p>
<div class="sourceCode" id="cb331"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb331-1" data-line-number="1"><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>))</a>
<a class="sourceLine" id="cb331-2" data-line-number="2">    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">repos =</span> <span class="st">&quot;https://cran.r-project.org&quot;</span>)</a>
<a class="sourceLine" id="cb331-3" data-line-number="3"></a>
<a class="sourceLine" id="cb331-4" data-line-number="4"><span class="co">## for example...</span></a>
<a class="sourceLine" id="cb331-5" data-line-number="5">pkgs &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Biostrings&quot;</span>, <span class="st">&quot;GenomicRanges&quot;</span>, <span class="st">&quot;rentrez&quot;</span>, <span class="st">&quot;genbankr&quot;</span>)</a>
<a class="sourceLine" id="cb331-6" data-line-number="6">BiocManager<span class="op">::</span><span class="kw">install</span>(pkgs)</a></code></pre></div></li>
</ul>
</div>
<div id="representing-dna-sequences" class="section level3">
<h3><span class="header-section-number">5.2.2</span> Representing DNA sequences</h3>
<p>Sample information</p>
<ul>
<li><p>Visit the <a href="https://www.ncbi.nlm.nih.gov/projects/genome/sars-cov-2-seqs">NCBI SARS-CoV-2</a> site for orientation. We’ll retrieve and analyse some of the samples sequenced in New York state</p></li>
<li><p>Read a csv file summarizing records available from the US NCBI.</p>
<div class="sourceCode" id="cb332"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb332-1" data-line-number="1">url &lt;-<span class="st"> &quot;https://raw.githubusercontent.com/mtmorgan/QuaRantine/master/assets/05-genbank-accessions.csv&quot;</span></a>
<a class="sourceLine" id="cb332-2" data-line-number="2">csv_file &lt;-<span class="st"> </span><span class="kw">file.path</span>(workdir, <span class="kw">basename</span>(url))</a>
<a class="sourceLine" id="cb332-3" data-line-number="3"><span class="cf">if</span> ( <span class="op">!</span><span class="kw">file.exists</span>(csv_file) )</a>
<a class="sourceLine" id="cb332-4" data-line-number="4">    <span class="kw">download.file</span>(url, csv_file)</a>
<a class="sourceLine" id="cb332-5" data-line-number="5"></a>
<a class="sourceLine" id="cb332-6" data-line-number="6">records &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw">read_csv</span>(csv_file)</a>
<a class="sourceLine" id="cb332-7" data-line-number="7"><span class="co">## Parsed with column specification:</span></a>
<a class="sourceLine" id="cb332-8" data-line-number="8"><span class="co">## cols(</span></a>
<a class="sourceLine" id="cb332-9" data-line-number="9"><span class="co">##   accession = col_character(),</span></a>
<a class="sourceLine" id="cb332-10" data-line-number="10"><span class="co">##   link = col_character(),</span></a>
<a class="sourceLine" id="cb332-11" data-line-number="11"><span class="co">##   date = col_date(format = &quot;&quot;),</span></a>
<a class="sourceLine" id="cb332-12" data-line-number="12"><span class="co">##   country = col_character(),</span></a>
<a class="sourceLine" id="cb332-13" data-line-number="13"><span class="co">##   region = col_character()</span></a>
<a class="sourceLine" id="cb332-14" data-line-number="14"><span class="co">## )</span></a>
<a class="sourceLine" id="cb332-15" data-line-number="15">records</a>
<a class="sourceLine" id="cb332-16" data-line-number="16"><span class="co">## # A tibble: 2,434 x 5</span></a>
<a class="sourceLine" id="cb332-17" data-line-number="17"><span class="co">##    accession link                                    date       country   region</span></a>
<a class="sourceLine" id="cb332-18" data-line-number="18"><span class="co">##    &lt;chr&gt;     &lt;chr&gt;                                   &lt;date&gt;     &lt;chr&gt;     &lt;chr&gt; </span></a>
<a class="sourceLine" id="cb332-19" data-line-number="19"><span class="co">##  1 NC_045512 https://www.ncbi.nlm.nih.gov/nuccore/N… NA         China     &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb332-20" data-line-number="20"><span class="co">##  2 MT447189  https://www.ncbi.nlm.nih.gov/nuccore/M… 2020-03-02 Uzbekist… &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb332-21" data-line-number="21"><span class="co">##  3 MT447188  https://www.ncbi.nlm.nih.gov/nuccore/M… 2020-03-02 Uzbekist… &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb332-22" data-line-number="22"><span class="co">##  4 MT447177  https://www.ncbi.nlm.nih.gov/nuccore/M… 2020-03-26 Iran      &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb332-23" data-line-number="23"><span class="co">##  5 MT447176  https://www.ncbi.nlm.nih.gov/nuccore/M… 2020-03-20 Thailand  &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb332-24" data-line-number="24"><span class="co">##  6 MT447175  https://www.ncbi.nlm.nih.gov/nuccore/M… 2020-03-19 Thailand  &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb332-25" data-line-number="25"><span class="co">##  7 MT447174  https://www.ncbi.nlm.nih.gov/nuccore/M… 2020-03-01 Thailand  &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb332-26" data-line-number="26"><span class="co">##  8 MT447173  https://www.ncbi.nlm.nih.gov/nuccore/M… 2020-03-18 Thailand  &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb332-27" data-line-number="27"><span class="co">##  9 MT447172  https://www.ncbi.nlm.nih.gov/nuccore/M… 2020-03-18 Thailand  &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb332-28" data-line-number="28"><span class="co">## 10 MT447171  https://www.ncbi.nlm.nih.gov/nuccore/M… 2020-03-16 Thailand  &lt;NA&gt;  </span></a>
<a class="sourceLine" id="cb332-29" data-line-number="29"><span class="co">## # … with 2,424 more rows</span></a></code></pre></div></li>
</ul>
<p>FASTA sequence files</p>
<ul>
<li><p>Download a FASTA file containing DNA sequences of all samples sequenced in New York State (this doesn’t make much sense biologically!)</p>
<div class="sourceCode" id="cb333"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb333-1" data-line-number="1">url &lt;-<span class="st"> &quot;https://raw.githubusercontent.com/mtmorgan/QuaRantine/master/assets/05-SARS-CoV2-NY.fasta&quot;</span></a>
<a class="sourceLine" id="cb333-2" data-line-number="2">fasta_file &lt;-<span class="st"> </span><span class="kw">file.path</span>(workdir, <span class="kw">basename</span>(url))</a>
<a class="sourceLine" id="cb333-3" data-line-number="3"><span class="cf">if</span> ( <span class="op">!</span><span class="kw">file.exists</span>(fasta_file) )</a>
<a class="sourceLine" id="cb333-4" data-line-number="4">    <span class="kw">download.file</span>(url, fasta_file)</a></code></pre></div></li>
<li><p>This is a plain text file with a simple format. Each DNA sequence is on a line starting with an identifier, <code>&gt;MT...</code>, and then lines representing the DNA (or RNA or amino acid) sequence.</p>
<div class="sourceCode" id="cb334"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb334-1" data-line-number="1"><span class="kw">readLines</span>(fasta_file, <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb334-2" data-line-number="2"><span class="co">## [1] &quot;&gt;MT434817&quot;                                                                       </span></a>
<a class="sourceLine" id="cb334-3" data-line-number="3"><span class="co">## [2] &quot;ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAA&quot;</span></a>
<a class="sourceLine" id="cb334-4" data-line-number="4"><span class="co">## [3] &quot;AATCTGTGTGGCTGTCACTCGGCTGCATGCTTAGTGCACTCACGCAGTATAATTAATAACTAATTACTGTCGTTGACAGG&quot;</span></a>
<a class="sourceLine" id="cb334-5" data-line-number="5"><span class="co">## [4] &quot;ACACGAGTAACTCGTCTATCTTCTGCAGGCTGCTTACGGTTTCGTCCGTGTTGCAGCCGATCATCAGCACATCTAGGTTT&quot;</span></a>
<a class="sourceLine" id="cb334-6" data-line-number="6"><span class="co">## [5] &quot;TGTCCGGGTGTGACCGAAAGGTAAGATGGAGAGCCTTGTCCCTGGTTTCAACGAGAAAACACACGTCCAACTCAGTTTGC&quot;</span></a></code></pre></div></li>
<li><p>Load the <a href="https://bioconductor.org/packages/Biostrings">Biostrings</a> package</p>
<div class="sourceCode" id="cb335"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb335-1" data-line-number="1"><span class="kw">library</span>(Biostrings)</a>
<a class="sourceLine" id="cb335-2" data-line-number="2"><span class="kw">options</span>(<span class="dt">Biostrings.coloring =</span> <span class="ot">FALSE</span>)</a></code></pre></div></li>
<li><p>Read the DNA sequences and explore basic properties</p>
<div class="sourceCode" id="cb336"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb336-1" data-line-number="1">dna &lt;-<span class="st"> </span><span class="kw">readDNAStringSet</span>(fasta_file)</a>
<a class="sourceLine" id="cb336-2" data-line-number="2">dna</a>
<a class="sourceLine" id="cb336-3" data-line-number="3"><span class="co">##   A DNAStringSet instance of length 245</span></a>
<a class="sourceLine" id="cb336-4" data-line-number="4"><span class="co">##       width seq                                             names               </span></a>
<a class="sourceLine" id="cb336-5" data-line-number="5"><span class="co">##   [1] 29873 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434817</span></a>
<a class="sourceLine" id="cb336-6" data-line-number="6"><span class="co">##   [2] 29879 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434816</span></a>
<a class="sourceLine" id="cb336-7" data-line-number="7"><span class="co">##   [3] 29873 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434815</span></a>
<a class="sourceLine" id="cb336-8" data-line-number="8"><span class="co">##   [4] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434814</span></a>
<a class="sourceLine" id="cb336-9" data-line-number="9"><span class="co">##   [5] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT434813</span></a>
<a class="sourceLine" id="cb336-10" data-line-number="10"><span class="co">##   ...   ... ...</span></a>
<a class="sourceLine" id="cb336-11" data-line-number="11"><span class="co">## [241] 29724 GATCTGTTCTCTAAACGAACTT...TGTACAGTGAACAATGCTAGGG MT370833</span></a>
<a class="sourceLine" id="cb336-12" data-line-number="12"><span class="co">## [242] 29717 TCTAAACGAACTTTAAAATCTG...TACAGTGAACAATGCTAGGGAG MT370832</span></a>
<a class="sourceLine" id="cb336-13" data-line-number="13"><span class="co">## [243] 29716 TCTAAACGAACTTTAAAATCTG...GTACAGTGAACAATGCTAGGGA MT370831</span></a>
<a class="sourceLine" id="cb336-14" data-line-number="14"><span class="co">## [244] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT325627</span></a>
<a class="sourceLine" id="cb336-15" data-line-number="15"><span class="co">## [245] 29882 ATTAAAGGTTTATACCTTCCCA...GGAGAATGACAAAAAAAAAAAA MT304486</span></a></code></pre></div></li>
<li><p>Perform simple manipulations</p>
<div class="sourceCode" id="cb337"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb337-1" data-line-number="1"><span class="co">## how many records?</span></a>
<a class="sourceLine" id="cb337-2" data-line-number="2"><span class="kw">length</span>(dna)</a>
<a class="sourceLine" id="cb337-3" data-line-number="3"><span class="co">## [1] 245</span></a>
<a class="sourceLine" id="cb337-4" data-line-number="4"></a>
<a class="sourceLine" id="cb337-5" data-line-number="5"><span class="co">## subset</span></a>
<a class="sourceLine" id="cb337-6" data-line-number="6">dna[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a>
<a class="sourceLine" id="cb337-7" data-line-number="7"><span class="co">##   A DNAStringSet instance of length 3</span></a>
<a class="sourceLine" id="cb337-8" data-line-number="8"><span class="co">##     width seq                                               names               </span></a>
<a class="sourceLine" id="cb337-9" data-line-number="9"><span class="co">## [1] 29873 ATTAAAGGTTTATACCTTCCCAG...AGGAGAATGACAAAAAAAAAAAA MT434817</span></a>
<a class="sourceLine" id="cb337-10" data-line-number="10"><span class="co">## [2] 29879 ATTAAAGGTTTATACCTTCCCAG...AGGAGAATGACAAAAAAAAAAAA MT434816</span></a>
<a class="sourceLine" id="cb337-11" data-line-number="11"><span class="co">## [3] 29873 ATTAAAGGTTTATACCTTCCCAG...AGGAGAATGACAAAAAAAAAAAA MT434815</span></a>
<a class="sourceLine" id="cb337-12" data-line-number="12">dna[<span class="kw">c</span>(<span class="st">&#39;MT434817&#39;</span>, <span class="st">&#39;MT434815&#39;</span>)]</a>
<a class="sourceLine" id="cb337-13" data-line-number="13"><span class="co">##   A DNAStringSet instance of length 2</span></a>
<a class="sourceLine" id="cb337-14" data-line-number="14"><span class="co">##     width seq                                               names               </span></a>
<a class="sourceLine" id="cb337-15" data-line-number="15"><span class="co">## [1] 29873 ATTAAAGGTTTATACCTTCCCAG...AGGAGAATGACAAAAAAAAAAAA MT434817</span></a>
<a class="sourceLine" id="cb337-16" data-line-number="16"><span class="co">## [2] 29873 ATTAAAGGTTTATACCTTCCCAG...AGGAGAATGACAAAAAAAAAAAA MT434815</span></a>
<a class="sourceLine" id="cb337-17" data-line-number="17"></a>
<a class="sourceLine" id="cb337-18" data-line-number="18"><span class="co">## extract</span></a>
<a class="sourceLine" id="cb337-19" data-line-number="19">dna[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb337-20" data-line-number="20"><span class="co">##   29873-letter &quot;DNAString&quot; instance</span></a>
<a class="sourceLine" id="cb337-21" data-line-number="21"><span class="co">## seq: ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAAC...TTAATAGCTTCTTAGGAGAATGACAAAAAAAAAAAA</span></a>
<a class="sourceLine" id="cb337-22" data-line-number="22"></a>
<a class="sourceLine" id="cb337-23" data-line-number="23"><span class="co">## number of characters in each record</span></a>
<a class="sourceLine" id="cb337-24" data-line-number="24"><span class="kw">head</span>(<span class="kw">nchar</span>(dna))</a>
<a class="sourceLine" id="cb337-25" data-line-number="25"><span class="co">## [1] 29873 29879 29873 29882 29882 29882</span></a>
<a class="sourceLine" id="cb337-26" data-line-number="26"><span class="kw">hist</span>(<span class="kw">nchar</span>(dna)) <span class="co"># histogram of sequence lengths</span></a></code></pre></div>
<p><img src="05-Bioinformatics_files/figure-html/unnamed-chunk-25-1.png" width="672" /></p></li>
</ul>
</div>
<div id="working-with-genbankrecord-objects" class="section level3">
<h3><span class="header-section-number">5.2.3</span> Working with <code>GenBankRecord</code> objects</h3>
<p>Data is often represented in complicated formats that do not fit into the tidy ‘table’ format of a CSV file. An example is the information displayed on the <a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_045512">SARS-CoV-2 reference genome</a> web page.</p>
<p>Often ‘someone’ has written software to manipulate this data in a more convient way.</p>
<ul>
<li><p>Attach the <a href="https://cran.r-project.org/package=rentrez">rentrez</a> package, which provides a way to query the NCBI’s ‘Entrez’ collection of data bases. Use <code>BiocManager::install('rentrez')</code> if you need to install the package</p>
<div class="sourceCode" id="cb338"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb338-1" data-line-number="1"><span class="kw">library</span>(rentrez)</a>
<a class="sourceLine" id="cb338-2" data-line-number="2"></a>
<a class="sourceLine" id="cb338-3" data-line-number="3"><span class="co">## who&#39;s responsible?</span></a>
<a class="sourceLine" id="cb338-4" data-line-number="4"><span class="kw">maintainer</span>(<span class="st">&quot;rentrez&quot;</span>)</a>
<a class="sourceLine" id="cb338-5" data-line-number="5"><span class="co">## [1] &quot;David Winter &lt;david.winter@gmail.com&gt;&quot;</span></a>
<a class="sourceLine" id="cb338-6" data-line-number="6"></a>
<a class="sourceLine" id="cb338-7" data-line-number="7"><span class="co">## how to cite?</span></a>
<a class="sourceLine" id="cb338-8" data-line-number="8"><span class="kw">citation</span>(<span class="st">&quot;rentrez&quot;</span>)</a>
<a class="sourceLine" id="cb338-9" data-line-number="9"><span class="co">## </span></a>
<a class="sourceLine" id="cb338-10" data-line-number="10"><span class="co">## To cite rentrez in publications use:</span></a>
<a class="sourceLine" id="cb338-11" data-line-number="11"><span class="co">## </span></a>
<a class="sourceLine" id="cb338-12" data-line-number="12"><span class="co">##   Winter, D. J. (2017) rentrez: an R package for the NCBI eUtils API</span></a>
<a class="sourceLine" id="cb338-13" data-line-number="13"><span class="co">##   The R Journal 9(2):520-526</span></a>
<a class="sourceLine" id="cb338-14" data-line-number="14"><span class="co">## </span></a>
<a class="sourceLine" id="cb338-15" data-line-number="15"><span class="co">## A BibTeX entry for LaTeX users is</span></a>
<a class="sourceLine" id="cb338-16" data-line-number="16"><span class="co">## </span></a>
<a class="sourceLine" id="cb338-17" data-line-number="17"><span class="co">##   @Article{,</span></a>
<a class="sourceLine" id="cb338-18" data-line-number="18"><span class="co">##     title = {{rentrez}: an R package for the NCBI eUtils API},</span></a>
<a class="sourceLine" id="cb338-19" data-line-number="19"><span class="co">##     author = {David J. Winter},</span></a>
<a class="sourceLine" id="cb338-20" data-line-number="20"><span class="co">##     journal = {The R Journal},</span></a>
<a class="sourceLine" id="cb338-21" data-line-number="21"><span class="co">##     year = {2017},</span></a>
<a class="sourceLine" id="cb338-22" data-line-number="22"><span class="co">##     volume = {9},</span></a>
<a class="sourceLine" id="cb338-23" data-line-number="23"><span class="co">##     issue = {2},</span></a>
<a class="sourceLine" id="cb338-24" data-line-number="24"><span class="co">##     pages = {520--526},</span></a>
<a class="sourceLine" id="cb338-25" data-line-number="25"><span class="co">##   }</span></a></code></pre></div></li>
<li><p>The accession number for the SARS-CoV-2 reference sequence is <code>NC_045512</code>. Use <code>entrez_search()</code> to search the NCBI ‘nuccore’ database to discover information about this accession.</p>
<div class="sourceCode" id="cb339"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb339-1" data-line-number="1">accession &lt;-<span class="st"> &quot;NC_045512&quot;</span></a>
<a class="sourceLine" id="cb339-2" data-line-number="2">nuccore_record &lt;-<span class="st"> </span><span class="kw">entrez_search</span>(<span class="st">&quot;nuccore&quot;</span>, accession)</a>
<a class="sourceLine" id="cb339-3" data-line-number="3">id &lt;-<span class="st"> </span>nuccore_record<span class="op">$</span>ids</a>
<a class="sourceLine" id="cb339-4" data-line-number="4">id</a>
<a class="sourceLine" id="cb339-5" data-line-number="5"><span class="co">## [1] &quot;1798174254&quot;</span></a></code></pre></div></li>
<li><p>The <code>id</code> is an identifier that can be used to actually retrieve the genbank record</p>
<div class="sourceCode" id="cb340"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb340-1" data-line-number="1">gb_record &lt;-</a>
<a class="sourceLine" id="cb340-2" data-line-number="2"><span class="st">    </span><span class="kw">entrez_fetch</span>(<span class="st">&quot;nuccore&quot;</span>, id, <span class="dt">rettype =</span> <span class="st">&quot;gbwithparts&quot;</span>, <span class="dt">retmode =</span> <span class="st">&quot;text&quot;</span>)</a></code></pre></div></li>
<li><p>The record contains all the text in the record; if you like visualize it with</p>
<div class="sourceCode" id="cb341"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb341-1" data-line-number="1"><span class="kw">cat</span>(gb_record)</a></code></pre></div></li>
</ul>
<p>Represent the data as <code>GenBankR</code> object</p>
<ul>
<li><p>Rather than try to ‘munge’ (e.g., copy and paste) relevant information from the GenBank record into R, use the <a href="https://bioconductor.org/packages/genbankr">genbankr</a> package. Again, install with <code>BiocManager::install(&quot;genbankr&quot;)</code> if necessary</p>
<div class="sourceCode" id="cb342"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb342-1" data-line-number="1"><span class="kw">library</span>(genbankr)</a>
<a class="sourceLine" id="cb342-2" data-line-number="2"></a>
<a class="sourceLine" id="cb342-3" data-line-number="3"><span class="kw">maintainer</span>(<span class="st">&quot;genbankr&quot;</span>)</a>
<a class="sourceLine" id="cb342-4" data-line-number="4"><span class="co">## [1] &quot;Gabriel Becker &lt;becker.gabriel@gene.com&gt;&quot;</span></a>
<a class="sourceLine" id="cb342-5" data-line-number="5"></a>
<a class="sourceLine" id="cb342-6" data-line-number="6"><span class="kw">citation</span>(<span class="st">&quot;genbankr&quot;</span>)</a>
<a class="sourceLine" id="cb342-7" data-line-number="7"><span class="co">## </span></a>
<a class="sourceLine" id="cb342-8" data-line-number="8"><span class="co">## To cite package &#39;genbankr&#39; in publications use:</span></a>
<a class="sourceLine" id="cb342-9" data-line-number="9"><span class="co">## </span></a>
<a class="sourceLine" id="cb342-10" data-line-number="10"><span class="co">##   Gabriel Becker and Michael Lawrence (2019). genbankr: Parsing GenBank</span></a>
<a class="sourceLine" id="cb342-11" data-line-number="11"><span class="co">##   files into semantically useful objects. R package version 1.14.0.</span></a>
<a class="sourceLine" id="cb342-12" data-line-number="12"><span class="co">## </span></a>
<a class="sourceLine" id="cb342-13" data-line-number="13"><span class="co">## A BibTeX entry for LaTeX users is</span></a>
<a class="sourceLine" id="cb342-14" data-line-number="14"><span class="co">## </span></a>
<a class="sourceLine" id="cb342-15" data-line-number="15"><span class="co">##   @Manual{,</span></a>
<a class="sourceLine" id="cb342-16" data-line-number="16"><span class="co">##     title = {genbankr: Parsing GenBank files into semantically useful objects},</span></a>
<a class="sourceLine" id="cb342-17" data-line-number="17"><span class="co">##     author = {Gabriel Becker and Michael Lawrence},</span></a>
<a class="sourceLine" id="cb342-18" data-line-number="18"><span class="co">##     year = {2019},</span></a>
<a class="sourceLine" id="cb342-19" data-line-number="19"><span class="co">##     note = {R package version 1.14.0},</span></a>
<a class="sourceLine" id="cb342-20" data-line-number="20"><span class="co">##   }</span></a></code></pre></div></li>
<li><p>Use the function <code>readGenBank()</code> to parse the complicated text of <code>gb_record</code></p>
<div class="sourceCode" id="cb343"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb343-1" data-line-number="1">gb &lt;-<span class="st"> </span><span class="kw">readGenBank</span>(<span class="dt">text =</span> gb_record)</a>
<a class="sourceLine" id="cb343-2" data-line-number="2">gb</a>
<a class="sourceLine" id="cb343-3" data-line-number="3"><span class="co">## GenBank Annotations</span></a>
<a class="sourceLine" id="cb343-4" data-line-number="4"><span class="co">## Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome. </span></a>
<a class="sourceLine" id="cb343-5" data-line-number="5"><span class="co">## Accession: NC_045512</span></a>
<a class="sourceLine" id="cb343-6" data-line-number="6"><span class="co">## 1 Sequence(s) with total length length: 29903</span></a>
<a class="sourceLine" id="cb343-7" data-line-number="7"><span class="co">## 11 genes</span></a>
<a class="sourceLine" id="cb343-8" data-line-number="8"><span class="co">## 12 transcripts</span></a>
<a class="sourceLine" id="cb343-9" data-line-number="9"><span class="co">## 13 exons/cds elements</span></a>
<a class="sourceLine" id="cb343-10" data-line-number="10"><span class="co">## 0 variations</span></a>
<a class="sourceLine" id="cb343-11" data-line-number="11"><span class="co">## 34 other features</span></a></code></pre></div></li>
<li><p>Extract the reference sequence from the object</p>
<div class="sourceCode" id="cb344"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb344-1" data-line-number="1">ref &lt;-<span class="st"> </span><span class="kw">getSeq</span>(gb)</a>
<a class="sourceLine" id="cb344-2" data-line-number="2">ref</a>
<a class="sourceLine" id="cb344-3" data-line-number="3"><span class="co">##   A DNAStringSet instance of length 1</span></a>
<a class="sourceLine" id="cb344-4" data-line-number="4"><span class="co">##     width seq                                               names               </span></a>
<a class="sourceLine" id="cb344-5" data-line-number="5"><span class="co">## [1] 29903 ATTAAAGGTTTATACCTTCCCAG...AAAAAAAAAAAAAAAAAAAAAAA Severe acute resp...</span></a></code></pre></div></li>
<li><p>Extract the coding sequences from the reference sequence; the <code>cds()</code> command is explained a little later.</p>
<div class="sourceCode" id="cb345"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb345-1" data-line-number="1">cds &lt;-<span class="st"> </span><span class="kw">getSeq</span>(ref, <span class="kw">cds</span>(gb))</a>
<a class="sourceLine" id="cb345-2" data-line-number="2">cds</a>
<a class="sourceLine" id="cb345-3" data-line-number="3"><span class="co">##   A DNAStringSet instance of length 13</span></a>
<a class="sourceLine" id="cb345-4" data-line-number="4"><span class="co">##      width seq</span></a>
<a class="sourceLine" id="cb345-5" data-line-number="5"><span class="co">##  [1] 13203 ATGGAGAGCCTTGTCCCTGGTTTCAACGAGAAA...CTTCAGTCAGCTGATGCACAATCGTTTTTAAAC</span></a>
<a class="sourceLine" id="cb345-6" data-line-number="6"><span class="co">##  [2]  8088 CGGGTTTGCGGTGTAAGTGCAGCCCGTCTTACA...GTTATTTCTAGTGATGTTCTTGTTAACAACTAA</span></a>
<a class="sourceLine" id="cb345-7" data-line-number="7"><span class="co">##  [3] 13218 ATGGAGAGCCTTGTCCCTGGTTTCAACGAGAAA...GCACAATCGTTTTTAAACGGGTTTGCGGTGTAA</span></a>
<a class="sourceLine" id="cb345-8" data-line-number="8"><span class="co">##  [4]  3822 ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTC...GTGCTCAAAGGAGTCAAATTACATTACACATAA</span></a>
<a class="sourceLine" id="cb345-9" data-line-number="9"><span class="co">##  [5]   828 ATGGATTTGTTTATGAGAATCTTCACAATTGGA...GAACCGACGACGACTACTAGCGTGCCTTTGTAA</span></a>
<a class="sourceLine" id="cb345-10" data-line-number="10"><span class="co">##  ...   ... ...</span></a>
<a class="sourceLine" id="cb345-11" data-line-number="11"><span class="co">##  [9]   366 ATGAAAATTATTCTTTTCTTGGCACTGATAACA...CTTTGCTTCACACTCAAAAGAAAGACAGAATGA</span></a>
<a class="sourceLine" id="cb345-12" data-line-number="12"><span class="co">## [10]   132 ATGATTGAACTTTCATTAATTGACTTCTATTTG...CTGCAAGATCATAATGAAACTTGTCACGCCTAA</span></a>
<a class="sourceLine" id="cb345-13" data-line-number="13"><span class="co">## [11]   366 ATGAAATTTCTTGTTTTCTTAGGAATCATCACA...CATGACGTTCGTGTTGTTTTAGATTTCATCTAA</span></a>
<a class="sourceLine" id="cb345-14" data-line-number="14"><span class="co">## [12]  1260 ATGTCTGATAATGGACCCCAAAATCAGCGAAAT...TCCATGAGCAGTGCTGACTCAACTCAGGCCTAA</span></a>
<a class="sourceLine" id="cb345-15" data-line-number="15"><span class="co">## [13]   117 ATGGGCTATATAAACGTTTTCGCTTTTCCGTTT...CAAGTAGATGTAGTTAACTTTAATCTCACATAG</span></a></code></pre></div></li>
<li><p>There are 13 coding sequences in the genome. Translate these to their amino acid sequences</p>
<div class="sourceCode" id="cb346"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb346-1" data-line-number="1"><span class="kw">translate</span>(cds)</a>
<a class="sourceLine" id="cb346-2" data-line-number="2"><span class="co">##   A AAStringSet instance of length 13</span></a>
<a class="sourceLine" id="cb346-3" data-line-number="3"><span class="co">##      width seq</span></a>
<a class="sourceLine" id="cb346-4" data-line-number="4"><span class="co">##  [1]  4401 MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGD...VCTVCGMWKGYGCSCDQLREPMLQSADAQSFLN</span></a>
<a class="sourceLine" id="cb346-5" data-line-number="5"><span class="co">##  [2]  2696 RVCGVSAARLTPCGTGTSTDVVYRAFDIYNDKV...INDMILSLLSKGRLIIRENNRVVISSDVLVNN*</span></a>
<a class="sourceLine" id="cb346-6" data-line-number="6"><span class="co">##  [3]  4406 MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGD...GMWKGYGCSCDQLREPMLQSADAQSFLNGFAV*</span></a>
<a class="sourceLine" id="cb346-7" data-line-number="7"><span class="co">##  [4]  1274 MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFT...SCLKGCCSCGSCCKFDEDDSEPVLKGVKLHYT*</span></a>
<a class="sourceLine" id="cb346-8" data-line-number="8"><span class="co">##  [5]   276 MDLFMRIFTIGTVTLKQGEIKDATPSDFVRATA...VQIHTIDGSSGVVNPVMEPIYDEPTTTTSVPL*</span></a>
<a class="sourceLine" id="cb346-9" data-line-number="9"><span class="co">##  ...   ... ...</span></a>
<a class="sourceLine" id="cb346-10" data-line-number="10"><span class="co">##  [9]   122 MKIILFLALITLATCELYHYQECVRGTTVLLKE...QEEVQELYSPIFLIVAAIVFITLCFTLKRKTE*</span></a>
<a class="sourceLine" id="cb346-11" data-line-number="11"><span class="co">## [10]    44 MIELSLIDFYLCFLAFLLFLVLIMLIIFWFSLELQDHNETCHA*</span></a>
<a class="sourceLine" id="cb346-12" data-line-number="12"><span class="co">## [11]   122 MKFLVFLGIITTVAAFHQECSLQSCTQHQPYVV...CQEPKLGSLVVRCSFYEDFLEYHDVRVVLDFI*</span></a>
<a class="sourceLine" id="cb346-13" data-line-number="13"><span class="co">## [12]   420 MSDNGPQNQRNAPRITFGGPSDSTGSNQNGERS...KQQTVTLLPAADLDDFSKQLQQSMSSADSTQA*</span></a>
<a class="sourceLine" id="cb346-14" data-line-number="14"><span class="co">## [13]    39 MGYINVFAFPFTIYSLLLCRMNSRNYIAQVDVVNFNLT*</span></a></code></pre></div></li>
</ul>
</div>
<div id="genomic-ranges-and-dna-sequences" class="section level3">
<h3><span class="header-section-number">5.2.4</span> Genomic ranges and DNA sequences</h3>
<p>Genomic ranges</p>
<ul>
<li><p>Attach the <a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a> package. If necessary, install with <code>BiocManager::install(&quot;GenomicRanges&quot;)</code></p>
<div class="sourceCode" id="cb347"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb347-1" data-line-number="1"><span class="kw">library</span>(GenomicRanges)</a></code></pre></div></li>
<li><p>Extract the coordinates of the coding sequences contained in the GenBank accession</p>
<div class="sourceCode" id="cb348"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb348-1" data-line-number="1">cds &lt;-<span class="st"> </span><span class="kw">cds</span>(gb)</a></code></pre></div></li>
<li><p>Look at the genomic ranges defined in <code>cds</code></p>
<div class="sourceCode" id="cb349"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb349-1" data-line-number="1"><span class="kw">granges</span>(cds)</a>
<a class="sourceLine" id="cb349-2" data-line-number="2"><span class="co">## GRanges object with 13 ranges and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb349-3" data-line-number="3"><span class="co">##                                              seqnames      ranges strand</span></a>
<a class="sourceLine" id="cb349-4" data-line-number="4"><span class="co">##                                                 &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;</span></a>
<a class="sourceLine" id="cb349-5" data-line-number="5"><span class="co">##    [1] Severe acute respiratory syndrome coronavirus2   266-13468      +</span></a>
<a class="sourceLine" id="cb349-6" data-line-number="6"><span class="co">##    [2] Severe acute respiratory syndrome coronavirus2 13468-21555      +</span></a>
<a class="sourceLine" id="cb349-7" data-line-number="7"><span class="co">##    [3] Severe acute respiratory syndrome coronavirus2   266-13483      +</span></a>
<a class="sourceLine" id="cb349-8" data-line-number="8"><span class="co">##    [4] Severe acute respiratory syndrome coronavirus2 21563-25384      +</span></a>
<a class="sourceLine" id="cb349-9" data-line-number="9"><span class="co">##    [5] Severe acute respiratory syndrome coronavirus2 25393-26220      +</span></a>
<a class="sourceLine" id="cb349-10" data-line-number="10"><span class="co">##    ...                                            ...         ...    ...</span></a>
<a class="sourceLine" id="cb349-11" data-line-number="11"><span class="co">##    [9] Severe acute respiratory syndrome coronavirus2 27394-27759      +</span></a>
<a class="sourceLine" id="cb349-12" data-line-number="12"><span class="co">##   [10] Severe acute respiratory syndrome coronavirus2 27756-27887      +</span></a>
<a class="sourceLine" id="cb349-13" data-line-number="13"><span class="co">##   [11] Severe acute respiratory syndrome coronavirus2 27894-28259      +</span></a>
<a class="sourceLine" id="cb349-14" data-line-number="14"><span class="co">##   [12] Severe acute respiratory syndrome coronavirus2 28274-29533      +</span></a>
<a class="sourceLine" id="cb349-15" data-line-number="15"><span class="co">##   [13] Severe acute respiratory syndrome coronavirus2 29558-29674      +</span></a>
<a class="sourceLine" id="cb349-16" data-line-number="16"><span class="co">##   -------</span></a>
<a class="sourceLine" id="cb349-17" data-line-number="17"><span class="co">##   seqinfo: 1 sequence from NC_045512.2 genome</span></a></code></pre></div></li>
<li><p>There are 13 coding sequences in the reference genome. The sequence is named ‘Severe acute respiratory syndrome coronavirus2’. The first region starts at position 266 of the reference genome, and goes to position 13468. The coding sequence is on the ‘+’ strand.</p></li>
<li><p>The full <code>cds</code> object has additional detail about each coding sequence. Display the full object (the display ‘wraps’ around several lines</p>
<div class="sourceCode" id="cb350"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb350-1" data-line-number="1">cds</a>
<a class="sourceLine" id="cb350-2" data-line-number="2"><span class="co">## GRanges object with 13 ranges and 14 metadata columns:</span></a>
<a class="sourceLine" id="cb350-3" data-line-number="3"><span class="co">##                                              seqnames      ranges strand |</span></a>
<a class="sourceLine" id="cb350-4" data-line-number="4"><span class="co">##                                                 &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; |</span></a>
<a class="sourceLine" id="cb350-5" data-line-number="5"><span class="co">##    [1] Severe acute respiratory syndrome coronavirus2   266-13468      + |</span></a>
<a class="sourceLine" id="cb350-6" data-line-number="6"><span class="co">##    [2] Severe acute respiratory syndrome coronavirus2 13468-21555      + |</span></a>
<a class="sourceLine" id="cb350-7" data-line-number="7"><span class="co">##    [3] Severe acute respiratory syndrome coronavirus2   266-13483      + |</span></a>
<a class="sourceLine" id="cb350-8" data-line-number="8"><span class="co">##    [4] Severe acute respiratory syndrome coronavirus2 21563-25384      + |</span></a>
<a class="sourceLine" id="cb350-9" data-line-number="9"><span class="co">##    [5] Severe acute respiratory syndrome coronavirus2 25393-26220      + |</span></a>
<a class="sourceLine" id="cb350-10" data-line-number="10"><span class="co">##    ...                                            ...         ...    ... .</span></a>
<a class="sourceLine" id="cb350-11" data-line-number="11"><span class="co">##    [9] Severe acute respiratory syndrome coronavirus2 27394-27759      + |</span></a>
<a class="sourceLine" id="cb350-12" data-line-number="12"><span class="co">##   [10] Severe acute respiratory syndrome coronavirus2 27756-27887      + |</span></a>
<a class="sourceLine" id="cb350-13" data-line-number="13"><span class="co">##   [11] Severe acute respiratory syndrome coronavirus2 27894-28259      + |</span></a>
<a class="sourceLine" id="cb350-14" data-line-number="14"><span class="co">##   [12] Severe acute respiratory syndrome coronavirus2 28274-29533      + |</span></a>
<a class="sourceLine" id="cb350-15" data-line-number="15"><span class="co">##   [13] Severe acute respiratory syndrome coronavirus2 29558-29674      + |</span></a>
<a class="sourceLine" id="cb350-16" data-line-number="16"><span class="co">##               type        gene   locus_tag ribosomal_slippage</span></a>
<a class="sourceLine" id="cb350-17" data-line-number="17"><span class="co">##        &lt;character&gt; &lt;character&gt; &lt;character&gt;          &lt;logical&gt;</span></a>
<a class="sourceLine" id="cb350-18" data-line-number="18"><span class="co">##    [1]         CDS      ORF1ab  GU280_gp01               TRUE</span></a>
<a class="sourceLine" id="cb350-19" data-line-number="19"><span class="co">##    [2]         CDS      ORF1ab  GU280_gp01               TRUE</span></a>
<a class="sourceLine" id="cb350-20" data-line-number="20"><span class="co">##    [3]         CDS      ORF1ab  GU280_gp01              FALSE</span></a>
<a class="sourceLine" id="cb350-21" data-line-number="21"><span class="co">##    [4]         CDS           S  GU280_gp02              FALSE</span></a>
<a class="sourceLine" id="cb350-22" data-line-number="22"><span class="co">##    [5]         CDS       ORF3a  GU280_gp03              FALSE</span></a>
<a class="sourceLine" id="cb350-23" data-line-number="23"><span class="co">##    ...         ...         ...         ...                ...</span></a>
<a class="sourceLine" id="cb350-24" data-line-number="24"><span class="co">##    [9]         CDS       ORF7a  GU280_gp07              FALSE</span></a>
<a class="sourceLine" id="cb350-25" data-line-number="25"><span class="co">##   [10]         CDS       ORF7b  GU280_gp08              FALSE</span></a>
<a class="sourceLine" id="cb350-26" data-line-number="26"><span class="co">##   [11]         CDS        ORF8  GU280_gp09              FALSE</span></a>
<a class="sourceLine" id="cb350-27" data-line-number="27"><span class="co">##   [12]         CDS           N  GU280_gp10              FALSE</span></a>
<a class="sourceLine" id="cb350-28" data-line-number="28"><span class="co">##   [13]         CDS       ORF10  GU280_gp11              FALSE</span></a>
<a class="sourceLine" id="cb350-29" data-line-number="29"><span class="co">##                                                note codon_start</span></a>
<a class="sourceLine" id="cb350-30" data-line-number="30"><span class="co">##                                         &lt;character&gt;   &lt;numeric&gt;</span></a>
<a class="sourceLine" id="cb350-31" data-line-number="31"><span class="co">##    [1] pp1ab; translated by -1 ribosomal frameshift           1</span></a>
<a class="sourceLine" id="cb350-32" data-line-number="32"><span class="co">##    [2] pp1ab; translated by -1 ribosomal frameshift           1</span></a>
<a class="sourceLine" id="cb350-33" data-line-number="33"><span class="co">##    [3]                                         pp1a           1</span></a>
<a class="sourceLine" id="cb350-34" data-line-number="34"><span class="co">##    [4]            structural protein; spike protein           1</span></a>
<a class="sourceLine" id="cb350-35" data-line-number="35"><span class="co">##    [5]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb350-36" data-line-number="36"><span class="co">##    ...                                          ...         ...</span></a>
<a class="sourceLine" id="cb350-37" data-line-number="37"><span class="co">##    [9]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb350-38" data-line-number="38"><span class="co">##   [10]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb350-39" data-line-number="39"><span class="co">##   [11]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb350-40" data-line-number="40"><span class="co">##   [12]                     ORF9; structural protein           1</span></a>
<a class="sourceLine" id="cb350-41" data-line-number="41"><span class="co">##   [13]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb350-42" data-line-number="42"><span class="co">##                            product     protein_id         db_xref   translation</span></a>
<a class="sourceLine" id="cb350-43" data-line-number="43"><span class="co">##                        &lt;character&gt;    &lt;character&gt; &lt;CharacterList&gt; &lt;AAStringSet&gt;</span></a>
<a class="sourceLine" id="cb350-44" data-line-number="44"><span class="co">##    [1]          ORF1ab polyprotein YP_009724389.1 GeneID:43740578     MES...VNN</span></a>
<a class="sourceLine" id="cb350-45" data-line-number="45"><span class="co">##    [2]          ORF1ab polyprotein YP_009724389.1 GeneID:43740578     MES...VNN</span></a>
<a class="sourceLine" id="cb350-46" data-line-number="46"><span class="co">##    [3]           ORF1a polyprotein YP_009725295.1 GeneID:43740578     MES...FAV</span></a>
<a class="sourceLine" id="cb350-47" data-line-number="47"><span class="co">##    [4]        surface glycoprotein YP_009724390.1 GeneID:43740568     MFV...HYT</span></a>
<a class="sourceLine" id="cb350-48" data-line-number="48"><span class="co">##    [5]               ORF3a protein YP_009724391.1 GeneID:43740569     MDL...VPL</span></a>
<a class="sourceLine" id="cb350-49" data-line-number="49"><span class="co">##    ...                         ...            ...             ...           ...</span></a>
<a class="sourceLine" id="cb350-50" data-line-number="50"><span class="co">##    [9]               ORF7a protein YP_009724395.1 GeneID:43740573     MKI...KTE</span></a>
<a class="sourceLine" id="cb350-51" data-line-number="51"><span class="co">##   [10]                       ORF7b YP_009725318.1 GeneID:43740574     MIE...CHA</span></a>
<a class="sourceLine" id="cb350-52" data-line-number="52"><span class="co">##   [11]                ORF8 protein YP_009724396.1 GeneID:43740577     MKF...DFI</span></a>
<a class="sourceLine" id="cb350-53" data-line-number="53"><span class="co">##   [12] nucleocapsid phosphoprotein YP_009724397.2 GeneID:43740575     MSD...TQA</span></a>
<a class="sourceLine" id="cb350-54" data-line-number="54"><span class="co">##   [13]               ORF10 protein YP_009725255.1 GeneID:43740576     MGY...NLT</span></a>
<a class="sourceLine" id="cb350-55" data-line-number="55"><span class="co">##            loctype       gene_synonym     gene_id transcript_id</span></a>
<a class="sourceLine" id="cb350-56" data-line-number="56"><span class="co">##        &lt;character&gt;    &lt;CharacterList&gt; &lt;character&gt;   &lt;character&gt;</span></a>
<a class="sourceLine" id="cb350-57" data-line-number="57"><span class="co">##    [1]      normal                 NA  GU280_gp01  GU280_gp01.1</span></a>
<a class="sourceLine" id="cb350-58" data-line-number="58"><span class="co">##    [2]      normal                 NA  GU280_gp01  GU280_gp01.1</span></a>
<a class="sourceLine" id="cb350-59" data-line-number="59"><span class="co">##    [3]      normal                 NA  GU280_gp01  GU280_gp01.2</span></a>
<a class="sourceLine" id="cb350-60" data-line-number="60"><span class="co">##    [4]      normal spike glycoprotein  GU280_gp02  GU280_gp02.1</span></a>
<a class="sourceLine" id="cb350-61" data-line-number="61"><span class="co">##    [5]      normal                 NA  GU280_gp03  GU280_gp03.1</span></a>
<a class="sourceLine" id="cb350-62" data-line-number="62"><span class="co">##    ...         ...                ...         ...           ...</span></a>
<a class="sourceLine" id="cb350-63" data-line-number="63"><span class="co">##    [9]      normal                 NA  GU280_gp07  GU280_gp07.1</span></a>
<a class="sourceLine" id="cb350-64" data-line-number="64"><span class="co">##   [10]      normal                 NA  GU280_gp08  GU280_gp08.1</span></a>
<a class="sourceLine" id="cb350-65" data-line-number="65"><span class="co">##   [11]      normal                 NA  GU280_gp09  GU280_gp09.1</span></a>
<a class="sourceLine" id="cb350-66" data-line-number="66"><span class="co">##   [12]      normal                 NA  GU280_gp10  GU280_gp10.1</span></a>
<a class="sourceLine" id="cb350-67" data-line-number="67"><span class="co">##   [13]      normal                 NA  GU280_gp11  GU280_gp11.1</span></a>
<a class="sourceLine" id="cb350-68" data-line-number="68"><span class="co">##   -------</span></a>
<a class="sourceLine" id="cb350-69" data-line-number="69"><span class="co">##   seqinfo: 1 sequence from NC_045512.2 genome</span></a></code></pre></div>
<p>or coerce to a tibble and explore</p>
<div class="sourceCode" id="cb351"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb351-1" data-line-number="1"><span class="kw">library</span>(tibble)</a>
<a class="sourceLine" id="cb351-2" data-line-number="2"><span class="kw">library</span>(dplyr)</a>
<a class="sourceLine" id="cb351-3" data-line-number="3"><span class="co">## </span></a>
<a class="sourceLine" id="cb351-4" data-line-number="4"><span class="co">## Attaching package: &#39;dplyr&#39;</span></a>
<a class="sourceLine" id="cb351-5" data-line-number="5"><span class="co">## The following objects are masked from &#39;package:GenomicRanges&#39;:</span></a>
<a class="sourceLine" id="cb351-6" data-line-number="6"><span class="co">## </span></a>
<a class="sourceLine" id="cb351-7" data-line-number="7"><span class="co">##     intersect, setdiff, union</span></a>
<a class="sourceLine" id="cb351-8" data-line-number="8"><span class="co">## The following object is masked from &#39;package:GenomeInfoDb&#39;:</span></a>
<a class="sourceLine" id="cb351-9" data-line-number="9"><span class="co">## </span></a>
<a class="sourceLine" id="cb351-10" data-line-number="10"><span class="co">##     intersect</span></a>
<a class="sourceLine" id="cb351-11" data-line-number="11"><span class="co">## The following objects are masked from &#39;package:Biostrings&#39;:</span></a>
<a class="sourceLine" id="cb351-12" data-line-number="12"><span class="co">## </span></a>
<a class="sourceLine" id="cb351-13" data-line-number="13"><span class="co">##     collapse, intersect, setdiff, setequal, union</span></a>
<a class="sourceLine" id="cb351-14" data-line-number="14"><span class="co">## The following object is masked from &#39;package:XVector&#39;:</span></a>
<a class="sourceLine" id="cb351-15" data-line-number="15"><span class="co">## </span></a>
<a class="sourceLine" id="cb351-16" data-line-number="16"><span class="co">##     slice</span></a>
<a class="sourceLine" id="cb351-17" data-line-number="17"><span class="co">## The following objects are masked from &#39;package:IRanges&#39;:</span></a>
<a class="sourceLine" id="cb351-18" data-line-number="18"><span class="co">## </span></a>
<a class="sourceLine" id="cb351-19" data-line-number="19"><span class="co">##     collapse, desc, intersect, setdiff, slice, union</span></a>
<a class="sourceLine" id="cb351-20" data-line-number="20"><span class="co">## The following objects are masked from &#39;package:S4Vectors&#39;:</span></a>
<a class="sourceLine" id="cb351-21" data-line-number="21"><span class="co">## </span></a>
<a class="sourceLine" id="cb351-22" data-line-number="22"><span class="co">##     first, intersect, rename, setdiff, setequal, union</span></a>
<a class="sourceLine" id="cb351-23" data-line-number="23"><span class="co">## The following objects are masked from &#39;package:BiocGenerics&#39;:</span></a>
<a class="sourceLine" id="cb351-24" data-line-number="24"><span class="co">## </span></a>
<a class="sourceLine" id="cb351-25" data-line-number="25"><span class="co">##     combine, intersect, setdiff, union</span></a>
<a class="sourceLine" id="cb351-26" data-line-number="26"><span class="co">## The following objects are masked from &#39;package:stats&#39;:</span></a>
<a class="sourceLine" id="cb351-27" data-line-number="27"><span class="co">## </span></a>
<a class="sourceLine" id="cb351-28" data-line-number="28"><span class="co">##     filter, lag</span></a>
<a class="sourceLine" id="cb351-29" data-line-number="29"><span class="co">## The following objects are masked from &#39;package:base&#39;:</span></a>
<a class="sourceLine" id="cb351-30" data-line-number="30"><span class="co">## </span></a>
<a class="sourceLine" id="cb351-31" data-line-number="31"><span class="co">##     intersect, setdiff, setequal, union</span></a>
<a class="sourceLine" id="cb351-32" data-line-number="32"><span class="kw">as_tibble</span>(cds)</a>
<a class="sourceLine" id="cb351-33" data-line-number="33"><span class="co">## # A tibble: 13 x 19</span></a>
<a class="sourceLine" id="cb351-34" data-line-number="34"><span class="co">##    seqnames start   end width strand type  gene  locus_tag ribosomal_slipp…</span></a>
<a class="sourceLine" id="cb351-35" data-line-number="35"><span class="co">##    &lt;fct&gt;    &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;fct&gt;  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;     &lt;lgl&gt;           </span></a>
<a class="sourceLine" id="cb351-36" data-line-number="36"><span class="co">##  1 Severe …   266 13468 13203 +      CDS   ORF1… GU280_gp… TRUE            </span></a>
<a class="sourceLine" id="cb351-37" data-line-number="37"><span class="co">##  2 Severe … 13468 21555  8088 +      CDS   ORF1… GU280_gp… TRUE            </span></a>
<a class="sourceLine" id="cb351-38" data-line-number="38"><span class="co">##  3 Severe …   266 13483 13218 +      CDS   ORF1… GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb351-39" data-line-number="39"><span class="co">##  4 Severe … 21563 25384  3822 +      CDS   S     GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb351-40" data-line-number="40"><span class="co">##  5 Severe … 25393 26220   828 +      CDS   ORF3a GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb351-41" data-line-number="41"><span class="co">##  6 Severe … 26245 26472   228 +      CDS   E     GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb351-42" data-line-number="42"><span class="co">##  7 Severe … 26523 27191   669 +      CDS   M     GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb351-43" data-line-number="43"><span class="co">##  8 Severe … 27202 27387   186 +      CDS   ORF6  GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb351-44" data-line-number="44"><span class="co">##  9 Severe … 27394 27759   366 +      CDS   ORF7a GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb351-45" data-line-number="45"><span class="co">## 10 Severe … 27756 27887   132 +      CDS   ORF7b GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb351-46" data-line-number="46"><span class="co">## 11 Severe … 27894 28259   366 +      CDS   ORF8  GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb351-47" data-line-number="47"><span class="co">## 12 Severe … 28274 29533  1260 +      CDS   N     GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb351-48" data-line-number="48"><span class="co">## 13 Severe … 29558 29674   117 +      CDS   ORF10 GU280_gp… FALSE           </span></a>
<a class="sourceLine" id="cb351-49" data-line-number="49"><span class="co">## # … with 10 more variables: note &lt;chr&gt;, codon_start &lt;dbl&gt;, product &lt;chr&gt;,</span></a>
<a class="sourceLine" id="cb351-50" data-line-number="50"><span class="co">## #   protein_id &lt;chr&gt;, db_xref &lt;I&lt;list&gt;&gt;, translation &lt;chr&gt;, loctype &lt;chr&gt;,</span></a>
<a class="sourceLine" id="cb351-51" data-line-number="51"><span class="co">## #   gene_synonym &lt;I&lt;list&gt;&gt;, gene_id &lt;chr&gt;, transcript_id &lt;chr&gt;</span></a></code></pre></div>
<p>For instance, we can find the gene associated with each coding sequence, using the following equivalent commands</p>
<div class="sourceCode" id="cb352"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb352-1" data-line-number="1">cds<span class="op">$</span>gene</a>
<a class="sourceLine" id="cb352-2" data-line-number="2"><span class="co">##  [1] &quot;ORF1ab&quot; &quot;ORF1ab&quot; &quot;ORF1ab&quot; &quot;S&quot;      &quot;ORF3a&quot;  &quot;E&quot;      &quot;M&quot;      &quot;ORF6&quot;  </span></a>
<a class="sourceLine" id="cb352-3" data-line-number="3"><span class="co">##  [9] &quot;ORF7a&quot;  &quot;ORF7b&quot;  &quot;ORF8&quot;   &quot;N&quot;      &quot;ORF10&quot;</span></a>
<a class="sourceLine" id="cb352-4" data-line-number="4"></a>
<a class="sourceLine" id="cb352-5" data-line-number="5"><span class="kw">subset</span>(cds, , gene)</a>
<a class="sourceLine" id="cb352-6" data-line-number="6"><span class="co">## GRanges object with 13 ranges and 1 metadata column:</span></a>
<a class="sourceLine" id="cb352-7" data-line-number="7"><span class="co">##                                              seqnames      ranges strand |</span></a>
<a class="sourceLine" id="cb352-8" data-line-number="8"><span class="co">##                                                 &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; |</span></a>
<a class="sourceLine" id="cb352-9" data-line-number="9"><span class="co">##    [1] Severe acute respiratory syndrome coronavirus2   266-13468      + |</span></a>
<a class="sourceLine" id="cb352-10" data-line-number="10"><span class="co">##    [2] Severe acute respiratory syndrome coronavirus2 13468-21555      + |</span></a>
<a class="sourceLine" id="cb352-11" data-line-number="11"><span class="co">##    [3] Severe acute respiratory syndrome coronavirus2   266-13483      + |</span></a>
<a class="sourceLine" id="cb352-12" data-line-number="12"><span class="co">##    [4] Severe acute respiratory syndrome coronavirus2 21563-25384      + |</span></a>
<a class="sourceLine" id="cb352-13" data-line-number="13"><span class="co">##    [5] Severe acute respiratory syndrome coronavirus2 25393-26220      + |</span></a>
<a class="sourceLine" id="cb352-14" data-line-number="14"><span class="co">##    ...                                            ...         ...    ... .</span></a>
<a class="sourceLine" id="cb352-15" data-line-number="15"><span class="co">##    [9] Severe acute respiratory syndrome coronavirus2 27394-27759      + |</span></a>
<a class="sourceLine" id="cb352-16" data-line-number="16"><span class="co">##   [10] Severe acute respiratory syndrome coronavirus2 27756-27887      + |</span></a>
<a class="sourceLine" id="cb352-17" data-line-number="17"><span class="co">##   [11] Severe acute respiratory syndrome coronavirus2 27894-28259      + |</span></a>
<a class="sourceLine" id="cb352-18" data-line-number="18"><span class="co">##   [12] Severe acute respiratory syndrome coronavirus2 28274-29533      + |</span></a>
<a class="sourceLine" id="cb352-19" data-line-number="19"><span class="co">##   [13] Severe acute respiratory syndrome coronavirus2 29558-29674      + |</span></a>
<a class="sourceLine" id="cb352-20" data-line-number="20"><span class="co">##               gene</span></a>
<a class="sourceLine" id="cb352-21" data-line-number="21"><span class="co">##        &lt;character&gt;</span></a>
<a class="sourceLine" id="cb352-22" data-line-number="22"><span class="co">##    [1]      ORF1ab</span></a>
<a class="sourceLine" id="cb352-23" data-line-number="23"><span class="co">##    [2]      ORF1ab</span></a>
<a class="sourceLine" id="cb352-24" data-line-number="24"><span class="co">##    [3]      ORF1ab</span></a>
<a class="sourceLine" id="cb352-25" data-line-number="25"><span class="co">##    [4]           S</span></a>
<a class="sourceLine" id="cb352-26" data-line-number="26"><span class="co">##    [5]       ORF3a</span></a>
<a class="sourceLine" id="cb352-27" data-line-number="27"><span class="co">##    ...         ...</span></a>
<a class="sourceLine" id="cb352-28" data-line-number="28"><span class="co">##    [9]       ORF7a</span></a>
<a class="sourceLine" id="cb352-29" data-line-number="29"><span class="co">##   [10]       ORF7b</span></a>
<a class="sourceLine" id="cb352-30" data-line-number="30"><span class="co">##   [11]        ORF8</span></a>
<a class="sourceLine" id="cb352-31" data-line-number="31"><span class="co">##   [12]           N</span></a>
<a class="sourceLine" id="cb352-32" data-line-number="32"><span class="co">##   [13]       ORF10</span></a>
<a class="sourceLine" id="cb352-33" data-line-number="33"><span class="co">##   -------</span></a>
<a class="sourceLine" id="cb352-34" data-line-number="34"><span class="co">##   seqinfo: 1 sequence from NC_045512.2 genome</span></a>
<a class="sourceLine" id="cb352-35" data-line-number="35"></a>
<a class="sourceLine" id="cb352-36" data-line-number="36"><span class="kw">as_tibble</span>(cds) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb352-37" data-line-number="37"><span class="st">    </span><span class="co">## info on gene and protein product</span></a>
<a class="sourceLine" id="cb352-38" data-line-number="38"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw">select</span>(gene, product)</a>
<a class="sourceLine" id="cb352-39" data-line-number="39"><span class="co">## # A tibble: 13 x 2</span></a>
<a class="sourceLine" id="cb352-40" data-line-number="40"><span class="co">##    gene   product                    </span></a>
<a class="sourceLine" id="cb352-41" data-line-number="41"><span class="co">##    &lt;chr&gt;  &lt;chr&gt;                      </span></a>
<a class="sourceLine" id="cb352-42" data-line-number="42"><span class="co">##  1 ORF1ab ORF1ab polyprotein         </span></a>
<a class="sourceLine" id="cb352-43" data-line-number="43"><span class="co">##  2 ORF1ab ORF1ab polyprotein         </span></a>
<a class="sourceLine" id="cb352-44" data-line-number="44"><span class="co">##  3 ORF1ab ORF1a polyprotein          </span></a>
<a class="sourceLine" id="cb352-45" data-line-number="45"><span class="co">##  4 S      surface glycoprotein       </span></a>
<a class="sourceLine" id="cb352-46" data-line-number="46"><span class="co">##  5 ORF3a  ORF3a protein              </span></a>
<a class="sourceLine" id="cb352-47" data-line-number="47"><span class="co">##  6 E      envelope protein           </span></a>
<a class="sourceLine" id="cb352-48" data-line-number="48"><span class="co">##  7 M      membrane glycoprotein      </span></a>
<a class="sourceLine" id="cb352-49" data-line-number="49"><span class="co">##  8 ORF6   ORF6 protein               </span></a>
<a class="sourceLine" id="cb352-50" data-line-number="50"><span class="co">##  9 ORF7a  ORF7a protein              </span></a>
<a class="sourceLine" id="cb352-51" data-line-number="51"><span class="co">## 10 ORF7b  ORF7b                      </span></a>
<a class="sourceLine" id="cb352-52" data-line-number="52"><span class="co">## 11 ORF8   ORF8 protein               </span></a>
<a class="sourceLine" id="cb352-53" data-line-number="53"><span class="co">## 12 N      nucleocapsid phosphoprotein</span></a>
<a class="sourceLine" id="cb352-54" data-line-number="54"><span class="co">## 13 ORF10  ORF10 protein</span></a></code></pre></div></li>
</ul>
<p>Using genomic ranges to extract DNA sequences</p>
<ul>
<li><p>Recall the sequence of the entire genome</p>
<div class="sourceCode" id="cb353"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb353-1" data-line-number="1">ref &lt;-<span class="st"> </span><span class="kw">getSeq</span>(gb)</a>
<a class="sourceLine" id="cb353-2" data-line-number="2">ref</a>
<a class="sourceLine" id="cb353-3" data-line-number="3"><span class="co">##   A DNAStringSet instance of length 1</span></a>
<a class="sourceLine" id="cb353-4" data-line-number="4"><span class="co">##     width seq                                               names               </span></a>
<a class="sourceLine" id="cb353-5" data-line-number="5"><span class="co">## [1] 29903 ATTAAAGGTTTATACCTTCCCAG...AAAAAAAAAAAAAAAAAAAAAAA Severe acute resp...</span></a></code></pre></div></li>
<li><p>Use <code>getSeq()</code> to extract the <code>DNAStringSet()</code> corresponding to the genomic ranges specified by <code>cds</code></p>
<div class="sourceCode" id="cb354"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb354-1" data-line-number="1">dna &lt;-<span class="st"> </span><span class="kw">getSeq</span>(ref, cds)</a>
<a class="sourceLine" id="cb354-2" data-line-number="2">dna</a>
<a class="sourceLine" id="cb354-3" data-line-number="3"><span class="co">##   A DNAStringSet instance of length 13</span></a>
<a class="sourceLine" id="cb354-4" data-line-number="4"><span class="co">##      width seq</span></a>
<a class="sourceLine" id="cb354-5" data-line-number="5"><span class="co">##  [1] 13203 ATGGAGAGCCTTGTCCCTGGTTTCAACGAGAAA...CTTCAGTCAGCTGATGCACAATCGTTTTTAAAC</span></a>
<a class="sourceLine" id="cb354-6" data-line-number="6"><span class="co">##  [2]  8088 CGGGTTTGCGGTGTAAGTGCAGCCCGTCTTACA...GTTATTTCTAGTGATGTTCTTGTTAACAACTAA</span></a>
<a class="sourceLine" id="cb354-7" data-line-number="7"><span class="co">##  [3] 13218 ATGGAGAGCCTTGTCCCTGGTTTCAACGAGAAA...GCACAATCGTTTTTAAACGGGTTTGCGGTGTAA</span></a>
<a class="sourceLine" id="cb354-8" data-line-number="8"><span class="co">##  [4]  3822 ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTC...GTGCTCAAAGGAGTCAAATTACATTACACATAA</span></a>
<a class="sourceLine" id="cb354-9" data-line-number="9"><span class="co">##  [5]   828 ATGGATTTGTTTATGAGAATCTTCACAATTGGA...GAACCGACGACGACTACTAGCGTGCCTTTGTAA</span></a>
<a class="sourceLine" id="cb354-10" data-line-number="10"><span class="co">##  ...   ... ...</span></a>
<a class="sourceLine" id="cb354-11" data-line-number="11"><span class="co">##  [9]   366 ATGAAAATTATTCTTTTCTTGGCACTGATAACA...CTTTGCTTCACACTCAAAAGAAAGACAGAATGA</span></a>
<a class="sourceLine" id="cb354-12" data-line-number="12"><span class="co">## [10]   132 ATGATTGAACTTTCATTAATTGACTTCTATTTG...CTGCAAGATCATAATGAAACTTGTCACGCCTAA</span></a>
<a class="sourceLine" id="cb354-13" data-line-number="13"><span class="co">## [11]   366 ATGAAATTTCTTGTTTTCTTAGGAATCATCACA...CATGACGTTCGTGTTGTTTTAGATTTCATCTAA</span></a>
<a class="sourceLine" id="cb354-14" data-line-number="14"><span class="co">## [12]  1260 ATGTCTGATAATGGACCCCAAAATCAGCGAAAT...TCCATGAGCAGTGCTGACTCAACTCAGGCCTAA</span></a>
<a class="sourceLine" id="cb354-15" data-line-number="15"><span class="co">## [13]   117 ATGGGCTATATAAACGTTTTCGCTTTTCCGTTT...CAAGTAGATGTAGTTAACTTTAATCTCACATAG</span></a></code></pre></div></li>
<li><p>It might be useful to coordinate information about each coding sequence with the sequence itself</p>
<div class="sourceCode" id="cb355"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb355-1" data-line-number="1">cds<span class="op">$</span>DNA &lt;-<span class="st"> </span>dna</a>
<a class="sourceLine" id="cb355-2" data-line-number="2">cds</a>
<a class="sourceLine" id="cb355-3" data-line-number="3"><span class="co">## GRanges object with 13 ranges and 15 metadata columns:</span></a>
<a class="sourceLine" id="cb355-4" data-line-number="4"><span class="co">##                                              seqnames      ranges strand |</span></a>
<a class="sourceLine" id="cb355-5" data-line-number="5"><span class="co">##                                                 &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; |</span></a>
<a class="sourceLine" id="cb355-6" data-line-number="6"><span class="co">##    [1] Severe acute respiratory syndrome coronavirus2   266-13468      + |</span></a>
<a class="sourceLine" id="cb355-7" data-line-number="7"><span class="co">##    [2] Severe acute respiratory syndrome coronavirus2 13468-21555      + |</span></a>
<a class="sourceLine" id="cb355-8" data-line-number="8"><span class="co">##    [3] Severe acute respiratory syndrome coronavirus2   266-13483      + |</span></a>
<a class="sourceLine" id="cb355-9" data-line-number="9"><span class="co">##    [4] Severe acute respiratory syndrome coronavirus2 21563-25384      + |</span></a>
<a class="sourceLine" id="cb355-10" data-line-number="10"><span class="co">##    [5] Severe acute respiratory syndrome coronavirus2 25393-26220      + |</span></a>
<a class="sourceLine" id="cb355-11" data-line-number="11"><span class="co">##    ...                                            ...         ...    ... .</span></a>
<a class="sourceLine" id="cb355-12" data-line-number="12"><span class="co">##    [9] Severe acute respiratory syndrome coronavirus2 27394-27759      + |</span></a>
<a class="sourceLine" id="cb355-13" data-line-number="13"><span class="co">##   [10] Severe acute respiratory syndrome coronavirus2 27756-27887      + |</span></a>
<a class="sourceLine" id="cb355-14" data-line-number="14"><span class="co">##   [11] Severe acute respiratory syndrome coronavirus2 27894-28259      + |</span></a>
<a class="sourceLine" id="cb355-15" data-line-number="15"><span class="co">##   [12] Severe acute respiratory syndrome coronavirus2 28274-29533      + |</span></a>
<a class="sourceLine" id="cb355-16" data-line-number="16"><span class="co">##   [13] Severe acute respiratory syndrome coronavirus2 29558-29674      + |</span></a>
<a class="sourceLine" id="cb355-17" data-line-number="17"><span class="co">##               type        gene   locus_tag ribosomal_slippage</span></a>
<a class="sourceLine" id="cb355-18" data-line-number="18"><span class="co">##        &lt;character&gt; &lt;character&gt; &lt;character&gt;          &lt;logical&gt;</span></a>
<a class="sourceLine" id="cb355-19" data-line-number="19"><span class="co">##    [1]         CDS      ORF1ab  GU280_gp01               TRUE</span></a>
<a class="sourceLine" id="cb355-20" data-line-number="20"><span class="co">##    [2]         CDS      ORF1ab  GU280_gp01               TRUE</span></a>
<a class="sourceLine" id="cb355-21" data-line-number="21"><span class="co">##    [3]         CDS      ORF1ab  GU280_gp01              FALSE</span></a>
<a class="sourceLine" id="cb355-22" data-line-number="22"><span class="co">##    [4]         CDS           S  GU280_gp02              FALSE</span></a>
<a class="sourceLine" id="cb355-23" data-line-number="23"><span class="co">##    [5]         CDS       ORF3a  GU280_gp03              FALSE</span></a>
<a class="sourceLine" id="cb355-24" data-line-number="24"><span class="co">##    ...         ...         ...         ...                ...</span></a>
<a class="sourceLine" id="cb355-25" data-line-number="25"><span class="co">##    [9]         CDS       ORF7a  GU280_gp07              FALSE</span></a>
<a class="sourceLine" id="cb355-26" data-line-number="26"><span class="co">##   [10]         CDS       ORF7b  GU280_gp08              FALSE</span></a>
<a class="sourceLine" id="cb355-27" data-line-number="27"><span class="co">##   [11]         CDS        ORF8  GU280_gp09              FALSE</span></a>
<a class="sourceLine" id="cb355-28" data-line-number="28"><span class="co">##   [12]         CDS           N  GU280_gp10              FALSE</span></a>
<a class="sourceLine" id="cb355-29" data-line-number="29"><span class="co">##   [13]         CDS       ORF10  GU280_gp11              FALSE</span></a>
<a class="sourceLine" id="cb355-30" data-line-number="30"><span class="co">##                                                note codon_start</span></a>
<a class="sourceLine" id="cb355-31" data-line-number="31"><span class="co">##                                         &lt;character&gt;   &lt;numeric&gt;</span></a>
<a class="sourceLine" id="cb355-32" data-line-number="32"><span class="co">##    [1] pp1ab; translated by -1 ribosomal frameshift           1</span></a>
<a class="sourceLine" id="cb355-33" data-line-number="33"><span class="co">##    [2] pp1ab; translated by -1 ribosomal frameshift           1</span></a>
<a class="sourceLine" id="cb355-34" data-line-number="34"><span class="co">##    [3]                                         pp1a           1</span></a>
<a class="sourceLine" id="cb355-35" data-line-number="35"><span class="co">##    [4]            structural protein; spike protein           1</span></a>
<a class="sourceLine" id="cb355-36" data-line-number="36"><span class="co">##    [5]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb355-37" data-line-number="37"><span class="co">##    ...                                          ...         ...</span></a>
<a class="sourceLine" id="cb355-38" data-line-number="38"><span class="co">##    [9]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb355-39" data-line-number="39"><span class="co">##   [10]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb355-40" data-line-number="40"><span class="co">##   [11]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb355-41" data-line-number="41"><span class="co">##   [12]                     ORF9; structural protein           1</span></a>
<a class="sourceLine" id="cb355-42" data-line-number="42"><span class="co">##   [13]                                         &lt;NA&gt;           1</span></a>
<a class="sourceLine" id="cb355-43" data-line-number="43"><span class="co">##                            product     protein_id         db_xref   translation</span></a>
<a class="sourceLine" id="cb355-44" data-line-number="44"><span class="co">##                        &lt;character&gt;    &lt;character&gt; &lt;CharacterList&gt; &lt;AAStringSet&gt;</span></a>
<a class="sourceLine" id="cb355-45" data-line-number="45"><span class="co">##    [1]          ORF1ab polyprotein YP_009724389.1 GeneID:43740578     MES...VNN</span></a>
<a class="sourceLine" id="cb355-46" data-line-number="46"><span class="co">##    [2]          ORF1ab polyprotein YP_009724389.1 GeneID:43740578     MES...VNN</span></a>
<a class="sourceLine" id="cb355-47" data-line-number="47"><span class="co">##    [3]           ORF1a polyprotein YP_009725295.1 GeneID:43740578     MES...FAV</span></a>
<a class="sourceLine" id="cb355-48" data-line-number="48"><span class="co">##    [4]        surface glycoprotein YP_009724390.1 GeneID:43740568     MFV...HYT</span></a>
<a class="sourceLine" id="cb355-49" data-line-number="49"><span class="co">##    [5]               ORF3a protein YP_009724391.1 GeneID:43740569     MDL...VPL</span></a>
<a class="sourceLine" id="cb355-50" data-line-number="50"><span class="co">##    ...                         ...            ...             ...           ...</span></a>
<a class="sourceLine" id="cb355-51" data-line-number="51"><span class="co">##    [9]               ORF7a protein YP_009724395.1 GeneID:43740573     MKI...KTE</span></a>
<a class="sourceLine" id="cb355-52" data-line-number="52"><span class="co">##   [10]                       ORF7b YP_009725318.1 GeneID:43740574     MIE...CHA</span></a>
<a class="sourceLine" id="cb355-53" data-line-number="53"><span class="co">##   [11]                ORF8 protein YP_009724396.1 GeneID:43740577     MKF...DFI</span></a>
<a class="sourceLine" id="cb355-54" data-line-number="54"><span class="co">##   [12] nucleocapsid phosphoprotein YP_009724397.2 GeneID:43740575     MSD...TQA</span></a>
<a class="sourceLine" id="cb355-55" data-line-number="55"><span class="co">##   [13]               ORF10 protein YP_009725255.1 GeneID:43740576     MGY...NLT</span></a>
<a class="sourceLine" id="cb355-56" data-line-number="56"><span class="co">##            loctype       gene_synonym     gene_id transcript_id            DNA</span></a>
<a class="sourceLine" id="cb355-57" data-line-number="57"><span class="co">##        &lt;character&gt;    &lt;CharacterList&gt; &lt;character&gt;   &lt;character&gt; &lt;DNAStringSet&gt;</span></a>
<a class="sourceLine" id="cb355-58" data-line-number="58"><span class="co">##    [1]      normal                 NA  GU280_gp01  GU280_gp01.1      ATG...AAC</span></a>
<a class="sourceLine" id="cb355-59" data-line-number="59"><span class="co">##    [2]      normal                 NA  GU280_gp01  GU280_gp01.1      CGG...TAA</span></a>
<a class="sourceLine" id="cb355-60" data-line-number="60"><span class="co">##    [3]      normal                 NA  GU280_gp01  GU280_gp01.2      ATG...TAA</span></a>
<a class="sourceLine" id="cb355-61" data-line-number="61"><span class="co">##    [4]      normal spike glycoprotein  GU280_gp02  GU280_gp02.1      ATG...TAA</span></a>
<a class="sourceLine" id="cb355-62" data-line-number="62"><span class="co">##    [5]      normal                 NA  GU280_gp03  GU280_gp03.1      ATG...TAA</span></a>
<a class="sourceLine" id="cb355-63" data-line-number="63"><span class="co">##    ...         ...                ...         ...           ...            ...</span></a>
<a class="sourceLine" id="cb355-64" data-line-number="64"><span class="co">##    [9]      normal                 NA  GU280_gp07  GU280_gp07.1      ATG...TGA</span></a>
<a class="sourceLine" id="cb355-65" data-line-number="65"><span class="co">##   [10]      normal                 NA  GU280_gp08  GU280_gp08.1      ATG...TAA</span></a>
<a class="sourceLine" id="cb355-66" data-line-number="66"><span class="co">##   [11]      normal                 NA  GU280_gp09  GU280_gp09.1      ATG...TAA</span></a>
<a class="sourceLine" id="cb355-67" data-line-number="67"><span class="co">##   [12]      normal                 NA  GU280_gp10  GU280_gp10.1      ATG...TAA</span></a>
<a class="sourceLine" id="cb355-68" data-line-number="68"><span class="co">##   [13]      normal                 NA  GU280_gp11  GU280_gp11.1      ATG...TAG</span></a>
<a class="sourceLine" id="cb355-69" data-line-number="69"><span class="co">##   -------</span></a>
<a class="sourceLine" id="cb355-70" data-line-number="70"><span class="co">##   seqinfo: 1 sequence from NC_045512.2 genome</span></a></code></pre></div></li>
</ul>
</div>
</div>
<div id="day-31-sequence-alignment-and-visualization" class="section level2">
<h2><span class="header-section-number">5.3</span> Day 31 Sequence alignment and visualization</h2>
<div id="s-gene-exploration" class="section level3">
<h3><span class="header-section-number">5.3.1</span> ‘S’ gene exploration</h3>
<p>Read the ‘S’ gene sequence extracted from all genbank records into a <code>DNAStringSet</code></p>
<div class="sourceCode" id="cb356"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb356-1" data-line-number="1">url &lt;-<span class="st"> &quot;https://raw.githubusercontent.com/mtmorgan/QuaRantine/master/assets/05-SARS-CoV2-S.fasta&quot;</span></a>
<a class="sourceLine" id="cb356-2" data-line-number="2">fasta_file &lt;-<span class="st"> </span><span class="kw">file.path</span>(workdir, <span class="kw">basename</span>(url))</a>
<a class="sourceLine" id="cb356-3" data-line-number="3"><span class="cf">if</span> ( <span class="op">!</span><span class="kw">file.exists</span>(fasta_file) )</a>
<a class="sourceLine" id="cb356-4" data-line-number="4">    <span class="kw">download.file</span>(url, fasta_file)</a>
<a class="sourceLine" id="cb356-5" data-line-number="5">dna &lt;-<span class="st"> </span><span class="kw">readDNAStringSet</span>(fasta_file)</a>
<a class="sourceLine" id="cb356-6" data-line-number="6">dna</a>
<a class="sourceLine" id="cb356-7" data-line-number="7"><span class="co">##   A DNAStringSet instance of length 2289</span></a>
<a class="sourceLine" id="cb356-8" data-line-number="8"><span class="co">##        width seq                                            names               </span></a>
<a class="sourceLine" id="cb356-9" data-line-number="9"><span class="co">##    [1]  3822 ATGTTTGTTTTTCTTGTTTTAT...GTCAAATTACATTACACATAA LC528232</span></a>
<a class="sourceLine" id="cb356-10" data-line-number="10"><span class="co">##    [2]  3822 ATGTTTGTTTTTCTTGTTTTAT...GTCAAATTACATTACACATAA LC528233</span></a>
<a class="sourceLine" id="cb356-11" data-line-number="11"><span class="co">##    [3]  3822 ATGTTTGTTTTTCTTGTTTTAT...GTCAAATTACATTACACATAA LC529905</span></a>
<a class="sourceLine" id="cb356-12" data-line-number="12"><span class="co">##    [4]  3822 ATGTTTGTTTTTCTTGTTTTAT...GTCAAATTACATTACACATAA LC534418</span></a>
<a class="sourceLine" id="cb356-13" data-line-number="13"><span class="co">##    [5]  3822 ATGTTTGTTTTTCTTGTTTTAT...GTCAAATTACATTACACATAA LC534419</span></a>
<a class="sourceLine" id="cb356-14" data-line-number="14"><span class="co">##    ...   ... ...</span></a>
<a class="sourceLine" id="cb356-15" data-line-number="15"><span class="co">## [2285]  3822 ATGTTTGTTTTTCTTGTTTTAT...GTCAAATTACATTACACATAA MT447174</span></a>
<a class="sourceLine" id="cb356-16" data-line-number="16"><span class="co">## [2286]  3822 ATGTTTGTTTTTCTTGTTTTAT...GTCAAATTACATTACACATAA MT447175</span></a>
<a class="sourceLine" id="cb356-17" data-line-number="17"><span class="co">## [2287]  3822 ATGTTTGTTTTTCTTGTTTTAT...GTCAAATTACATTACACATAA MT447176</span></a>
<a class="sourceLine" id="cb356-18" data-line-number="18"><span class="co">## [2288]  3822 ATGTTTGTTTTTCTTGTTTTAT...GTCAAATTACATTACACATAA MT447177</span></a>
<a class="sourceLine" id="cb356-19" data-line-number="19"><span class="co">## [2289]  3822 ATGTTTGTTTTTCTTGTTTTAT...GTCAAATTACATTACACATAA NC_045512</span></a></code></pre></div>
<p>Explore the data. How many sequences? How many unique sequences?</p>
<div class="sourceCode" id="cb357"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb357-1" data-line-number="1"><span class="kw">length</span>(dna)</a>
<a class="sourceLine" id="cb357-2" data-line-number="2"><span class="co">## [1] 2289</span></a>
<a class="sourceLine" id="cb357-3" data-line-number="3"></a>
<a class="sourceLine" id="cb357-4" data-line-number="4">dna <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">unique</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">length</span>()</a>
<a class="sourceLine" id="cb357-5" data-line-number="5"><span class="co">## [1] 427</span></a></code></pre></div>
<p>Note that the DNA alphabet contains ‘ambiguity’ letters, corresponding to base sequences that were uncertain, e.g., ‘A’ or ‘T’. Check out the help page for <code>?translate</code> to see options for translating ambiguity codes. How many unique protein sequences?</p>
<div class="sourceCode" id="cb358"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb358-1" data-line-number="1">IUPAC_CODE_MAP</a>
<a class="sourceLine" id="cb358-2" data-line-number="2"><span class="co">##      A      C      G      T      M      R      W      S      Y      K      V </span></a>
<a class="sourceLine" id="cb358-3" data-line-number="3"><span class="co">##    &quot;A&quot;    &quot;C&quot;    &quot;G&quot;    &quot;T&quot;   &quot;AC&quot;   &quot;AG&quot;   &quot;AT&quot;   &quot;CG&quot;   &quot;CT&quot;   &quot;GT&quot;  &quot;ACG&quot; </span></a>
<a class="sourceLine" id="cb358-4" data-line-number="4"><span class="co">##      H      D      B      N </span></a>
<a class="sourceLine" id="cb358-5" data-line-number="5"><span class="co">##  &quot;ACT&quot;  &quot;AGT&quot;  &quot;CGT&quot; &quot;ACGT&quot;</span></a>
<a class="sourceLine" id="cb358-6" data-line-number="6"></a>
<a class="sourceLine" id="cb358-7" data-line-number="7">dna <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">alphabetFrequency</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">colSums</span>()</a>
<a class="sourceLine" id="cb358-8" data-line-number="8"><span class="co">##       A       C       G       T       M       R       W       S       Y       K </span></a>
<a class="sourceLine" id="cb358-9" data-line-number="9"><span class="co">## 2563642 1649084 1604491 2898077       2       9       6       0      17       6 </span></a>
<a class="sourceLine" id="cb358-10" data-line-number="10"><span class="co">##       V       H       D       B       N       -       +       . </span></a>
<a class="sourceLine" id="cb358-11" data-line-number="11"><span class="co">##       1       0       1       1   33212       0       0       0</span></a>
<a class="sourceLine" id="cb358-12" data-line-number="12"></a>
<a class="sourceLine" id="cb358-13" data-line-number="13">aa &lt;-<span class="st"> </span>dna <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">translate</span>(<span class="dt">if.fuzzy.codon =</span> <span class="st">&quot;solve&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">unique</span>()</a>
<a class="sourceLine" id="cb358-14" data-line-number="14"></a>
<a class="sourceLine" id="cb358-15" data-line-number="15">aa</a>
<a class="sourceLine" id="cb358-16" data-line-number="16"><span class="co">##   A AAStringSet instance of length 343</span></a>
<a class="sourceLine" id="cb358-17" data-line-number="17"><span class="co">##       width seq                                             names               </span></a>
<a class="sourceLine" id="cb358-18" data-line-number="18"><span class="co">##   [1]  1274 MFVFLVLLPLVSSQCVNLTTRT...CCKFDEDDSEPVLKGVKLHYT* LC528232</span></a>
<a class="sourceLine" id="cb358-19" data-line-number="19"><span class="co">##   [2]  1274 MFVFLVLLPLVSSQCVNLTTRT...CCKFDEDDSEPVLKGVKLHYT* MT007544</span></a>
<a class="sourceLine" id="cb358-20" data-line-number="20"><span class="co">##   [3]  1273 MFVFLVLLPLVSSQCVNLTTRT...CCKFDEDDSEPVLKGVKLHYT* MT012098</span></a>
<a class="sourceLine" id="cb358-21" data-line-number="21"><span class="co">##   [4]  1274 MFVFLVLLPLVSSQCVNLTTRT...CCKFDEDDSEPVLKGVKLHYT* MT020781</span></a>
<a class="sourceLine" id="cb358-22" data-line-number="22"><span class="co">##   [5]  1274 MFVFLVLLPLVSSQCVNLTTRT...CCKFDEDDSEPVLKGVKLHYT* MT039890</span></a>
<a class="sourceLine" id="cb358-23" data-line-number="23"><span class="co">##   ...   ... ...</span></a>
<a class="sourceLine" id="cb358-24" data-line-number="24"><span class="co">## [339]  1274 MFVFLVLLPLVSSQCVNLTTRT...CCKFDEDDSEPVLKGVKLHYT* MT446361</span></a>
<a class="sourceLine" id="cb358-25" data-line-number="25"><span class="co">## [340]  1274 MFVFLVLLPLVSSQCVNLTTRT...CCKFDEDDSEPVLKGVKLHYT* MT447160</span></a>
<a class="sourceLine" id="cb358-26" data-line-number="26"><span class="co">## [341]  1274 MFVFLVLLPLVSSQCVNLTTRT...CCKFDEDDSEPVLKGVKLHYT* MT447163</span></a>
<a class="sourceLine" id="cb358-27" data-line-number="27"><span class="co">## [342]  1274 MFVFLVLLPLVSSQCVNLTTRT...CCKFDEDDSEPVLKGVKLHYT* MT447168</span></a>
<a class="sourceLine" id="cb358-28" data-line-number="28"><span class="co">## [343]  1274 MFVFLVLLPLVSSQCVNLTTRI...CCKFDEDDSEPVLKGVKLHYT* MT447177</span></a></code></pre></div>
<p>The <code>table()</code> function counts the number of occurrences of each element, so</p>
<div class="sourceCode" id="cb359"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb359-1" data-line-number="1"><span class="kw">nchar</span>(aa) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">table</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as_tibble</span>()</a>
<a class="sourceLine" id="cb359-2" data-line-number="2"><span class="co">## # A tibble: 2 x 2</span></a>
<a class="sourceLine" id="cb359-3" data-line-number="3"><span class="co">##   .         n</span></a>
<a class="sourceLine" id="cb359-4" data-line-number="4"><span class="co">##   &lt;chr&gt; &lt;int&gt;</span></a>
<a class="sourceLine" id="cb359-5" data-line-number="5"><span class="co">## 1 1273      3</span></a>
<a class="sourceLine" id="cb359-6" data-line-number="6"><span class="co">## 2 1274    340</span></a></code></pre></div>
<p>shows us that most sequences have 1274 amino acids.</p>
<p>The <code>names(aa)</code> are misleading – these are the identifiers of a representative accession, but several accessions may map to the same identifier. The following code creates a table that indicates which cluster each original identifier belongs to; the names of <code>aa</code> are updated accordingly.</p>
<div class="sourceCode" id="cb360"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb360-1" data-line-number="1">all &lt;-<span class="st"> </span>dna <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">translate</span>(<span class="dt">if.fuzzy.codon =</span> <span class="st">&quot;solve&quot;</span>)</a>
<a class="sourceLine" id="cb360-2" data-line-number="2"><span class="co">## match(x, y) finds the index of each element of x in y.</span></a>
<a class="sourceLine" id="cb360-3" data-line-number="3"><span class="co">## match(c(&quot;apple&quot;, &quot;pear&quot;), c(&quot;pear&quot;, &quot;apple&quot;, &quot;banana&quot;)) finds</span></a>
<a class="sourceLine" id="cb360-4" data-line-number="4"><span class="co">## that &#39;apple&#39; in x matches the second element of y, and &#39;pear&#39;</span></a>
<a class="sourceLine" id="cb360-5" data-line-number="5"><span class="co">## matches the first element of y. So the result is c(2, 1).</span></a>
<a class="sourceLine" id="cb360-6" data-line-number="6">id &lt;-<span class="st"> </span><span class="kw">match</span>(all, aa)</a>
<a class="sourceLine" id="cb360-7" data-line-number="7">clusters &lt;-<span class="st"> </span><span class="kw">tibble</span>(</a>
<a class="sourceLine" id="cb360-8" data-line-number="8">    <span class="dt">accession =</span> <span class="kw">names</span>(all),</a>
<a class="sourceLine" id="cb360-9" data-line-number="9">    <span class="dt">label =</span> <span class="kw">as.character</span>(id)</a>
<a class="sourceLine" id="cb360-10" data-line-number="10">)</a>
<a class="sourceLine" id="cb360-11" data-line-number="11"><span class="kw">names</span>(aa) &lt;-<span class="st"> </span><span class="kw">match</span>(aa, aa)</a></code></pre></div>
<p>With this information, it’s easy to count the number of times each sequence was represented. The reference sequence belongs to cluster 1.</p>
<div class="sourceCode" id="cb361"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb361-1" data-line-number="1">clusters <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(accession <span class="op">==</span><span class="st"> &quot;NC_045512&quot;</span>)</a>
<a class="sourceLine" id="cb361-2" data-line-number="2"><span class="co">## # A tibble: 1 x 2</span></a>
<a class="sourceLine" id="cb361-3" data-line-number="3"><span class="co">##   accession label</span></a>
<a class="sourceLine" id="cb361-4" data-line-number="4"><span class="co">##   &lt;chr&gt;     &lt;chr&gt;</span></a>
<a class="sourceLine" id="cb361-5" data-line-number="5"><span class="co">## 1 NC_045512 1</span></a>
<a class="sourceLine" id="cb361-6" data-line-number="6"></a>
<a class="sourceLine" id="cb361-7" data-line-number="7">clusters <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb361-8" data-line-number="8"><span class="st">    </span><span class="kw">count</span>(label) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb361-9" data-line-number="9"><span class="st">    </span><span class="kw">arrange</span>(<span class="kw">desc</span>(n))</a>
<a class="sourceLine" id="cb361-10" data-line-number="10"><span class="co">## # A tibble: 343 x 2</span></a>
<a class="sourceLine" id="cb361-11" data-line-number="11"><span class="co">##    label     n</span></a>
<a class="sourceLine" id="cb361-12" data-line-number="12"><span class="co">##    &lt;chr&gt; &lt;int&gt;</span></a>
<a class="sourceLine" id="cb361-13" data-line-number="13"><span class="co">##  1 11     1117</span></a>
<a class="sourceLine" id="cb361-14" data-line-number="14"><span class="co">##  2 1       723</span></a>
<a class="sourceLine" id="cb361-15" data-line-number="15"><span class="co">##  3 126      21</span></a>
<a class="sourceLine" id="cb361-16" data-line-number="16"><span class="co">##  4 34       10</span></a>
<a class="sourceLine" id="cb361-17" data-line-number="17"><span class="co">##  5 341      10</span></a>
<a class="sourceLine" id="cb361-18" data-line-number="18"><span class="co">##  6 125       6</span></a>
<a class="sourceLine" id="cb361-19" data-line-number="19"><span class="co">##  7 167       6</span></a>
<a class="sourceLine" id="cb361-20" data-line-number="20"><span class="co">##  8 47        6</span></a>
<a class="sourceLine" id="cb361-21" data-line-number="21"><span class="co">##  9 59        6</span></a>
<a class="sourceLine" id="cb361-22" data-line-number="22"><span class="co">## 10 170       5</span></a>
<a class="sourceLine" id="cb361-23" data-line-number="23"><span class="co">## # … with 333 more rows</span></a></code></pre></div>
</div>
<div id="sequence-alignment" class="section level3">
<h3><span class="header-section-number">5.3.2</span> Sequence alignment</h3>
<p>Setup</p>
<ul>
<li><p>Create a working directory, if necessary. E.g.,</p>
<div class="sourceCode" id="cb362"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb362-1" data-line-number="1">workdir &lt;-<span class="st"> &quot;workdir&quot;</span></a>
<a class="sourceLine" id="cb362-2" data-line-number="2"><span class="cf">if</span> (<span class="op">!</span><span class="kw">dir.exists</span>(workdir))</a>
<a class="sourceLine" id="cb362-3" data-line-number="3">    <span class="kw">dir.create</span>(workdir)</a></code></pre></div></li>
<li><p>Install (if necessary, using <code>BiocManager::install(&quot;DECIPHER&quot;)</code>) and attach the <a href="https://bioconductor.org/packages/DECIPHER">DECIPHER</a> package</p>
<div class="sourceCode" id="cb363"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb363-1" data-line-number="1"><span class="kw">library</span>(DECIPHER)</a>
<a class="sourceLine" id="cb363-2" data-line-number="2"></a>
<a class="sourceLine" id="cb363-3" data-line-number="3"><span class="kw">maintainer</span>(<span class="st">&quot;DECIPHER&quot;</span>)</a>
<a class="sourceLine" id="cb363-4" data-line-number="4"><span class="co">## [1] &quot;Erik Wright &lt;eswright@pitt.edu&gt;&quot;</span></a>
<a class="sourceLine" id="cb363-5" data-line-number="5"></a>
<a class="sourceLine" id="cb363-6" data-line-number="6"><span class="kw">citation</span>(<span class="st">&quot;DECIPHER&quot;</span>)</a>
<a class="sourceLine" id="cb363-7" data-line-number="7"><span class="co">## </span></a>
<a class="sourceLine" id="cb363-8" data-line-number="8"><span class="co">## Wright ES (2016). &quot;Using DECIPHER v2.0 to Analyze Big Biological</span></a>
<a class="sourceLine" id="cb363-9" data-line-number="9"><span class="co">## Sequence Data in R.&quot; _The R Journal_, *8*(1), 352-359.</span></a>
<a class="sourceLine" id="cb363-10" data-line-number="10"><span class="co">## </span></a>
<a class="sourceLine" id="cb363-11" data-line-number="11"><span class="co">## A BibTeX entry for LaTeX users is</span></a>
<a class="sourceLine" id="cb363-12" data-line-number="12"><span class="co">## </span></a>
<a class="sourceLine" id="cb363-13" data-line-number="13"><span class="co">##   @Article{,</span></a>
<a class="sourceLine" id="cb363-14" data-line-number="14"><span class="co">##     title = {Using DECIPHER v2.0 to Analyze Big Biological Sequence Data in R},</span></a>
<a class="sourceLine" id="cb363-15" data-line-number="15"><span class="co">##     author = {Erik S. Wright},</span></a>
<a class="sourceLine" id="cb363-16" data-line-number="16"><span class="co">##     journal = {The R Journal},</span></a>
<a class="sourceLine" id="cb363-17" data-line-number="17"><span class="co">##     year = {2016},</span></a>
<a class="sourceLine" id="cb363-18" data-line-number="18"><span class="co">##     volume = {8},</span></a>
<a class="sourceLine" id="cb363-19" data-line-number="19"><span class="co">##     number = {1},</span></a>
<a class="sourceLine" id="cb363-20" data-line-number="20"><span class="co">##     pages = {352-359},</span></a>
<a class="sourceLine" id="cb363-21" data-line-number="21"><span class="co">##   }</span></a></code></pre></div></li>
<li><p>This package contains tools for multiple sequence alignment. Explore the introductory and other vignettes in this package.</p>
<div class="sourceCode" id="cb364"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb364-1" data-line-number="1"><span class="kw">browseVignettes</span>(<span class="dt">package =</span> <span class="st">&quot;DECIPHER&quot;</span>)</a></code></pre></div></li>
<li><p>Check out one of the help pages, e.g.,</p>
<div class="sourceCode" id="cb365"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb365-1" data-line-number="1">?AlignSeqs</a></code></pre></div>
<p>Use <code>example(AlignSeqs)</code> to run the examples on the help page.</p></li>
</ul>
<p>Perform multiple alignment</p>
<ul>
<li><p>Perform multiple alignment on the unique amino acid sequences using DECIPHER; this has minimal effect, with single insertions in the few sequences with unusual <code>nchar()</code>.</p>
<div class="sourceCode" id="cb366"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb366-1" data-line-number="1">aln &lt;-<span class="st"> </span><span class="kw">AlignSeqs</span>(aa, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb366-2" data-line-number="2">aln</a>
<a class="sourceLine" id="cb366-3" data-line-number="3"><span class="co">##   A AAStringSet instance of length 343</span></a>
<a class="sourceLine" id="cb366-4" data-line-number="4"><span class="co">##       width seq                                             names               </span></a>
<a class="sourceLine" id="cb366-5" data-line-number="5"><span class="co">##   [1]  1274 MFVFLVLLPLVSSQCVNLTTRT...CCKFDEDDSEPVLKGVKLHYT* 1</span></a>
<a class="sourceLine" id="cb366-6" data-line-number="6"><span class="co">##   [2]  1274 MFVFLVLLPLVSSQCVNLTTRT...CCKFDEDDSEPVLKGVKLHYT* 2</span></a>
<a class="sourceLine" id="cb366-7" data-line-number="7"><span class="co">##   [3]  1274 MFVFLVLLPLVSSQCVNLTTRT...CCKFDEDDSEPVLKGVKLHYT* 3</span></a>
<a class="sourceLine" id="cb366-8" data-line-number="8"><span class="co">##   [4]  1274 MFVFLVLLPLVSSQCVNLTTRT...CCKFDEDDSEPVLKGVKLHYT* 4</span></a>
<a class="sourceLine" id="cb366-9" data-line-number="9"><span class="co">##   [5]  1274 MFVFLVLLPLVSSQCVNLTTRT...CCKFDEDDSEPVLKGVKLHYT* 5</span></a>
<a class="sourceLine" id="cb366-10" data-line-number="10"><span class="co">##   ...   ... ...</span></a>
<a class="sourceLine" id="cb366-11" data-line-number="11"><span class="co">## [339]  1274 MFVFLVLLPLVSSQCVNLTTRT...CCKFDEDDSEPVLKGVKLHYT* 339</span></a>
<a class="sourceLine" id="cb366-12" data-line-number="12"><span class="co">## [340]  1274 MFVFLVLLPLVSSQCVNLTTRT...CCKFDEDDSEPVLKGVKLHYT* 340</span></a>
<a class="sourceLine" id="cb366-13" data-line-number="13"><span class="co">## [341]  1274 MFVFLVLLPLVSSQCVNLTTRT...CCKFDEDDSEPVLKGVKLHYT* 341</span></a>
<a class="sourceLine" id="cb366-14" data-line-number="14"><span class="co">## [342]  1274 MFVFLVLLPLVSSQCVNLTTRT...CCKFDEDDSEPVLKGVKLHYT* 342</span></a>
<a class="sourceLine" id="cb366-15" data-line-number="15"><span class="co">## [343]  1274 MFVFLVLLPLVSSQCVNLTTRI...CCKFDEDDSEPVLKGVKLHYT* 343</span></a>
<a class="sourceLine" id="cb366-16" data-line-number="16"></a>
<a class="sourceLine" id="cb366-17" data-line-number="17"><span class="kw">nchar</span>(aln) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">table</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as_tibble</span>()</a>
<a class="sourceLine" id="cb366-18" data-line-number="18"><span class="co">## # A tibble: 1 x 2</span></a>
<a class="sourceLine" id="cb366-19" data-line-number="19"><span class="co">##   .         n</span></a>
<a class="sourceLine" id="cb366-20" data-line-number="20"><span class="co">##   &lt;chr&gt; &lt;int&gt;</span></a>
<a class="sourceLine" id="cb366-21" data-line-number="21"><span class="co">## 1 1274    343</span></a></code></pre></div></li>
<li><p>The ‘distance matrix’ measures the proximity (Hamming distance) of each sequence to other sequences in sequence space. The result is a large matrix of pairwise distances</p>
<div class="sourceCode" id="cb367"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb367-1" data-line-number="1">dist &lt;-<span class="st"> </span><span class="kw">DistanceMatrix</span>(aln, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb367-2" data-line-number="2"><span class="kw">class</span>(dist)</a>
<a class="sourceLine" id="cb367-3" data-line-number="3"><span class="co">## [1] &quot;matrix&quot;</span></a>
<a class="sourceLine" id="cb367-4" data-line-number="4"><span class="kw">dim</span>(dist)</a>
<a class="sourceLine" id="cb367-5" data-line-number="5"><span class="co">## [1] 343 343</span></a>
<a class="sourceLine" id="cb367-6" data-line-number="6">dist[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb367-7" data-line-number="7"><span class="co">##              1            2           3            4            5</span></a>
<a class="sourceLine" id="cb367-8" data-line-number="8"><span class="co">## 1 0.0000000000 0.0007849294 0.001569859 0.0007849294 0.0007849294</span></a>
<a class="sourceLine" id="cb367-9" data-line-number="9"><span class="co">## 2 0.0007849294 0.0000000000 0.002354788 0.0015698587 0.0015698587</span></a>
<a class="sourceLine" id="cb367-10" data-line-number="10"><span class="co">## 3 0.0015698587 0.0023547881 0.000000000 0.0023547881 0.0023547881</span></a>
<a class="sourceLine" id="cb367-11" data-line-number="11"><span class="co">## 4 0.0007849294 0.0015698587 0.002354788 0.0000000000 0.0015698587</span></a>
<a class="sourceLine" id="cb367-12" data-line-number="12"><span class="co">## 5 0.0007849294 0.0015698587 0.002354788 0.0015698587 0.0000000000</span></a></code></pre></div></li>
<li><p>Use the distance matrix to cluster sequences using the ‘Neighbor Joining’ algorithm; visualize the result (see <code>?plot.dendrogram</code> for plotting options when the first argument is of class <code>dendrogram</code>)</p>
<div class="sourceCode" id="cb368"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb368-1" data-line-number="1">dendrogram &lt;-<span class="st"> </span><span class="kw">IdClusters</span>(dist, <span class="dt">method =</span> <span class="st">&quot;NJ&quot;</span>, <span class="dt">type =</span> <span class="st">&quot;dendrogram&quot;</span>)</a>
<a class="sourceLine" id="cb368-2" data-line-number="2"><span class="co">## ================================================================================</span></a>
<a class="sourceLine" id="cb368-3" data-line-number="3"><span class="co">## </span></a>
<a class="sourceLine" id="cb368-4" data-line-number="4"><span class="co">## Time difference of 0.61 secs</span></a>
<a class="sourceLine" id="cb368-5" data-line-number="5">dendrogram</a>
<a class="sourceLine" id="cb368-6" data-line-number="6"><span class="co">## &#39;dendrogram&#39; with 2 branches and 343 members total, at height 0.009898669</span></a>
<a class="sourceLine" id="cb368-7" data-line-number="7"><span class="kw">plot</span>(dendrogram, <span class="dt">leaflab =</span> <span class="st">&quot;none&quot;</span>)</a></code></pre></div>
<p><img src="05-Bioinformatics_files/figure-html/unnamed-chunk-56-1.png" width="672" /></p></li>
</ul>
<p>Save the tree to the working directory using the ‘Newick’ format</p>
<div class="sourceCode" id="cb369"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb369-1" data-line-number="1">dendrogram_file &lt;-<span class="st"> </span><span class="kw">file.path</span>(workdir, <span class="st">&quot;SARS-CoV2-S.newick&quot;</span>)</a>
<a class="sourceLine" id="cb369-2" data-line-number="2"><span class="kw">WriteDendrogram</span>(dendrogram, dendrogram_file, <span class="dt">quoteLabels =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
</div>
<div id="phylogenetic-tree-visualization" class="section level3">
<h3><span class="header-section-number">5.3.3</span> Phylogenetic tree visualization</h3>
<p>Setup</p>
<ul>
<li><p>Install (if necessary, using <code>BiocManager::install(&quot;ggtree&quot;)</code>) and attach the <a href="https://bioconductor.org/packages/ggtree">ggtree</a> package</p>
<div class="sourceCode" id="cb370"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb370-1" data-line-number="1"><span class="kw">library</span>(ggtree)</a>
<a class="sourceLine" id="cb370-2" data-line-number="2"></a>
<a class="sourceLine" id="cb370-3" data-line-number="3"><span class="kw">maintainer</span>(<span class="st">&quot;ggtree&quot;</span>)</a>
<a class="sourceLine" id="cb370-4" data-line-number="4"><span class="co">## [1] &quot;Guangchuang Yu &lt;guangchuangyu@gmail.com&gt;&quot;</span></a>
<a class="sourceLine" id="cb370-5" data-line-number="5"></a>
<a class="sourceLine" id="cb370-6" data-line-number="6"><span class="kw">citation</span>(<span class="st">&quot;ggtree&quot;</span>)</a>
<a class="sourceLine" id="cb370-7" data-line-number="7"><span class="co">## </span></a>
<a class="sourceLine" id="cb370-8" data-line-number="8"><span class="co">## To cite ggtree in publications use:</span></a>
<a class="sourceLine" id="cb370-9" data-line-number="9"><span class="co">## </span></a>
<a class="sourceLine" id="cb370-10" data-line-number="10"><span class="co">##   Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk</span></a>
<a class="sourceLine" id="cb370-11" data-line-number="11"><span class="co">##   Lam. ggtree: an R package for visualization and annotation of</span></a>
<a class="sourceLine" id="cb370-12" data-line-number="12"><span class="co">##   phylogenetic trees with their covariates and other associated data.</span></a>
<a class="sourceLine" id="cb370-13" data-line-number="13"><span class="co">##   Methods in Ecology and Evolution 2017, 8(1):28-36</span></a>
<a class="sourceLine" id="cb370-14" data-line-number="14"><span class="co">## </span></a>
<a class="sourceLine" id="cb370-15" data-line-number="15"><span class="co">##   Guangchuang Yu, Tommy Tsan-Yuk Lam, Huachen Zhu, Yi Guan. Two methods</span></a>
<a class="sourceLine" id="cb370-16" data-line-number="16"><span class="co">##   for mapping and visualizing associated data on phylogeny using</span></a>
<a class="sourceLine" id="cb370-17" data-line-number="17"><span class="co">##   ggtree. Molecular Biology and Evolution 2018, 35(2):3041-3043. doi:</span></a>
<a class="sourceLine" id="cb370-18" data-line-number="18"><span class="co">##   10.1093/molbev/msy194</span></a>
<a class="sourceLine" id="cb370-19" data-line-number="19"><span class="co">## </span></a>
<a class="sourceLine" id="cb370-20" data-line-number="20"><span class="co">## To see these entries in BibTeX format, use &#39;print(&lt;citation&gt;,</span></a>
<a class="sourceLine" id="cb370-21" data-line-number="21"><span class="co">## bibtex=TRUE)&#39;, &#39;toBibtex(.)&#39;, or set</span></a>
<a class="sourceLine" id="cb370-22" data-line-number="22"><span class="co">## &#39;options(citation.bibtex.max=999)&#39;.</span></a></code></pre></div></li>
<li><p>The visualizations this package provides are not well-suited to <em>R</em>’s help pages; check out the online <a href="https://yulab-smu.github.io/treedata-book">ggtree book</a> to orient yourself.</p></li>
</ul>
<p>Read the data from the Newick-format file created by DECIPHER</p>
<div class="sourceCode" id="cb371"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb371-1" data-line-number="1">dendrogram_file &lt;-<span class="st"> </span><span class="kw">file.path</span>(workdir, <span class="st">&quot;SARS-CoV2-S.newick&quot;</span>)</a>
<a class="sourceLine" id="cb371-2" data-line-number="2">tree &lt;-<span class="st"> </span>treeio<span class="op">::</span><span class="kw">read.newick</span>(dendrogram_file)</a>
<a class="sourceLine" id="cb371-3" data-line-number="3">tree</a>
<a class="sourceLine" id="cb371-4" data-line-number="4"><span class="co">## </span></a>
<a class="sourceLine" id="cb371-5" data-line-number="5"><span class="co">## Phylogenetic tree with 343 tips and 288 internal nodes.</span></a>
<a class="sourceLine" id="cb371-6" data-line-number="6"><span class="co">## </span></a>
<a class="sourceLine" id="cb371-7" data-line-number="7"><span class="co">## Tip labels:</span></a>
<a class="sourceLine" id="cb371-8" data-line-number="8"><span class="co">##  228, 267, 298, 294, 77, 189, ...</span></a>
<a class="sourceLine" id="cb371-9" data-line-number="9"><span class="co">## </span></a>
<a class="sourceLine" id="cb371-10" data-line-number="10"><span class="co">## Rooted; includes branch lengths.</span></a>
<a class="sourceLine" id="cb371-11" data-line-number="11">tibble<span class="op">::</span><span class="kw">as_tibble</span>(tree)</a>
<a class="sourceLine" id="cb371-12" data-line-number="12"><span class="co">## # A tibble: 631 x 4</span></a>
<a class="sourceLine" id="cb371-13" data-line-number="13"><span class="co">##    parent  node branch.length label</span></a>
<a class="sourceLine" id="cb371-14" data-line-number="14"><span class="co">##     &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt;</span></a>
<a class="sourceLine" id="cb371-15" data-line-number="15"><span class="co">##  1    346     1      0        228  </span></a>
<a class="sourceLine" id="cb371-16" data-line-number="16"><span class="co">##  2    346     2      0.00385  267  </span></a>
<a class="sourceLine" id="cb371-17" data-line-number="17"><span class="co">##  3    347     3      0        298  </span></a>
<a class="sourceLine" id="cb371-18" data-line-number="18"><span class="co">##  4    348     4      0        294  </span></a>
<a class="sourceLine" id="cb371-19" data-line-number="19"><span class="co">##  5    349     5      0        77   </span></a>
<a class="sourceLine" id="cb371-20" data-line-number="20"><span class="co">##  6    349     6      0.00600  189  </span></a>
<a class="sourceLine" id="cb371-21" data-line-number="21"><span class="co">##  7    350     7      0        295  </span></a>
<a class="sourceLine" id="cb371-22" data-line-number="22"><span class="co">##  8    351     8      0.000580 280  </span></a>
<a class="sourceLine" id="cb371-23" data-line-number="23"><span class="co">##  9    352     9      0        297  </span></a>
<a class="sourceLine" id="cb371-24" data-line-number="24"><span class="co">## 10    353    10      0        87   </span></a>
<a class="sourceLine" id="cb371-25" data-line-number="25"><span class="co">## # … with 621 more rows</span></a></code></pre></div>
<p>Plot the tree, adding tip labels to the longest branches</p>
<div class="sourceCode" id="cb372"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb372-1" data-line-number="1"><span class="kw">ggtree</span>(tree) <span class="op">+</span></a>
<a class="sourceLine" id="cb372-2" data-line-number="2"><span class="st">    </span><span class="kw">geom_tiplab</span>(</a>
<a class="sourceLine" id="cb372-3" data-line-number="3">        <span class="kw">aes</span>(</a>
<a class="sourceLine" id="cb372-4" data-line-number="4">            <span class="dt">subset =</span> branch.length <span class="op">&gt;</span><span class="st"> </span><span class="fl">.001</span>,</a>
<a class="sourceLine" id="cb372-5" data-line-number="5">            <span class="dt">label =</span> label</a>
<a class="sourceLine" id="cb372-6" data-line-number="6">        ),</a>
<a class="sourceLine" id="cb372-7" data-line-number="7">        <span class="dt">size =</span> <span class="dv">3</span></a>
<a class="sourceLine" id="cb372-8" data-line-number="8">    ) <span class="op">+</span></a>
<a class="sourceLine" id="cb372-9" data-line-number="9"><span class="st">    </span>ggplot2<span class="op">::</span><span class="kw">ggtitle</span>(<span class="st">&quot;SARS-CoV-2 S-locus phylogeny; genbank, 10 May 2020&quot;</span>)</a>
<a class="sourceLine" id="cb372-10" data-line-number="10"><span class="co">## Warning: `data_frame()` is deprecated as of tibble 1.1.0.</span></a>
<a class="sourceLine" id="cb372-11" data-line-number="11"><span class="co">## Please use `tibble()` instead.</span></a>
<a class="sourceLine" id="cb372-12" data-line-number="12"><span class="co">## This warning is displayed once every 8 hours.</span></a>
<a class="sourceLine" id="cb372-13" data-line-number="13"><span class="co">## Call `lifecycle::last_warnings()` to see where this warning was generated.</span></a></code></pre></div>
<p><img src="05-Bioinformatics_files/figure-html/unnamed-chunk-60-1.png" width="672" /></p>
<p>Color sequences by region</p>
<ul>
<li>Find nodes representing just a single sequence – ‘singletons’</li>
<li>Use GenBank country to identify which country these were sequenced from</li>
<li>Color lineages according to singleton country of origin</li>
</ul>
</div>
</div>
<div id="day-32-single-cell-expression-data" class="section level2">
<h2><span class="header-section-number">5.4</span> Day 32 Single-cell expression data</h2>
<p>This is a very preliminary foray into COVID-19 single-cell RNA-seq gene expression data. It is derived from a preprint by <a href="https://www.biorxiv.org/content/10.1101/2020.04.19.049254v2">Muus et al</a>, including a ‘Terra’ (NHGRI) cloud computing <a href="https://anvil.terra.bio/#workspaces/kco-incubator/COVID-19_cross_tissue_analysis">workspace</a>.</p>
<p>RNA-seq is a technology for measuring RNA expression levels as a proxy for gene expression. Single-cell RNA seq measures RNA expression levels at the single-cell resolution, sometimes in conjunction with imaging or other spatial information. Often, the overall pattern of expression of a single cell can be used to classify the cell to a particular type, e.g., epithelial.</p>
<p>Many interesting questions present themselves in the context of COVID-19. It would be interesting to investigate patterns of expression of human genes known to interact with the SARS-CoV-2 virus, especially in tissues likely to be involved in the introduction of the virus to the human body. One could imagine further studies identifying, e.g., genetic variants that influence susceptibility or response to infection.</p>
<p>Our goal is to do some preliminary visualization of scRNA seq at human genes whose product is known to interact with virus proteins. A much more comprehensive guide is available as <a href="https://osca.bioconductor.org/">Orchestrating Single Cell Analysis with <em>Bioconductor</em></a>.</p>
<div id="setup-2" class="section level3">
<h3><span class="header-section-number">5.4.1</span> Setup</h3>
<p>This module requires a recent version of <em>R</em> and <em>Bioconductor</em>. It also requires some advanced packages that may be difficult to install. Definitely it is necessary that</p>
<div class="sourceCode" id="cb373"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb373-1" data-line-number="1"><span class="kw">as.package_version</span>(R.version) <span class="op">&gt;=</span><span class="st"> &quot;3.6&quot;</span></a>
<a class="sourceLine" id="cb373-2" data-line-number="2"><span class="co">## [1] TRUE</span></a>
<a class="sourceLine" id="cb373-3" data-line-number="3">BiocManager<span class="op">::</span><span class="kw">version</span>() <span class="op">&gt;=</span><span class="st"> &quot;3.10&quot;</span></a>
<a class="sourceLine" id="cb373-4" data-line-number="4"><span class="co">## [1] TRUE</span></a></code></pre></div>
<p>If those conditions are met, load the following libraries (use <code>BiocManager::install(&lt;pkgs&gt;)</code> if necessary)</p>
<div class="sourceCode" id="cb374"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb374-1" data-line-number="1"><span class="kw">library</span>(SingleCellExperiment)</a>
<a class="sourceLine" id="cb374-2" data-line-number="2"><span class="kw">library</span>(Matrix)</a>
<a class="sourceLine" id="cb374-3" data-line-number="3"><span class="kw">library</span>(rhdf5)</a>
<a class="sourceLine" id="cb374-4" data-line-number="4"><span class="kw">library</span>(dplyr)</a>
<a class="sourceLine" id="cb374-5" data-line-number="5"><span class="kw">library</span>(hexbin)</a></code></pre></div>
<p>Create a working directory for today.</p>
<div class="sourceCode" id="cb375"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb375-1" data-line-number="1">workdir &lt;-<span class="st"> &quot;workdir&quot;</span></a>
<a class="sourceLine" id="cb375-2" data-line-number="2"><span class="cf">if</span> (<span class="op">!</span><span class="kw">dir.exists</span>(workdir))</a>
<a class="sourceLine" id="cb375-3" data-line-number="3">    <span class="kw">dir.create</span>(workdir)</a></code></pre></div>
<p>A couple of scripts provide helper functions for working with the data. Download the following file to the working directory</p>
<div class="sourceCode" id="cb376"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb376-1" data-line-number="1">url &lt;-<span class="st"> &quot;https://raw.githubusercontent.com/mtmorgan/QuaRantine/master/assets/05-scRNAseq-import.R&quot;</span></a>
<a class="sourceLine" id="cb376-2" data-line-number="2">destination &lt;-<span class="st"> </span><span class="kw">file.path</span>(workdir, <span class="kw">basename</span>(url))</a>
<a class="sourceLine" id="cb376-3" data-line-number="3"><span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(destination))</a>
<a class="sourceLine" id="cb376-4" data-line-number="4">    <span class="kw">download.file</span>(url, destination)</a></code></pre></div>
<p>Take a quick look at the downloaded file (e.g., in RStudio File -&gt; open) and verify that it is an <em>R</em> script. Source the script in to your <em>R</em> session so that you have access to the functions in the file.</p>
<div class="sourceCode" id="cb377"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb377-1" data-line-number="1"><span class="kw">source</span>(destination)</a></code></pre></div>
<p>Download our sample data set</p>
<div class="sourceCode" id="cb378"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb378-1" data-line-number="1">url &lt;-<span class="st"> &quot;https://github.com/mtmorgan/QuaRantine/raw/master/assets/05-internal_nonsmokerslung.h5ad&quot;</span></a>
<a class="sourceLine" id="cb378-2" data-line-number="2">destination &lt;-<span class="st"> </span><span class="kw">file.path</span>(workdir, <span class="kw">basename</span>(url))</a>
<a class="sourceLine" id="cb378-3" data-line-number="3"><span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(destination))</a>
<a class="sourceLine" id="cb378-4" data-line-number="4">    <span class="kw">download.file</span>(url, destination)</a></code></pre></div>
<p>The file is an ‘hdf5’ file format, used to store relatively large data sets in a binary format that allows for faster and more robust input than a CSV file. <em>R</em> and other software packages can read hdf5 files, but Excel, etc.</p>
</div>
<div id="data-input-and-exploration" class="section level3">
<h3><span class="header-section-number">5.4.2</span> Data input and exploration</h3>
<p>There are two main components to the data. The first describes each observation (cell).</p>
<div class="sourceCode" id="cb379"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb379-1" data-line-number="1">obs &lt;-<span class="st"> </span><span class="kw">read_obs</span>(destination)</a>
<a class="sourceLine" id="cb379-2" data-line-number="2">obs</a>
<a class="sourceLine" id="cb379-3" data-line-number="3"><span class="co">## # A tibble: 240,511 x 7</span></a>
<a class="sourceLine" id="cb379-4" data-line-number="4"><span class="co">##      Age `Cell Type` Count Gender Patient_ID Tissue              cell_subset    </span></a>
<a class="sourceLine" id="cb379-5" data-line-number="5"><span class="co">##    &lt;dbl&gt; &lt;fct&gt;       &lt;int&gt; &lt;fct&gt;  &lt;fct&gt;      &lt;fct&gt;               &lt;fct&gt;          </span></a>
<a class="sourceLine" id="cb379-6" data-line-number="6"><span class="co">##  1    57 Basal           1 F      HU28       internal_nonsmoker… Epithelial (ba…</span></a>
<a class="sourceLine" id="cb379-7" data-line-number="7"><span class="co">##  2    57 Basal           1 F      HU28       internal_nonsmoker… Epithelial (ba…</span></a>
<a class="sourceLine" id="cb379-8" data-line-number="8"><span class="co">##  3    57 Basal           1 F      HU28       internal_nonsmoker… Epithelial (ba…</span></a>
<a class="sourceLine" id="cb379-9" data-line-number="9"><span class="co">##  4    57 Basal           1 F      HU28       internal_nonsmoker… Epithelial (ba…</span></a>
<a class="sourceLine" id="cb379-10" data-line-number="10"><span class="co">##  5    57 Secretory       1 F      HU28       internal_nonsmoker… Secretory      </span></a>
<a class="sourceLine" id="cb379-11" data-line-number="11"><span class="co">##  6    57 Basal           1 F      HU28       internal_nonsmoker… Epithelial (ba…</span></a>
<a class="sourceLine" id="cb379-12" data-line-number="12"><span class="co">##  7    57 Basal           1 F      HU28       internal_nonsmoker… Epithelial (ba…</span></a>
<a class="sourceLine" id="cb379-13" data-line-number="13"><span class="co">##  8    57 Basal           1 F      HU28       internal_nonsmoker… Epithelial (ba…</span></a>
<a class="sourceLine" id="cb379-14" data-line-number="14"><span class="co">##  9    57 Basal           1 F      HU28       internal_nonsmoker… Epithelial (ba…</span></a>
<a class="sourceLine" id="cb379-15" data-line-number="15"><span class="co">## 10    57 Secretory       1 F      HU28       internal_nonsmoker… Secretory      </span></a>
<a class="sourceLine" id="cb379-16" data-line-number="16"><span class="co">## # … with 240,501 more rows</span></a></code></pre></div>
<p>If <em>R</em> complains that it can’t find <code>read_obs</code>, then remember to <code>source()</code> the script downloaded earlier.</p>
<p>This experiment measured gene expression in 240511 cells. Use standard dplyr commands to explore information about the cells</p>
<ul>
<li><p>How many patients are represented? (we need to use <code>dplyr::count()</code> because another package attached to our search path also has a <code>count()</code> function)</p>
<div class="sourceCode" id="cb380"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb380-1" data-line-number="1">obs <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">count</span>(Patient_ID)</a>
<a class="sourceLine" id="cb380-2" data-line-number="2"><span class="co">## # A tibble: 7 x 2</span></a>
<a class="sourceLine" id="cb380-3" data-line-number="3"><span class="co">##   Patient_ID     n</span></a>
<a class="sourceLine" id="cb380-4" data-line-number="4"><span class="co">##   &lt;fct&gt;      &lt;int&gt;</span></a>
<a class="sourceLine" id="cb380-5" data-line-number="5"><span class="co">## 1 HU28       12301</span></a>
<a class="sourceLine" id="cb380-6" data-line-number="6"><span class="co">## 2 HU30       33971</span></a>
<a class="sourceLine" id="cb380-7" data-line-number="7"><span class="co">## 3 HU37       66594</span></a>
<a class="sourceLine" id="cb380-8" data-line-number="8"><span class="co">## 4 HU39       29908</span></a>
<a class="sourceLine" id="cb380-9" data-line-number="9"><span class="co">## 5 HU49       56575</span></a>
<a class="sourceLine" id="cb380-10" data-line-number="10"><span class="co">## 6 HU52       17768</span></a>
<a class="sourceLine" id="cb380-11" data-line-number="11"><span class="co">## 7 HU62       23394</span></a></code></pre></div></li>
<li><p>What is the age and gender of each patient?</p>
<div class="sourceCode" id="cb381"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb381-1" data-line-number="1">obs <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">distinct</span>(Patient_ID, Age, Gender)</a>
<a class="sourceLine" id="cb381-2" data-line-number="2"><span class="co">## # A tibble: 7 x 3</span></a>
<a class="sourceLine" id="cb381-3" data-line-number="3"><span class="co">##   Patient_ID   Age Gender</span></a>
<a class="sourceLine" id="cb381-4" data-line-number="4"><span class="co">##   &lt;fct&gt;      &lt;dbl&gt; &lt;fct&gt; </span></a>
<a class="sourceLine" id="cb381-5" data-line-number="5"><span class="co">## 1 HU28        57   F     </span></a>
<a class="sourceLine" id="cb381-6" data-line-number="6"><span class="co">## 2 HU39         0.3 M     </span></a>
<a class="sourceLine" id="cb381-7" data-line-number="7"><span class="co">## 3 HU62        46   M     </span></a>
<a class="sourceLine" id="cb381-8" data-line-number="8"><span class="co">## 4 HU30        59   F     </span></a>
<a class="sourceLine" id="cb381-9" data-line-number="9"><span class="co">## 5 HU49        18   F     </span></a>
<a class="sourceLine" id="cb381-10" data-line-number="10"><span class="co">## 6 HU37        42   F     </span></a>
<a class="sourceLine" id="cb381-11" data-line-number="11"><span class="co">## 7 HU52        66   F</span></a></code></pre></div></li>
<li><p>Which cell types are represented (the column contains a space in its name, which causes problems for <em>R</em> unless the column is quoted)</p>
<div class="sourceCode" id="cb382"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb382-1" data-line-number="1">obs <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">count</span>(<span class="st">`</span><span class="dt">Cell Type</span><span class="st">`</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">arrange</span>(<span class="kw">desc</span>(n))</a>
<a class="sourceLine" id="cb382-2" data-line-number="2"><span class="co">## # A tibble: 17 x 2</span></a>
<a class="sourceLine" id="cb382-3" data-line-number="3"><span class="co">##    `Cell Type`        n</span></a>
<a class="sourceLine" id="cb382-4" data-line-number="4"><span class="co">##    &lt;fct&gt;          &lt;int&gt;</span></a>
<a class="sourceLine" id="cb382-5" data-line-number="5"><span class="co">##  1 Basal          67554</span></a>
<a class="sourceLine" id="cb382-6" data-line-number="6"><span class="co">##  2 AT2            28326</span></a>
<a class="sourceLine" id="cb382-7" data-line-number="7"><span class="co">##  3 Fibroblast     27574</span></a>
<a class="sourceLine" id="cb382-8" data-line-number="8"><span class="co">##  4 T.NK.cells     26709</span></a>
<a class="sourceLine" id="cb382-9" data-line-number="9"><span class="co">##  5 Myeloid        19730</span></a>
<a class="sourceLine" id="cb382-10" data-line-number="10"><span class="co">##  6 Secretory      18643</span></a>
<a class="sourceLine" id="cb382-11" data-line-number="11"><span class="co">##  7 AT1            11722</span></a>
<a class="sourceLine" id="cb382-12" data-line-number="12"><span class="co">##  8 Endothelial     9611</span></a>
<a class="sourceLine" id="cb382-13" data-line-number="13"><span class="co">##  9 Ciliated        9445</span></a>
<a class="sourceLine" id="cb382-14" data-line-number="14"><span class="co">## 10 SmoothMuscle    9392</span></a>
<a class="sourceLine" id="cb382-15" data-line-number="15"><span class="co">## 11 Pericytes       3385</span></a>
<a class="sourceLine" id="cb382-16" data-line-number="16"><span class="co">## 12 B.cells         3232</span></a>
<a class="sourceLine" id="cb382-17" data-line-number="17"><span class="co">## 13 Mesothelium     2179</span></a>
<a class="sourceLine" id="cb382-18" data-line-number="18"><span class="co">## 14 Mast            1991</span></a>
<a class="sourceLine" id="cb382-19" data-line-number="19"><span class="co">## 15 Tuft.like        677</span></a>
<a class="sourceLine" id="cb382-20" data-line-number="20"><span class="co">## 16 Neuroendocrine   179</span></a>
<a class="sourceLine" id="cb382-21" data-line-number="21"><span class="co">## 17 Ionocytes        162</span></a></code></pre></div></li>
<li><p>Each cell was classified to cell type based on overall gene expression. How many cells of each type were observed?</p>
<div class="sourceCode" id="cb383"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb383-1" data-line-number="1">obs <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">count</span>(cell_subset)</a>
<a class="sourceLine" id="cb383-2" data-line-number="2"><span class="co">## # A tibble: 17 x 2</span></a>
<a class="sourceLine" id="cb383-3" data-line-number="3"><span class="co">##    cell_subset                            n</span></a>
<a class="sourceLine" id="cb383-4" data-line-number="4"><span class="co">##    &lt;fct&gt;                              &lt;int&gt;</span></a>
<a class="sourceLine" id="cb383-5" data-line-number="5"><span class="co">##  1 Endothelial cell                    9611</span></a>
<a class="sourceLine" id="cb383-6" data-line-number="6"><span class="co">##  2 Epithelial (basal)                 67554</span></a>
<a class="sourceLine" id="cb383-7" data-line-number="7"><span class="co">##  3 Epithelial (ciliated)               9445</span></a>
<a class="sourceLine" id="cb383-8" data-line-number="8"><span class="co">##  4 Epithelial cell (alveolar type I)  11722</span></a>
<a class="sourceLine" id="cb383-9" data-line-number="9"><span class="co">##  5 Epithelial cell (alveolar type II) 28326</span></a>
<a class="sourceLine" id="cb383-10" data-line-number="10"><span class="co">##  6 Fibroblast                         27574</span></a>
<a class="sourceLine" id="cb383-11" data-line-number="11"><span class="co">##  7 Immune (B cell)                     3232</span></a>
<a class="sourceLine" id="cb383-12" data-line-number="12"><span class="co">##  8 Immune (NKT cell)                  26709</span></a>
<a class="sourceLine" id="cb383-13" data-line-number="13"><span class="co">##  9 Immune (mast cell)                  1991</span></a>
<a class="sourceLine" id="cb383-14" data-line-number="14"><span class="co">## 10 Immune (myeloid)                   19730</span></a>
<a class="sourceLine" id="cb383-15" data-line-number="15"><span class="co">## 11 Ionocyte                             162</span></a>
<a class="sourceLine" id="cb383-16" data-line-number="16"><span class="co">## 12 Mesothelium                         2179</span></a>
<a class="sourceLine" id="cb383-17" data-line-number="17"><span class="co">## 13 Neuroendocrine                       179</span></a>
<a class="sourceLine" id="cb383-18" data-line-number="18"><span class="co">## 14 Pericyte                            3385</span></a>
<a class="sourceLine" id="cb383-19" data-line-number="19"><span class="co">## 15 Secretory                          18643</span></a>
<a class="sourceLine" id="cb383-20" data-line-number="20"><span class="co">## 16 Smooth muscle cell                  9392</span></a>
<a class="sourceLine" id="cb383-21" data-line-number="21"><span class="co">## 17 Tuft                                 677</span></a></code></pre></div></li>
</ul>
<p>Single-cell RNA assays expression across all genes, but the current data set has been restricted to 22 genes relevant to virus interaction. Read in the expression levels of these 22 genes across all cells.</p>
<div class="sourceCode" id="cb384"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb384-1" data-line-number="1">x &lt;-<span class="st"> </span><span class="kw">read_expression</span>(destination)</a>
<a class="sourceLine" id="cb384-2" data-line-number="2"><span class="kw">dim</span>(x)</a>
<a class="sourceLine" id="cb384-3" data-line-number="3"><span class="co">## [1]     22 240511</span></a>
<a class="sourceLine" id="cb384-4" data-line-number="4"><span class="kw">class</span>(x)</a>
<a class="sourceLine" id="cb384-5" data-line-number="5"><span class="co">## [1] &quot;dgCMatrix&quot;</span></a>
<a class="sourceLine" id="cb384-6" data-line-number="6"><span class="co">## attr(,&quot;package&quot;)</span></a>
<a class="sourceLine" id="cb384-7" data-line-number="7"><span class="co">## [1] &quot;Matrix&quot;</span></a>
<a class="sourceLine" id="cb384-8" data-line-number="8"></a>
<a class="sourceLine" id="cb384-9" data-line-number="9">x[, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a>
<a class="sourceLine" id="cb384-10" data-line-number="10"><span class="co">## 22 x 3 sparse Matrix of class &quot;dgCMatrix&quot;</span></a>
<a class="sourceLine" id="cb384-11" data-line-number="11"><span class="co">##         1B_cell57_HU28 1B_cell418_HU28 1B_cell611_HU28</span></a>
<a class="sourceLine" id="cb384-12" data-line-number="12"><span class="co">## ACE2         .               .               .        </span></a>
<a class="sourceLine" id="cb384-13" data-line-number="13"><span class="co">## TMPRSS2      0.6002854       0.7688452       .        </span></a>
<a class="sourceLine" id="cb384-14" data-line-number="14"><span class="co">## IL6          .               .               .        </span></a>
<a class="sourceLine" id="cb384-15" data-line-number="15"><span class="co">## IL6R         .               .               .        </span></a>
<a class="sourceLine" id="cb384-16" data-line-number="16"><span class="co">## IL6ST        .               0.7688452       .        </span></a>
<a class="sourceLine" id="cb384-17" data-line-number="17"><span class="co">## PCSK1        .               .               .        </span></a>
<a class="sourceLine" id="cb384-18" data-line-number="18"><span class="co">## PCSK2        .               .               .        </span></a>
<a class="sourceLine" id="cb384-19" data-line-number="19"><span class="co">## FURIN        .               .               .        </span></a>
<a class="sourceLine" id="cb384-20" data-line-number="20"><span class="co">## PCSK4        .               .               .        </span></a>
<a class="sourceLine" id="cb384-21" data-line-number="21"><span class="co">## PCSK5        .               .               .        </span></a>
<a class="sourceLine" id="cb384-22" data-line-number="22"><span class="co">## PCSK6        .               .               .        </span></a>
<a class="sourceLine" id="cb384-23" data-line-number="23"><span class="co">## PCSK7        .               .               0.7382636</span></a>
<a class="sourceLine" id="cb384-24" data-line-number="24"><span class="co">## C1R          .               .               .        </span></a>
<a class="sourceLine" id="cb384-25" data-line-number="25"><span class="co">## C2           .               .               .        </span></a>
<a class="sourceLine" id="cb384-26" data-line-number="26"><span class="co">## C3           .               .               .        </span></a>
<a class="sourceLine" id="cb384-27" data-line-number="27"><span class="co">## C5           .               .               .        </span></a>
<a class="sourceLine" id="cb384-28" data-line-number="28"><span class="co">## CFI          .               .               .        </span></a>
<a class="sourceLine" id="cb384-29" data-line-number="29"><span class="co">## CTSS         0.9727762       0.7688452       .        </span></a>
<a class="sourceLine" id="cb384-30" data-line-number="30"><span class="co">## CTSL         .               .               .        </span></a>
<a class="sourceLine" id="cb384-31" data-line-number="31"><span class="co">## CTSB         1.2435541       1.1983210       .        </span></a>
<a class="sourceLine" id="cb384-32" data-line-number="32"><span class="co">## CTSC         .               0.7688452       .        </span></a>
<a class="sourceLine" id="cb384-33" data-line-number="33"><span class="co">## CTSE         .               .               .</span></a></code></pre></div>
<p>This is a matrix of 22 x 240511 values, one entry for each gene and cell. However, single-cell assays are ‘sparse’, with a large number of zeros. To take advantage of this sparse data, <em>R</em> has represented the matrix as a special <em>sparse</em> matrix. The <code>.</code> in the display above represent zeros; the object <code>x</code> only contains information about the non-zero values.</p>
<p>How sparse is the data, i.e., what is the fraction of zeros in the data?</p>
<ul>
<li><p>Total number of elements</p>
<div class="sourceCode" id="cb385"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb385-1" data-line-number="1"><span class="kw">nrow</span>(x) <span class="op">*</span><span class="st"> </span><span class="kw">ncol</span>(x)</a>
<a class="sourceLine" id="cb385-2" data-line-number="2"><span class="co">## [1] 5291242</span></a></code></pre></div></li>
<li><p>Total number of zero elements (<code>x == 0</code> returns a logical vector of <code>TRUE</code> and <code>FALSE</code> values; <code>sum()</code> treats <code>TRUE</code> as 1, and <code>FALSE</code> as 0, so <code>sum(x -- 0)</code> is the number of zero elements)</p>
<div class="sourceCode" id="cb386"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb386-1" data-line-number="1"><span class="kw">sum</span>(x <span class="op">==</span><span class="st"> </span><span class="dv">0</span>)</a>
<a class="sourceLine" id="cb386-2" data-line-number="2"><span class="co">## [1] 4321307</span></a></code></pre></div></li>
<li><p>Fraction of zero elements</p>
<div class="sourceCode" id="cb387"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb387-1" data-line-number="1"><span class="kw">sum</span>(x <span class="op">==</span><span class="st"> </span><span class="dv">0</span>) <span class="op">/</span><span class="st"> </span>(<span class="kw">nrow</span>(x) <span class="op">*</span><span class="st"> </span><span class="kw">ncol</span>(x))</a>
<a class="sourceLine" id="cb387-2" data-line-number="2"><span class="co">## [1] 0.8166905</span></a></code></pre></div></li>
<li><p>More than 80% of the values are zeros! This primarily represents limitations of the technology, where relatively few ‘reads’ are available from each cell, so only a fraction of the genes being expressed are detected.</p></li>
<li><p>Use the function <code>rowMeans()</code> to calculate average expression of each gene; present this as a tibble</p>
<div class="sourceCode" id="cb388"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb388-1" data-line-number="1"><span class="kw">tibble</span>(</a>
<a class="sourceLine" id="cb388-2" data-line-number="2">    <span class="dt">gene =</span> <span class="kw">rownames</span>(x),</a>
<a class="sourceLine" id="cb388-3" data-line-number="3">    <span class="dt">average_expression =</span> <span class="kw">rowMeans</span>(x)</a>
<a class="sourceLine" id="cb388-4" data-line-number="4">)</a>
<a class="sourceLine" id="cb388-5" data-line-number="5"><span class="co">## # A tibble: 22 x 2</span></a>
<a class="sourceLine" id="cb388-6" data-line-number="6"><span class="co">##    gene    average_expression</span></a>
<a class="sourceLine" id="cb388-7" data-line-number="7"><span class="co">##    &lt;chr&gt;                &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb388-8" data-line-number="8"><span class="co">##  1 ACE2               0.00539</span></a>
<a class="sourceLine" id="cb388-9" data-line-number="9"><span class="co">##  2 TMPRSS2            0.0877 </span></a>
<a class="sourceLine" id="cb388-10" data-line-number="10"><span class="co">##  3 IL6                0.144  </span></a>
<a class="sourceLine" id="cb388-11" data-line-number="11"><span class="co">##  4 IL6R               0.0923 </span></a>
<a class="sourceLine" id="cb388-12" data-line-number="12"><span class="co">##  5 IL6ST              0.386  </span></a>
<a class="sourceLine" id="cb388-13" data-line-number="13"><span class="co">##  6 PCSK1              0.00302</span></a>
<a class="sourceLine" id="cb388-14" data-line-number="14"><span class="co">##  7 PCSK2              0.0117 </span></a>
<a class="sourceLine" id="cb388-15" data-line-number="15"><span class="co">##  8 FURIN              0.0908 </span></a>
<a class="sourceLine" id="cb388-16" data-line-number="16"><span class="co">##  9 PCSK4              0.00728</span></a>
<a class="sourceLine" id="cb388-17" data-line-number="17"><span class="co">## 10 PCSK5              0.0740 </span></a>
<a class="sourceLine" id="cb388-18" data-line-number="18"><span class="co">## # … with 12 more rows</span></a></code></pre></div></li>
</ul>
</div>
<div id="coordinating-cell-and-expression-data" class="section level3">
<h3><span class="header-section-number">5.4.3</span> Coordinating cell and expression data</h3>
<p>Notice that we have two separate data structures – a large matrix of expression values, and a tibble describing each cell. It seems error-prone to manage these data separately, e.g., what if we remove some rows from the cell annotations without removing the corresponding columns from the expression matrix?</p>
<p>A better strategy is to introduce an object that can manage both parts of the data in a coordinated way. This container is provided by the <a href="https://bioconductor.org/packages/SingleCellExperiment">SingleCellExperiment</a> package</p>
<div class="sourceCode" id="cb389"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb389-1" data-line-number="1"><span class="kw">library</span>(SingleCellExperiment)</a></code></pre></div>
<p>Use the <em>constructor</em> <code>SingleCellExperiment()</code> to create this object. The basic metaphor is that a <code>SingleCellExperiment</code> is like a matrix, but with column (and row) annotations. The column annotations are called <code>colData</code>.</p>
<div class="sourceCode" id="cb390"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb390-1" data-line-number="1">knitr<span class="op">::</span><span class="kw">include_graphics</span>(<span class="st">&#39;images/SummarizedExperiment.png&#39;</span>)</a></code></pre></div>
<p><img src="images/SummarizedExperiment.png" width="60%" /></p>
<p>Create and view a <code>SingleCellExperiment</code></p>
<div class="sourceCode" id="cb391"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb391-1" data-line-number="1">sce &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(x, <span class="dt">colData =</span> obs)</a>
<a class="sourceLine" id="cb391-2" data-line-number="2">sce</a>
<a class="sourceLine" id="cb391-3" data-line-number="3"><span class="co">## class: SingleCellExperiment </span></a>
<a class="sourceLine" id="cb391-4" data-line-number="4"><span class="co">## dim: 22 240511 </span></a>
<a class="sourceLine" id="cb391-5" data-line-number="5"><span class="co">## metadata(0):</span></a>
<a class="sourceLine" id="cb391-6" data-line-number="6"><span class="co">## assays(1): &#39;&#39;</span></a>
<a class="sourceLine" id="cb391-7" data-line-number="7"><span class="co">## rownames(22): ACE2 TMPRSS2 ... CTSC CTSE</span></a>
<a class="sourceLine" id="cb391-8" data-line-number="8"><span class="co">## rowData names(0):</span></a>
<a class="sourceLine" id="cb391-9" data-line-number="9"><span class="co">## colnames(240511): 1B_cell57_HU28 1B_cell418_HU28 ...</span></a>
<a class="sourceLine" id="cb391-10" data-line-number="10"><span class="co">##   RP2_neg_cell6794209_HU52 RP2_neg_cell6794225_HU52</span></a>
<a class="sourceLine" id="cb391-11" data-line-number="11"><span class="co">## colData names(7): Age Cell Type ... Tissue cell_subset</span></a>
<a class="sourceLine" id="cb391-12" data-line-number="12"><span class="co">## reducedDimNames(0):</span></a>
<a class="sourceLine" id="cb391-13" data-line-number="13"><span class="co">## spikeNames(0):</span></a>
<a class="sourceLine" id="cb391-14" data-line-number="14"><span class="co">## altExpNames(0):</span></a>
<a class="sourceLine" id="cb391-15" data-line-number="15"></a>
<a class="sourceLine" id="cb391-16" data-line-number="16"><span class="kw">dim</span>(sce)</a>
<a class="sourceLine" id="cb391-17" data-line-number="17"><span class="co">## [1]     22 240511</span></a></code></pre></div>
<p>Use <code>subset()</code> to select particular rows and columns. The subset operates in a coordinated fashion on both the cell annotations and expression data, so there is no risk of ‘book-keeping’ error.</p>
<div class="sourceCode" id="cb392"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb392-1" data-line-number="1">my_sce &lt;-</a>
<a class="sourceLine" id="cb392-2" data-line-number="2"><span class="st">    </span>sce <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb392-3" data-line-number="3"><span class="st">    </span><span class="kw">subset</span>(</a>
<a class="sourceLine" id="cb392-4" data-line-number="4">        <span class="kw">rownames</span>(.) <span class="op">%in%</span><span class="st"> &quot;ACE2&quot;</span>,</a>
<a class="sourceLine" id="cb392-5" data-line-number="5">        <span class="st">`</span><span class="dt">Cell Type</span><span class="st">`</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Fibroblast&quot;</span>, <span class="st">&quot;SmoothMuscle&quot;</span>)</a>
<a class="sourceLine" id="cb392-6" data-line-number="6">    )</a>
<a class="sourceLine" id="cb392-7" data-line-number="7"></a>
<a class="sourceLine" id="cb392-8" data-line-number="8">my_sce</a>
<a class="sourceLine" id="cb392-9" data-line-number="9"><span class="co">## class: SingleCellExperiment </span></a>
<a class="sourceLine" id="cb392-10" data-line-number="10"><span class="co">## dim: 1 36966 </span></a>
<a class="sourceLine" id="cb392-11" data-line-number="11"><span class="co">## metadata(0):</span></a>
<a class="sourceLine" id="cb392-12" data-line-number="12"><span class="co">## assays(1): &#39;&#39;</span></a>
<a class="sourceLine" id="cb392-13" data-line-number="13"><span class="co">## rownames(1): ACE2</span></a>
<a class="sourceLine" id="cb392-14" data-line-number="14"><span class="co">## rowData names(0):</span></a>
<a class="sourceLine" id="cb392-15" data-line-number="15"><span class="co">## colnames(36966): 1B_cell31801_HU28 1B_cell48220_HU28 ...</span></a>
<a class="sourceLine" id="cb392-16" data-line-number="16"><span class="co">##   RP2_neg_cell6778193_HU52 RP2_neg_cell6785080_HU52</span></a>
<a class="sourceLine" id="cb392-17" data-line-number="17"><span class="co">## colData names(7): Age Cell Type ... Tissue cell_subset</span></a>
<a class="sourceLine" id="cb392-18" data-line-number="18"><span class="co">## reducedDimNames(0):</span></a>
<a class="sourceLine" id="cb392-19" data-line-number="19"><span class="co">## spikeNames(0):</span></a>
<a class="sourceLine" id="cb392-20" data-line-number="20"><span class="co">## altExpNames(0):</span></a>
<a class="sourceLine" id="cb392-21" data-line-number="21"></a>
<a class="sourceLine" id="cb392-22" data-line-number="22"><span class="kw">dim</span>(my_sce)</a>
<a class="sourceLine" id="cb392-23" data-line-number="23"><span class="co">## [1]     1 36966</span></a></code></pre></div>
<p>For the exercises below, focus on the ACE2 gene. ACE2 is a gene on the outer surface of the cell membrane in lung and other tissue.</p>
<div class="sourceCode" id="cb393"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb393-1" data-line-number="1">ace2 &lt;-<span class="st"> </span><span class="kw">subset</span>(sce, <span class="kw">rownames</span>(sce) <span class="op">%in%</span><span class="st"> &quot;ACE2&quot;</span>)</a>
<a class="sourceLine" id="cb393-2" data-line-number="2">ace2</a>
<a class="sourceLine" id="cb393-3" data-line-number="3"><span class="co">## class: SingleCellExperiment </span></a>
<a class="sourceLine" id="cb393-4" data-line-number="4"><span class="co">## dim: 1 240511 </span></a>
<a class="sourceLine" id="cb393-5" data-line-number="5"><span class="co">## metadata(0):</span></a>
<a class="sourceLine" id="cb393-6" data-line-number="6"><span class="co">## assays(1): &#39;&#39;</span></a>
<a class="sourceLine" id="cb393-7" data-line-number="7"><span class="co">## rownames(1): ACE2</span></a>
<a class="sourceLine" id="cb393-8" data-line-number="8"><span class="co">## rowData names(0):</span></a>
<a class="sourceLine" id="cb393-9" data-line-number="9"><span class="co">## colnames(240511): 1B_cell57_HU28 1B_cell418_HU28 ...</span></a>
<a class="sourceLine" id="cb393-10" data-line-number="10"><span class="co">##   RP2_neg_cell6794209_HU52 RP2_neg_cell6794225_HU52</span></a>
<a class="sourceLine" id="cb393-11" data-line-number="11"><span class="co">## colData names(7): Age Cell Type ... Tissue cell_subset</span></a>
<a class="sourceLine" id="cb393-12" data-line-number="12"><span class="co">## reducedDimNames(0):</span></a>
<a class="sourceLine" id="cb393-13" data-line-number="13"><span class="co">## spikeNames(0):</span></a>
<a class="sourceLine" id="cb393-14" data-line-number="14"><span class="co">## altExpNames(0):</span></a></code></pre></div>
<p>We will perform some simple visualization, as we start to explore the data. Create a tibble containing the <code>cell_subset</code> annotation and the expression values of the ACE2 gene. We make use of <code>$</code> to access columns of the cell annotation (<code>colData</code>) and <code>assay()</code> to access the expression data. <code>as.vector()</code> changes the one-dimensional array returned by <code>assay()</code> into a numeric vector.</p>
<div class="sourceCode" id="cb394"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb394-1" data-line-number="1">tbl &lt;-<span class="st"> </span><span class="kw">tibble</span>(</a>
<a class="sourceLine" id="cb394-2" data-line-number="2">    <span class="dt">cell_subset =</span> ace2<span class="op">$</span>cell_subset,</a>
<a class="sourceLine" id="cb394-3" data-line-number="3">    <span class="dt">expression =</span> <span class="kw">as.vector</span>(<span class="kw">assay</span>(ace2))</a>
<a class="sourceLine" id="cb394-4" data-line-number="4">)</a>
<a class="sourceLine" id="cb394-5" data-line-number="5">tbl</a>
<a class="sourceLine" id="cb394-6" data-line-number="6"><span class="co">## # A tibble: 240,511 x 2</span></a>
<a class="sourceLine" id="cb394-7" data-line-number="7"><span class="co">##    cell_subset        expression</span></a>
<a class="sourceLine" id="cb394-8" data-line-number="8"><span class="co">##    &lt;fct&gt;                   &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb394-9" data-line-number="9"><span class="co">##  1 Epithelial (basal)          0</span></a>
<a class="sourceLine" id="cb394-10" data-line-number="10"><span class="co">##  2 Epithelial (basal)          0</span></a>
<a class="sourceLine" id="cb394-11" data-line-number="11"><span class="co">##  3 Epithelial (basal)          0</span></a>
<a class="sourceLine" id="cb394-12" data-line-number="12"><span class="co">##  4 Epithelial (basal)          0</span></a>
<a class="sourceLine" id="cb394-13" data-line-number="13"><span class="co">##  5 Secretory                   0</span></a>
<a class="sourceLine" id="cb394-14" data-line-number="14"><span class="co">##  6 Epithelial (basal)          0</span></a>
<a class="sourceLine" id="cb394-15" data-line-number="15"><span class="co">##  7 Epithelial (basal)          0</span></a>
<a class="sourceLine" id="cb394-16" data-line-number="16"><span class="co">##  8 Epithelial (basal)          0</span></a>
<a class="sourceLine" id="cb394-17" data-line-number="17"><span class="co">##  9 Epithelial (basal)          0</span></a>
<a class="sourceLine" id="cb394-18" data-line-number="18"><span class="co">## 10 Secretory                   0</span></a>
<a class="sourceLine" id="cb394-19" data-line-number="19"><span class="co">## # … with 240,501 more rows</span></a></code></pre></div>
<p>Summarize the number of cells, non-zero expression values, and average expression of ACE2 for each cell subset.</p>
<div class="sourceCode" id="cb395"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb395-1" data-line-number="1">tbl <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb395-2" data-line-number="2"><span class="st">    </span><span class="kw">group_by</span>(cell_subset) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb395-3" data-line-number="3"><span class="st">    </span><span class="kw">summarize</span>(</a>
<a class="sourceLine" id="cb395-4" data-line-number="4">        <span class="dt">n =</span> <span class="kw">n</span>(),</a>
<a class="sourceLine" id="cb395-5" data-line-number="5">        <span class="dt">n_non_zero =</span> <span class="kw">sum</span>(expression <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>),</a>
<a class="sourceLine" id="cb395-6" data-line-number="6">        <span class="dt">mean_expression =</span> <span class="kw">mean</span>(expression)</a>
<a class="sourceLine" id="cb395-7" data-line-number="7">    )</a>
<a class="sourceLine" id="cb395-8" data-line-number="8"><span class="co">## # A tibble: 17 x 4</span></a>
<a class="sourceLine" id="cb395-9" data-line-number="9"><span class="co">##    cell_subset                            n n_non_zero mean_expression</span></a>
<a class="sourceLine" id="cb395-10" data-line-number="10"><span class="co">##    &lt;fct&gt;                              &lt;int&gt;      &lt;int&gt;           &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb395-11" data-line-number="11"><span class="co">##  1 Endothelial cell                    9611          2       0.0000691</span></a>
<a class="sourceLine" id="cb395-12" data-line-number="12"><span class="co">##  2 Epithelial (basal)                 67554        854       0.00726  </span></a>
<a class="sourceLine" id="cb395-13" data-line-number="13"><span class="co">##  3 Epithelial (ciliated)               9445        229       0.0167   </span></a>
<a class="sourceLine" id="cb395-14" data-line-number="14"><span class="co">##  4 Epithelial cell (alveolar type I)  11722         67       0.00341  </span></a>
<a class="sourceLine" id="cb395-15" data-line-number="15"><span class="co">##  5 Epithelial cell (alveolar type II) 28326        594       0.00967  </span></a>
<a class="sourceLine" id="cb395-16" data-line-number="16"><span class="co">##  6 Fibroblast                         27574         25       0.000648 </span></a>
<a class="sourceLine" id="cb395-17" data-line-number="17"><span class="co">##  7 Immune (B cell)                     3232          1       0.000226 </span></a>
<a class="sourceLine" id="cb395-18" data-line-number="18"><span class="co">##  8 Immune (NKT cell)                  26709          5       0.000270 </span></a>
<a class="sourceLine" id="cb395-19" data-line-number="19"><span class="co">##  9 Immune (mast cell)                  1991          0       0        </span></a>
<a class="sourceLine" id="cb395-20" data-line-number="20"><span class="co">## 10 Immune (myeloid)                   19730          2       0.0000694</span></a>
<a class="sourceLine" id="cb395-21" data-line-number="21"><span class="co">## 11 Ionocyte                             162          0       0        </span></a>
<a class="sourceLine" id="cb395-22" data-line-number="22"><span class="co">## 12 Mesothelium                         2179          2       0.000418 </span></a>
<a class="sourceLine" id="cb395-23" data-line-number="23"><span class="co">## 13 Neuroendocrine                       179          0       0        </span></a>
<a class="sourceLine" id="cb395-24" data-line-number="24"><span class="co">## 14 Pericyte                            3385         23       0.00765  </span></a>
<a class="sourceLine" id="cb395-25" data-line-number="25"><span class="co">## 15 Secretory                          18643        533       0.0146   </span></a>
<a class="sourceLine" id="cb395-26" data-line-number="26"><span class="co">## 16 Smooth muscle cell                  9392          7       0.000642 </span></a>
<a class="sourceLine" id="cb395-27" data-line-number="27"><span class="co">## 17 Tuft                                 677          4       0.00291</span></a></code></pre></div>
<p>Pictures speak louder than words. Use the <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a> package</p>
<div class="sourceCode" id="cb396"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb396-1" data-line-number="1"><span class="kw">library</span>(ggplot2)</a></code></pre></div>
<p>to visualize as a horizontal boxplot the ACE2 expression values of each cell subtype. I had initially used <code>geom_boxplot()</code> without <code>coord_flip()</code>, but the cell subtype labels overlapped. I experimented with rotating the labels, but that just made them difficult to read. So <code>coord_flip()</code> makes the boxplot horizontal rather than vertical, the labels are easy to read, and the figure easy to interpret.</p>
<div class="sourceCode" id="cb397"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb397-1" data-line-number="1">tbl <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb397-2" data-line-number="2"><span class="st">    </span><span class="kw">filter</span>(expression <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb397-3" data-line-number="3"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(cell_subset, expression)) <span class="op">+</span></a>
<a class="sourceLine" id="cb397-4" data-line-number="4"><span class="st">        </span><span class="kw">geom_boxplot</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb397-5" data-line-number="5"><span class="st">        </span><span class="kw">coord_flip</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb397-6" data-line-number="6"><span class="st">        </span><span class="kw">ggtitle</span>(<span class="st">&quot;ACE2&quot;</span>)</a></code></pre></div>
<p><img src="05-Bioinformatics_files/figure-html/unnamed-chunk-85-1.png" width="672" /></p>
<p>I want to explore co-expression of two genes, ACE2 and TMPRSS2. I used <code>subset()</code> to select the relevant expression and cell annotation values, then created a tibble with gene, cell, and expression columns.</p>
<ul>
<li><p>The <code>rep()</code> function replicates each element of its first argument the number of times indicated by it’s second argument; explore it’s behavior with simple examples, e.g.,</p>
<div class="sourceCode" id="cb398"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb398-1" data-line-number="1"><span class="dv">1</span><span class="op">:</span><span class="dv">3</span></a>
<a class="sourceLine" id="cb398-2" data-line-number="2"><span class="co">## [1] 1 2 3</span></a>
<a class="sourceLine" id="cb398-3" data-line-number="3"></a>
<a class="sourceLine" id="cb398-4" data-line-number="4"><span class="kw">rep</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb398-5" data-line-number="5"><span class="co">## [1] 1 2 3 1 2 3</span></a>
<a class="sourceLine" id="cb398-6" data-line-number="6"></a>
<a class="sourceLine" id="cb398-7" data-line-number="7"><span class="kw">rep</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="dt">each =</span> <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb398-8" data-line-number="8"><span class="co">## [1] 1 1 2 2 3 3</span></a></code></pre></div></li>
<li><p>When a two-dimensional array is coerced to a vector, it is ‘flattened’ by appending columns</p>
<div class="sourceCode" id="cb399"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb399-1" data-line-number="1">m &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">12</span>, <span class="dt">nrow =</span> <span class="dv">4</span>)</a>
<a class="sourceLine" id="cb399-2" data-line-number="2">m</a>
<a class="sourceLine" id="cb399-3" data-line-number="3"><span class="co">##      [,1] [,2] [,3]</span></a>
<a class="sourceLine" id="cb399-4" data-line-number="4"><span class="co">## [1,]    1    5    9</span></a>
<a class="sourceLine" id="cb399-5" data-line-number="5"><span class="co">## [2,]    2    6   10</span></a>
<a class="sourceLine" id="cb399-6" data-line-number="6"><span class="co">## [3,]    3    7   11</span></a>
<a class="sourceLine" id="cb399-7" data-line-number="7"><span class="co">## [4,]    4    8   12</span></a>
<a class="sourceLine" id="cb399-8" data-line-number="8"></a>
<a class="sourceLine" id="cb399-9" data-line-number="9"><span class="kw">as.vector</span>(m)</a>
<a class="sourceLine" id="cb399-10" data-line-number="10"><span class="co">##  [1]  1  2  3  4  5  6  7  8  9 10 11 12</span></a></code></pre></div></li>
<li><p>Putting this together</p>
<div class="sourceCode" id="cb400"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb400-1" data-line-number="1">two_genes &lt;-<span class="st"> </span><span class="kw">subset</span>(sce, <span class="kw">rownames</span>(sce) <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;ACE2&quot;</span>, <span class="st">&quot;TMPRSS2&quot;</span>))</a>
<a class="sourceLine" id="cb400-2" data-line-number="2">tbl &lt;-<span class="st"> </span><span class="kw">tibble</span>(</a>
<a class="sourceLine" id="cb400-3" data-line-number="3">    <span class="dt">gene =</span> <span class="kw">rep</span>(<span class="kw">rownames</span>(two_genes), <span class="kw">ncol</span>(two_genes)),</a>
<a class="sourceLine" id="cb400-4" data-line-number="4">    <span class="dt">cell =</span> <span class="kw">rep</span>(<span class="kw">colnames</span>(two_genes), <span class="dt">each =</span> <span class="kw">nrow</span>(two_genes)),</a>
<a class="sourceLine" id="cb400-5" data-line-number="5">    <span class="dt">expression =</span> <span class="kw">as.vector</span>(<span class="kw">assay</span>(two_genes))</a>
<a class="sourceLine" id="cb400-6" data-line-number="6">)</a>
<a class="sourceLine" id="cb400-7" data-line-number="7">tbl</a>
<a class="sourceLine" id="cb400-8" data-line-number="8"><span class="co">## # A tibble: 481,022 x 3</span></a>
<a class="sourceLine" id="cb400-9" data-line-number="9"><span class="co">##    gene    cell            expression</span></a>
<a class="sourceLine" id="cb400-10" data-line-number="10"><span class="co">##    &lt;chr&gt;   &lt;chr&gt;                &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb400-11" data-line-number="11"><span class="co">##  1 ACE2    1B_cell57_HU28       0    </span></a>
<a class="sourceLine" id="cb400-12" data-line-number="12"><span class="co">##  2 TMPRSS2 1B_cell57_HU28       0.600</span></a>
<a class="sourceLine" id="cb400-13" data-line-number="13"><span class="co">##  3 ACE2    1B_cell418_HU28      0    </span></a>
<a class="sourceLine" id="cb400-14" data-line-number="14"><span class="co">##  4 TMPRSS2 1B_cell418_HU28      0.769</span></a>
<a class="sourceLine" id="cb400-15" data-line-number="15"><span class="co">##  5 ACE2    1B_cell611_HU28      0    </span></a>
<a class="sourceLine" id="cb400-16" data-line-number="16"><span class="co">##  6 TMPRSS2 1B_cell611_HU28      0    </span></a>
<a class="sourceLine" id="cb400-17" data-line-number="17"><span class="co">##  7 ACE2    1B_cell617_HU28      0    </span></a>
<a class="sourceLine" id="cb400-18" data-line-number="18"><span class="co">##  8 TMPRSS2 1B_cell617_HU28      0    </span></a>
<a class="sourceLine" id="cb400-19" data-line-number="19"><span class="co">##  9 ACE2    1B_cell885_HU28      0    </span></a>
<a class="sourceLine" id="cb400-20" data-line-number="20"><span class="co">## 10 TMPRSS2 1B_cell885_HU28      1.09 </span></a>
<a class="sourceLine" id="cb400-21" data-line-number="21"><span class="co">## # … with 481,012 more rows</span></a></code></pre></div></li>
</ul>
<p>I’d like to plot the expression of one gene against the expression of the other. It is convenient to transform the tidy tibble we just created to have a separate column for each gene. Use <code>tidyr::pivot_wider()</code> for this operation</p>
<div class="sourceCode" id="cb401"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb401-1" data-line-number="1">wider_tbl &lt;-</a>
<a class="sourceLine" id="cb401-2" data-line-number="2"><span class="st">    </span>tbl <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb401-3" data-line-number="3"><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">expression =</span> <span class="kw">log</span>(expression)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb401-4" data-line-number="4"><span class="st">    </span>tidyr<span class="op">::</span><span class="kw">pivot_wider</span>(<span class="dt">names_from=</span>gene, <span class="dt">values_from=</span>expression)</a>
<a class="sourceLine" id="cb401-5" data-line-number="5"></a>
<a class="sourceLine" id="cb401-6" data-line-number="6">wider_tbl</a>
<a class="sourceLine" id="cb401-7" data-line-number="7"><span class="co">## # A tibble: 240,511 x 3</span></a>
<a class="sourceLine" id="cb401-8" data-line-number="8"><span class="co">##    cell              ACE2   TMPRSS2</span></a>
<a class="sourceLine" id="cb401-9" data-line-number="9"><span class="co">##    &lt;chr&gt;            &lt;dbl&gt;     &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb401-10" data-line-number="10"><span class="co">##  1 1B_cell57_HU28    -Inf   -0.510 </span></a>
<a class="sourceLine" id="cb401-11" data-line-number="11"><span class="co">##  2 1B_cell418_HU28   -Inf   -0.263 </span></a>
<a class="sourceLine" id="cb401-12" data-line-number="12"><span class="co">##  3 1B_cell611_HU28   -Inf -Inf     </span></a>
<a class="sourceLine" id="cb401-13" data-line-number="13"><span class="co">##  4 1B_cell617_HU28   -Inf -Inf     </span></a>
<a class="sourceLine" id="cb401-14" data-line-number="14"><span class="co">##  5 1B_cell885_HU28   -Inf    0.0879</span></a>
<a class="sourceLine" id="cb401-15" data-line-number="15"><span class="co">##  6 1B_cell968_HU28   -Inf -Inf     </span></a>
<a class="sourceLine" id="cb401-16" data-line-number="16"><span class="co">##  7 1B_cell1008_HU28  -Inf -Inf     </span></a>
<a class="sourceLine" id="cb401-17" data-line-number="17"><span class="co">##  8 1B_cell1060_HU28  -Inf   -0.300 </span></a>
<a class="sourceLine" id="cb401-18" data-line-number="18"><span class="co">##  9 1B_cell1156_HU28  -Inf   -0.675 </span></a>
<a class="sourceLine" id="cb401-19" data-line-number="19"><span class="co">## 10 1B_cell1370_HU28  -Inf -Inf     </span></a>
<a class="sourceLine" id="cb401-20" data-line-number="20"><span class="co">## # … with 240,501 more rows</span></a></code></pre></div>
<p>We’re ready to plot expression values of TMPRSS2 against ACE2. If we were to use <code>geom_point()</code>, many points would over-plot and we would have a difficult time seeing overall pattern. So instead we use <code>geom_hex()</code> to divide the plot area into hexagonal regions, and to fill each region with a color that indicates the number of points in the region. <code>coord_fix()</code> ensures that the x- and y- axes have the same scale, and the plot is printed so that the ‘aspect ratio’ (height / width) is 1.</p>
<div class="sourceCode" id="cb402"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb402-1" data-line-number="1">wider_tbl <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb402-2" data-line-number="2"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(ACE2, TMPRSS2)) <span class="op">+</span></a>
<a class="sourceLine" id="cb402-3" data-line-number="3"><span class="st">    </span><span class="kw">geom_hex</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb402-4" data-line-number="4"><span class="st">    </span><span class="kw">coord_fixed</span>()</a>
<a class="sourceLine" id="cb402-5" data-line-number="5"><span class="co">## Warning: Removed 239728 rows containing non-finite values (stat_binhex).</span></a></code></pre></div>
<p><img src="05-Bioinformatics_files/figure-html/unnamed-chunk-90-1.png" width="672" /></p>
<p>The plot shows a positive relation between expression of the two genes.</p>
<p>A major caveat to any conclusion of co-expression from the figure is that <em>covariates</em> may introduce spurious correlations. For instance, sequencing of one cell can be more efficient than sequencing of another, so <em>all</em> genes in one cell might appear to have higher expression than in another cell, but for completely un-interesting reasons (i.e., a technological artifact of sequencing efficiency). We would need to do a more sophisticated analysis to reach robust conclusions.</p>
</div>
</div>
<div id="day-33-friday-zoom-check-in" class="section level2">
<h2><span class="header-section-number">5.5</span> Day 33 (Friday) Zoom check-in</h2>
<p>Biological background</p>
<ul>
<li><p>RNA virus (<a href="https://en.wikipedia.org/wiki/File:Coronavirus_virion_structure.svg">source</a>, license: <a href="https://creativecommons.org/licenses/by-sa/4.0/deed.en">CC BY-SA 4.0</a>)</p>
<p><img src="images/512px-Coronavirus_virion_structure.svg.png" width="60%" /></p></li>
<li><p>Key domains (<a href="https://en.wikipedia.org/wiki/File:SARS-CoV_genome_organization.jpg">source</a>, license: <a href="https://creativecommons.org/licenses/by/2.0/deed.en">CC BY 2</a>)</p>
<p><img src="images/SARS-CoV_genome_organization.jpg" width="100%" /></p></li>
</ul>
<div id="monday-troubles-10-minutes" class="section level3">
<h3><span class="header-section-number">5.5.1</span> Monday troubles (10 minutes)</h3>
<p>Where we ran into trouble on Monday</p>
<ul>
<li><p>Load specific packages</p>
<div class="sourceCode" id="cb403"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb403-1" data-line-number="1"><span class="kw">library</span>(Biostrings)</a>
<a class="sourceLine" id="cb403-2" data-line-number="2"><span class="kw">options</span>(<span class="dt">Biostrings.coloring =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb403-3" data-line-number="3"><span class="kw">library</span>(GenomicRanges)</a></code></pre></div></li>
<li><p>DNA sequences and genomic ranges</p>
<div class="sourceCode" id="cb404"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb404-1" data-line-number="1">sequences &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dt">chr1 =</span> <span class="st">&quot;CTGACT&quot;</span>, <span class="dt">chr2 =</span> <span class="st">&quot;CTGGGAACT&quot;</span>)</a>
<a class="sourceLine" id="cb404-2" data-line-number="2">dna &lt;-<span class="st"> </span><span class="kw">DNAStringSet</span>(sequences)</a>
<a class="sourceLine" id="cb404-3" data-line-number="3"></a>
<a class="sourceLine" id="cb404-4" data-line-number="4">regions_of_interest &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;chr1:1-3&quot;</span>, <span class="st">&quot;chr1:3-6&quot;</span>, <span class="st">&quot;chr2:4-6&quot;</span>)</a>
<a class="sourceLine" id="cb404-5" data-line-number="5">roi &lt;-<span class="st"> </span><span class="kw">GRanges</span>(regions_of_interest)</a></code></pre></div></li>
<li><p>So far so good, but…</p>
<div class="sourceCode" id="cb405"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb405-1" data-line-number="1"><span class="op">&gt;</span><span class="st"> </span><span class="kw">getSeq</span>(dna, roi)</a>
<a class="sourceLine" id="cb405-2" data-line-number="2">Error <span class="cf">in</span> (<span class="cf">function</span> (classes, fdef, mtable)  <span class="op">:</span><span class="st"> </span></a>
<a class="sourceLine" id="cb405-3" data-line-number="3"><span class="st">  </span>unable to find an inherited method <span class="cf">for</span> <span class="cf">function</span> <span class="st">&#39;getSeq&#39;</span> <span class="cf">for</span> signature <span class="st">&#39;&quot;DNAStringSet&quot;&#39;</span></a></code></pre></div></li>
<li><p>What went wrong? The <code>getSeq()</code> <em>generic</em> is defined in <a href="https://bioconductor.org/packages/Biostrings">Biostrings</a>, but the <code>getSeq()</code> <em>method</em> <code>getSeq,DNAStringSet-method</code> is not! Hence ‘unable to find an inherited method … for signature …’.</p></li>
<li><p>Why did it work in the QuaRantine document? I had also loaded another package</p>
<div class="sourceCode" id="cb406"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb406-1" data-line-number="1"><span class="kw">library</span>(genbankr)</a></code></pre></div>
<ul>
<li><a href="https://bioconductor.org/packages/genbankr">genbankr</a> depends on the package <a href="https://bioconductor.org/packages/BSgenome">BSgenome</a></li>
<li><a href="https://bioconductor.org/packages/BSgenome">BSgenome</a> defines a method <code>getSeq,DNAStringSet-method</code>.</li>
<li><p>The generic in <a href="https://bioconductor.org/packages/Biostrings">Biostrings</a> now finds the method in <a href="https://bioconductor.org/packages/BSgenome">BSgenome</a>!</p>
<div class="sourceCode" id="cb407"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb407-1" data-line-number="1"><span class="kw">getSeq</span>(dna, roi)</a>
<a class="sourceLine" id="cb407-2" data-line-number="2"><span class="co">##   A DNAStringSet instance of length 3</span></a>
<a class="sourceLine" id="cb407-3" data-line-number="3"><span class="co">##     width seq</span></a>
<a class="sourceLine" id="cb407-4" data-line-number="4"><span class="co">## [1]     3 CTG</span></a>
<a class="sourceLine" id="cb407-5" data-line-number="5"><span class="co">## [2]     4 GACT</span></a>
<a class="sourceLine" id="cb407-6" data-line-number="6"><span class="co">## [3]     3 GGA</span></a></code></pre></div></li>
</ul></li>
</ul>
</div>
<div id="review-and-trouble-shoot-40-minutes-1" class="section level3">
<h3><span class="header-section-number">5.5.2</span> Review and trouble shoot (40 minutes)</h3>
<p><a href="https://bioconductor.org">Bioconductor</a></p>
<ul>
<li><p>Useful resource: statistical analysis and comprehension of high-throughput genomic data.</p></li>
<li><p><code>BiocManager::install()</code>, <code>BiocManager::version()</code>, <code>BiocManager::valid()</code></p></li>
<li><p>Key concepts: package, vignette, class, method.</p></li>
</ul>
<p>SARS-CoV-2 sequence data</p>
<ul>
<li><p>Querying Entrez for <a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_045512">NC_045512</a> and parsing GenBank records</p>
<div class="sourceCode" id="cb408"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb408-1" data-line-number="1"><span class="kw">library</span>(genbankr)</a>
<a class="sourceLine" id="cb408-2" data-line-number="2">gb &lt;-<span class="st"> </span><span class="kw">readGenBank</span>(<span class="kw">GBAccession</span>(<span class="st">&quot;NC_045512&quot;</span>))</a>
<a class="sourceLine" id="cb408-3" data-line-number="3"><span class="co">## No exons read from genbank file. Assuming sections of CDS are full exons</span></a>
<a class="sourceLine" id="cb408-4" data-line-number="4"><span class="co">## No transcript features (mRNA) found, using spans of CDSs</span></a>
<a class="sourceLine" id="cb408-5" data-line-number="5">gb</a>
<a class="sourceLine" id="cb408-6" data-line-number="6"><span class="co">## GenBank Annotations</span></a>
<a class="sourceLine" id="cb408-7" data-line-number="7"><span class="co">## Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome. </span></a>
<a class="sourceLine" id="cb408-8" data-line-number="8"><span class="co">## Accession: NC_045512</span></a>
<a class="sourceLine" id="cb408-9" data-line-number="9"><span class="co">## 1 Sequence(s) with total length length: 29903</span></a>
<a class="sourceLine" id="cb408-10" data-line-number="10"><span class="co">## 11 genes</span></a>
<a class="sourceLine" id="cb408-11" data-line-number="11"><span class="co">## 12 transcripts</span></a>
<a class="sourceLine" id="cb408-12" data-line-number="12"><span class="co">## 13 exons/cds elements</span></a>
<a class="sourceLine" id="cb408-13" data-line-number="13"><span class="co">## 0 variations</span></a>
<a class="sourceLine" id="cb408-14" data-line-number="14"><span class="co">## 34 other features</span></a></code></pre></div></li>
<li><p>Extract whole-genome sequence and annotation</p>
<div class="sourceCode" id="cb409"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb409-1" data-line-number="1">ref &lt;-<span class="st"> </span><span class="kw">getSeq</span>(gb)</a>
<a class="sourceLine" id="cb409-2" data-line-number="2">ref</a>
<a class="sourceLine" id="cb409-3" data-line-number="3"><span class="co">##   A DNAStringSet instance of length 1</span></a>
<a class="sourceLine" id="cb409-4" data-line-number="4"><span class="co">##     width seq                                               names               </span></a>
<a class="sourceLine" id="cb409-5" data-line-number="5"><span class="co">## [1] 29903 ATTAAAGGTTTATACCTTCCCAG...AAAAAAAAAAAAAAAAAAAAAAA Severe acute resp...</span></a>
<a class="sourceLine" id="cb409-6" data-line-number="6"></a>
<a class="sourceLine" id="cb409-7" data-line-number="7">cds &lt;-<span class="st"> </span><span class="kw">cds</span>(gb)</a>
<a class="sourceLine" id="cb409-8" data-line-number="8"><span class="kw">getSeq</span>(ref, <span class="kw">subset</span>(cds, gene <span class="op">==</span><span class="st"> &quot;S&quot;</span>))</a>
<a class="sourceLine" id="cb409-9" data-line-number="9"><span class="co">##   A DNAStringSet instance of length 1</span></a>
<a class="sourceLine" id="cb409-10" data-line-number="10"><span class="co">##     width seq</span></a>
<a class="sourceLine" id="cb409-11" data-line-number="11"><span class="co">## [1]  3822 ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCT...GTGCTCAAAGGAGTCAAATTACATTACACATAA</span></a></code></pre></div></li>
<li><p>Multiple alignment and phylogenetic tree construction</p>
<div class="sourceCode" id="cb410"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb410-1" data-line-number="1"><span class="kw">library</span>(DECIPHER)</a>
<a class="sourceLine" id="cb410-2" data-line-number="2">url &lt;-<span class="st"> &quot;https://raw.githubusercontent.com/mtmorgan/QuaRantine/master/assets/05-SARS-CoV2-S.fasta&quot;</span></a>
<a class="sourceLine" id="cb410-3" data-line-number="3">fasta_file &lt;-<span class="st"> </span><span class="kw">file.path</span>(workdir, <span class="kw">basename</span>(url))</a>
<a class="sourceLine" id="cb410-4" data-line-number="4"><span class="cf">if</span> ( <span class="op">!</span><span class="kw">file.exists</span>(fasta_file) )</a>
<a class="sourceLine" id="cb410-5" data-line-number="5">    <span class="kw">download.file</span>(url, fasta_file)</a>
<a class="sourceLine" id="cb410-6" data-line-number="6">dna &lt;-<span class="st"> </span><span class="kw">readDNAStringSet</span>(fasta_file)</a>
<a class="sourceLine" id="cb410-7" data-line-number="7">aa &lt;-<span class="st"> </span>dna <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">translate</span>(<span class="dt">if.fuzzy.codon =</span> <span class="st">&quot;solve&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">unique</span>()</a>
<a class="sourceLine" id="cb410-8" data-line-number="8">aln &lt;-<span class="st"> </span><span class="kw">AlignSeqs</span>(aa, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb410-9" data-line-number="9">dist &lt;-<span class="st"> </span><span class="kw">DistanceMatrix</span>(aln, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb410-10" data-line-number="10">dendrogram &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb410-11" data-line-number="11"><span class="st">    </span><span class="kw">IdClusters</span>(dist, <span class="dt">method =</span> <span class="st">&quot;NJ&quot;</span>, <span class="dt">type =</span> <span class="st">&quot;dendrogram&quot;</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a></code></pre></div></li>
<li><p>Save intermediate result and take a deep breath</p>
<div class="sourceCode" id="cb411"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb411-1" data-line-number="1">dendrogram_file &lt;-<span class="st"> </span><span class="kw">file.path</span>(workdir, <span class="st">&quot;SARS-CoV2-S.newick&quot;</span>)</a>
<a class="sourceLine" id="cb411-2" data-line-number="2"><span class="kw">WriteDendrogram</span>(dendrogram, dendrogram_file, <span class="dt">quoteLabels =</span> <span class="ot">FALSE</span>)</a></code></pre></div></li>
<li><p>Visualization</p>
<div class="sourceCode" id="cb412"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb412-1" data-line-number="1"><span class="kw">library</span>(ggtree)</a>
<a class="sourceLine" id="cb412-2" data-line-number="2">tree &lt;-<span class="st"> </span>treeio<span class="op">::</span><span class="kw">read.newick</span>(dendrogram_file)</a>
<a class="sourceLine" id="cb412-3" data-line-number="3"><span class="kw">ggtree</span>(tree) <span class="op">+</span></a>
<a class="sourceLine" id="cb412-4" data-line-number="4"><span class="st">    </span><span class="kw">geom_tiplab</span>(</a>
<a class="sourceLine" id="cb412-5" data-line-number="5">        <span class="kw">aes</span>(<span class="dt">subset =</span> branch.length <span class="op">&gt;</span><span class="st"> </span><span class="fl">.001</span>, <span class="dt">label =</span> label), </a>
<a class="sourceLine" id="cb412-6" data-line-number="6">        <span class="dt">size =</span> <span class="dv">3</span></a>
<a class="sourceLine" id="cb412-7" data-line-number="7">    ) <span class="op">+</span></a>
<a class="sourceLine" id="cb412-8" data-line-number="8"><span class="st">    </span>ggplot2<span class="op">::</span><span class="kw">ggtitle</span>(<span class="st">&quot;SARS-CoV-2 S-locus phylogeny; genbank, 10 May 2020&quot;</span>)</a></code></pre></div>
<p><img src="05-Bioinformatics_files/figure-html/unnamed-chunk-102-1.png" width="672" /></p></li>
</ul>
<p>Single cell expression</p>
<ul>
<li><p>Retrieve script and file to a local working directory</p>
<div class="sourceCode" id="cb413"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb413-1" data-line-number="1">workdir &lt;-<span class="st"> &quot;workdir&quot;</span></a>
<a class="sourceLine" id="cb413-2" data-line-number="2"><span class="cf">if</span> (<span class="op">!</span><span class="kw">dir.exists</span>(workdir))</a>
<a class="sourceLine" id="cb413-3" data-line-number="3">    <span class="kw">dir.create</span>(workdir)</a>
<a class="sourceLine" id="cb413-4" data-line-number="4"></a>
<a class="sourceLine" id="cb413-5" data-line-number="5">url &lt;-<span class="st"> &quot;https://raw.githubusercontent.com/mtmorgan/QuaRantine/master/assets/05-scRNAseq-import.R&quot;</span></a>
<a class="sourceLine" id="cb413-6" data-line-number="6">destination &lt;-<span class="st"> </span><span class="kw">file.path</span>(workdir, <span class="kw">basename</span>(url))</a>
<a class="sourceLine" id="cb413-7" data-line-number="7"><span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(destination))</a>
<a class="sourceLine" id="cb413-8" data-line-number="8">    <span class="kw">download.file</span>(url, destination)</a>
<a class="sourceLine" id="cb413-9" data-line-number="9"></a>
<a class="sourceLine" id="cb413-10" data-line-number="10">url &lt;-<span class="st"> &quot;https://github.com/mtmorgan/QuaRantine/raw/master/assets/05-internal_nonsmokerslung.h5ad&quot;</span></a>
<a class="sourceLine" id="cb413-11" data-line-number="11">h5_file &lt;-<span class="st"> </span><span class="kw">file.path</span>(workdir, <span class="kw">basename</span>(url))</a>
<a class="sourceLine" id="cb413-12" data-line-number="12"><span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(h5_file))</a>
<a class="sourceLine" id="cb413-13" data-line-number="13">    <span class="kw">download.file</span>(url, h5_file)</a></code></pre></div></li>
<li><p>Input and explore cell annotations and expression values</p>
<div class="sourceCode" id="cb414"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb414-1" data-line-number="1">obs &lt;-<span class="st"> </span><span class="kw">read_obs</span>(h5_file)      <span class="co"># tibble</span></a>
<a class="sourceLine" id="cb414-2" data-line-number="2">x &lt;-<span class="st"> </span><span class="kw">read_expression</span>(h5_file) <span class="co"># sparse matrix</span></a></code></pre></div></li>
<li><p>Coordinate cell annotation and expression values as <code>SingleCellExperiment</code></p>
<p><img src="images/SummarizedExperiment.png" width="60%" /></p>
<div class="sourceCode" id="cb415"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb415-1" data-line-number="1">sce &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(x, <span class="dt">colData =</span> obs)</a>
<a class="sourceLine" id="cb415-2" data-line-number="2">sce</a>
<a class="sourceLine" id="cb415-3" data-line-number="3"><span class="co">## class: SingleCellExperiment </span></a>
<a class="sourceLine" id="cb415-4" data-line-number="4"><span class="co">## dim: 22 240511 </span></a>
<a class="sourceLine" id="cb415-5" data-line-number="5"><span class="co">## metadata(0):</span></a>
<a class="sourceLine" id="cb415-6" data-line-number="6"><span class="co">## assays(1): &#39;&#39;</span></a>
<a class="sourceLine" id="cb415-7" data-line-number="7"><span class="co">## rownames(22): ACE2 TMPRSS2 ... CTSC CTSE</span></a>
<a class="sourceLine" id="cb415-8" data-line-number="8"><span class="co">## rowData names(0):</span></a>
<a class="sourceLine" id="cb415-9" data-line-number="9"><span class="co">## colnames(240511): 1B_cell57_HU28 1B_cell418_HU28 ...</span></a>
<a class="sourceLine" id="cb415-10" data-line-number="10"><span class="co">##   RP2_neg_cell6794209_HU52 RP2_neg_cell6794225_HU52</span></a>
<a class="sourceLine" id="cb415-11" data-line-number="11"><span class="co">## colData names(7): Age Cell Type ... Tissue cell_subset</span></a>
<a class="sourceLine" id="cb415-12" data-line-number="12"><span class="co">## reducedDimNames(0):</span></a>
<a class="sourceLine" id="cb415-13" data-line-number="13"><span class="co">## spikeNames(0):</span></a>
<a class="sourceLine" id="cb415-14" data-line-number="14"><span class="co">## altExpNames(0):</span></a></code></pre></div></li>
<li><p>Etc., e.g., cell subtype expression</p>
<div class="sourceCode" id="cb416"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb416-1" data-line-number="1"><span class="kw">library</span>(ggplot2)</a>
<a class="sourceLine" id="cb416-2" data-line-number="2"></a>
<a class="sourceLine" id="cb416-3" data-line-number="3">ace2 &lt;-<span class="st"> </span><span class="kw">subset</span>(sce, <span class="kw">rownames</span>(sce) <span class="op">%in%</span><span class="st"> &quot;ACE2&quot;</span>)</a>
<a class="sourceLine" id="cb416-4" data-line-number="4">tbl &lt;-<span class="st"> </span><span class="kw">tibble</span>(</a>
<a class="sourceLine" id="cb416-5" data-line-number="5">    <span class="dt">cell_subset =</span> ace2<span class="op">$</span>cell_subset,</a>
<a class="sourceLine" id="cb416-6" data-line-number="6">    <span class="dt">expression =</span> <span class="kw">as.vector</span>(<span class="kw">assay</span>(ace2))</a>
<a class="sourceLine" id="cb416-7" data-line-number="7">)</a>
<a class="sourceLine" id="cb416-8" data-line-number="8"></a>
<a class="sourceLine" id="cb416-9" data-line-number="9">tbl <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb416-10" data-line-number="10"><span class="st">    </span><span class="kw">filter</span>(expression <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb416-11" data-line-number="11"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(cell_subset, expression)) <span class="op">+</span></a>
<a class="sourceLine" id="cb416-12" data-line-number="12"><span class="st">        </span><span class="kw">geom_boxplot</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb416-13" data-line-number="13"><span class="st">        </span><span class="kw">coord_flip</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb416-14" data-line-number="14"><span class="st">        </span><span class="kw">ggtitle</span>(<span class="st">&quot;ACE2&quot;</span>)</a></code></pre></div>
<p><img src="05-Bioinformatics_files/figure-html/unnamed-chunk-107-1.png" width="672" /></p></li>
<li><p>Many more single cell resources described in <a href="https://osca.bioconductor.org/">Orchestrating Single Cell Analysis with <em>Bioconductor</em></a>.</p></li>
</ul>
</div>
<div id="this-weekend-10-minutes" class="section level3">
<h3><span class="header-section-number">5.5.3</span> This weekend (10 minutes)</h3>
<p>Saturday</p>
<ul>
<li>Tree visualization: color lineages based on observed attributes</li>
</ul>
<p>Sunday</p>
<ul>
<li>Package exploration: identifying solutions for your research challenges.</li>
</ul>
</div>
</div>
<div id="day-34-revising-tree-visualization" class="section level2">
<h2><span class="header-section-number">5.6</span> Day 34 Revising tree visualization</h2>
<p>The goal for today is to return to the Wednesday challenge: to color the S-locus gene phylogeny based on the country the lineage was sequenced in. The steps are as follows:</p>
<ul>
<li>Find nodes representing just a single sequence – ‘singletons’</li>
<li>Use GenBank country to identify which country these were sequenced from</li>
<li>Color lineages according to singleton country of origin</li>
</ul>
</div>
<div id="day-35-exploring-bioconductor" class="section level2">
<h2><span class="header-section-number">5.7</span> Day 35 Exploring <a href="https://bioconductor.org">Bioconductor</a></h2>
<p>The exciting part of science is that we are always on the edge – on our own edge of new discovery, where no one has gone before. We’ve barely scratched the surface of <em>Bioconductor</em> this week. Explore <a href="https://bioconductor.org/packages">available packages</a> for solutions that might be appropriate for <strong>your particular problem</strong>.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="four.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="six.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["QuaRantine.pdf", "QuaRantine.epub"],
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
